// @wireio/sdk-core@1.0.18/lib/api/v1/History.js
import { Checksum256, Int32, Name, PublicKey, UInt32 } from "../../chain";
import { GetActionsResponse, GetControlledAccountsResponse, GetKeyAccountsResponse, GetTransactionResponse } from "./Types";
export class HistoryAPI {
    client;
    constructor(client) {
        this.client = client;
    }
    async get_actions(accountName, pos, offset) {
        return this.client.call({
            path: "/v1/history/get_actions",
            params: {
                account_name: Name.from(accountName),
                pos: Int32.from(pos),
                offset: Int32.from(offset)
            },
            responseType: GetActionsResponse
        });
    }
    async get_transaction(id, options = {}) {
        return this.client.call({
            path: "/v1/history/get_transaction",
            params: {
                id: Checksum256.from(id),
                block_num_hint: options.blockNumHint && UInt32.from(options.blockNumHint),
                traces: options.excludeTraces === true ? false : undefined
            },
            responseType: GetTransactionResponse
        });
    }
    async get_key_accounts(publicKey) {
        return this.client.call({
            path: "/v1/history/get_key_accounts",
            params: { public_key: PublicKey.from(publicKey) },
            responseType: GetKeyAccountsResponse
        });
    }
    async get_controlled_accounts(controllingAccount) {
        return this.client.call({
            path: "/v1/history/get_controlled_accounts",
            params: { controlling_account: Name.from(controllingAccount) },
            responseType: GetControlledAccountsResponse
        });
    }
}
