// @wireio/sdk-core@1.0.18/lib/api/v1/Types.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import pako from "pako";
import { ABI, Action, Asset, Authority, Blob, BlockId, BlockTimestamp, Bytes, Checksum256, Float64, Int32, Int64, KeyWeight, Name, PermissionLevel, PublicKey, Signature, Struct, TimePoint, TimePointSec, Transaction, TransactionHeader, TransactionReceipt, UInt16, UInt32, UInt64, Weight } from "../../chain";
import { Serializer } from "../../serializer";
let AccountLinkedAction = class AccountLinkedAction extends Struct {
};
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], AccountLinkedAction.prototype, "account", void 0);
__decorate([
    Struct.field("name", { optional: true }),
    __metadata("design:type", Name)
], AccountLinkedAction.prototype, "action", void 0);
AccountLinkedAction = __decorate([
    Struct.type("account_linked_action")
], AccountLinkedAction);
export { AccountLinkedAction };
let AccountPermission = class AccountPermission extends Struct {
};
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], AccountPermission.prototype, "perm_name", void 0);
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], AccountPermission.prototype, "parent", void 0);
__decorate([
    Struct.field(Authority),
    __metadata("design:type", Authority)
], AccountPermission.prototype, "required_auth", void 0);
__decorate([
    Struct.field(AccountLinkedAction, { optional: true, array: true }),
    __metadata("design:type", Array)
], AccountPermission.prototype, "linked_actions", void 0);
AccountPermission = __decorate([
    Struct.type("account_permission")
], AccountPermission);
export { AccountPermission };
let AccountResourceLimit = class AccountResourceLimit extends Struct {
};
__decorate([
    Struct.field("int64"),
    __metadata("design:type", Int64)
], AccountResourceLimit.prototype, "used", void 0);
__decorate([
    Struct.field("int64"),
    __metadata("design:type", Int64)
], AccountResourceLimit.prototype, "available", void 0);
__decorate([
    Struct.field("int64"),
    __metadata("design:type", Int64)
], AccountResourceLimit.prototype, "max", void 0);
__decorate([
    Struct.field("time_point", { optional: true }),
    __metadata("design:type", TimePoint)
], AccountResourceLimit.prototype, "last_usage_update_time", void 0);
__decorate([
    Struct.field("int64", { optional: true }),
    __metadata("design:type", Int64)
], AccountResourceLimit.prototype, "current_used", void 0);
AccountResourceLimit = __decorate([
    Struct.type("account_resource_limit")
], AccountResourceLimit);
export { AccountResourceLimit };
let AccountTotalResources = class AccountTotalResources extends Struct {
};
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], AccountTotalResources.prototype, "owner", void 0);
__decorate([
    Struct.field("asset"),
    __metadata("design:type", Asset)
], AccountTotalResources.prototype, "net_weight", void 0);
__decorate([
    Struct.field("asset"),
    __metadata("design:type", Asset)
], AccountTotalResources.prototype, "cpu_weight", void 0);
__decorate([
    Struct.field("uint64"),
    __metadata("design:type", UInt64)
], AccountTotalResources.prototype, "ram_bytes", void 0);
AccountTotalResources = __decorate([
    Struct.type("account_total_resources")
], AccountTotalResources);
export { AccountTotalResources };
let AccountSelfDelegatedBandwidth = class AccountSelfDelegatedBandwidth extends Struct {
};
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], AccountSelfDelegatedBandwidth.prototype, "from", void 0);
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], AccountSelfDelegatedBandwidth.prototype, "to", void 0);
__decorate([
    Struct.field("asset"),
    __metadata("design:type", Asset)
], AccountSelfDelegatedBandwidth.prototype, "net_weight", void 0);
__decorate([
    Struct.field("asset"),
    __metadata("design:type", Asset)
], AccountSelfDelegatedBandwidth.prototype, "cpu_weight", void 0);
AccountSelfDelegatedBandwidth = __decorate([
    Struct.type("account_self_delegated_bandwidth")
], AccountSelfDelegatedBandwidth);
export { AccountSelfDelegatedBandwidth };
let AccountRefundRequest = class AccountRefundRequest extends Struct {
};
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], AccountRefundRequest.prototype, "owner", void 0);
__decorate([
    Struct.field("time_point"),
    __metadata("design:type", TimePoint)
], AccountRefundRequest.prototype, "request_time", void 0);
__decorate([
    Struct.field("asset"),
    __metadata("design:type", Asset)
], AccountRefundRequest.prototype, "net_amount", void 0);
__decorate([
    Struct.field("asset"),
    __metadata("design:type", Asset)
], AccountRefundRequest.prototype, "cpu_amount", void 0);
AccountRefundRequest = __decorate([
    Struct.type("account_refund_request")
], AccountRefundRequest);
export { AccountRefundRequest };
let AccountVoterInfo = class AccountVoterInfo extends Struct {
    staked;
    flags1;
};
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], AccountVoterInfo.prototype, "owner", void 0);
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], AccountVoterInfo.prototype, "proxy", void 0);
__decorate([
    Struct.field("name", { array: true }),
    __metadata("design:type", Array)
], AccountVoterInfo.prototype, "producers", void 0);
__decorate([
    Struct.field("int64", { optional: true }),
    __metadata("design:type", Int64)
], AccountVoterInfo.prototype, "staked", void 0);
__decorate([
    Struct.field("float64"),
    __metadata("design:type", Float64)
], AccountVoterInfo.prototype, "last_vote_weight", void 0);
__decorate([
    Struct.field("float64"),
    __metadata("design:type", Float64)
], AccountVoterInfo.prototype, "proxied_vote_weight", void 0);
__decorate([
    Struct.field("bool"),
    __metadata("design:type", Boolean)
], AccountVoterInfo.prototype, "is_proxy", void 0);
__decorate([
    Struct.field("uint32", { optional: true }),
    __metadata("design:type", UInt32)
], AccountVoterInfo.prototype, "flags1", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], AccountVoterInfo.prototype, "reserved2", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], AccountVoterInfo.prototype, "reserved3", void 0);
AccountVoterInfo = __decorate([
    Struct.type("account_voter_info")
], AccountVoterInfo);
export { AccountVoterInfo };
let AccountRexInfoMaturities = class AccountRexInfoMaturities extends Struct {
    /** Expected results from after SYSIO.Contracts v1.9.0 */
    key;
    value;
    /** Expected results from before SYSIO.Contracts v1.9.0 */
    first;
    second;
};
__decorate([
    Struct.field("time_point", { optional: true }),
    __metadata("design:type", TimePoint)
], AccountRexInfoMaturities.prototype, "key", void 0);
__decorate([
    Struct.field("int64", { optional: true }),
    __metadata("design:type", Int64
    /** Expected results from before SYSIO.Contracts v1.9.0 */
    )
], AccountRexInfoMaturities.prototype, "value", void 0);
__decorate([
    Struct.field("time_point", { optional: true }),
    __metadata("design:type", TimePoint)
], AccountRexInfoMaturities.prototype, "first", void 0);
__decorate([
    Struct.field("int64", { optional: true }),
    __metadata("design:type", Int64)
], AccountRexInfoMaturities.prototype, "second", void 0);
AccountRexInfoMaturities = __decorate([
    Struct.type("account_rex_info_maturities")
], AccountRexInfoMaturities);
export { AccountRexInfoMaturities };
let AccountRexInfo = class AccountRexInfo extends Struct {
};
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], AccountRexInfo.prototype, "version", void 0);
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], AccountRexInfo.prototype, "owner", void 0);
__decorate([
    Struct.field("asset"),
    __metadata("design:type", Asset)
], AccountRexInfo.prototype, "vote_stake", void 0);
__decorate([
    Struct.field("asset"),
    __metadata("design:type", Asset)
], AccountRexInfo.prototype, "rex_balance", void 0);
__decorate([
    Struct.field("int64"),
    __metadata("design:type", Int64)
], AccountRexInfo.prototype, "matured_rex", void 0);
__decorate([
    Struct.field(AccountRexInfoMaturities, { array: true }),
    __metadata("design:type", Array)
], AccountRexInfo.prototype, "rex_maturities", void 0);
AccountRexInfo = __decorate([
    Struct.type("account_rex_info")
], AccountRexInfo);
export { AccountRexInfo };
let GetRawAbiResponse = class GetRawAbiResponse extends Struct {
};
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], GetRawAbiResponse.prototype, "account_name", void 0);
__decorate([
    Struct.field("checksum256"),
    __metadata("design:type", Checksum256)
], GetRawAbiResponse.prototype, "code_hash", void 0);
__decorate([
    Struct.field("checksum256"),
    __metadata("design:type", Checksum256)
], GetRawAbiResponse.prototype, "abi_hash", void 0);
__decorate([
    Struct.field(Blob),
    __metadata("design:type", Blob)
], GetRawAbiResponse.prototype, "abi", void 0);
GetRawAbiResponse = __decorate([
    Struct.type("get_raw_abi_response")
], GetRawAbiResponse);
export { GetRawAbiResponse };
let AccountObject = class AccountObject extends Struct {
    /** Account core token balance */
    core_liquid_balance;
    self_delegated_bandwidth;
    refund_request;
    voter_info;
    rex_info;
    getPermission(permission) {
        const name = Name.from(permission);
        const match = this.permissions.find(p => p.perm_name.equals(name));
        if (!match) {
            throw new Error(`Unknown permission ${name} on account ${this.account_name}.`);
        }
        return match;
    }
};
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name
    /** Highest block number on the chain */
    )
], AccountObject.prototype, "account_name", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32
    /** Highest block unix timestamp. */
    )
], AccountObject.prototype, "head_block_num", void 0);
__decorate([
    Struct.field("time_point"),
    __metadata("design:type", TimePoint
    /** Indicator of if this is a privileged system account */
    )
], AccountObject.prototype, "head_block_time", void 0);
__decorate([
    Struct.field("bool"),
    __metadata("design:type", Boolean)
], AccountObject.prototype, "privileged", void 0);
__decorate([
    Struct.field("time_point"),
    __metadata("design:type", TimePoint
    /** Account created as unix timestamp. */
    )
], AccountObject.prototype, "last_code_update", void 0);
__decorate([
    Struct.field("time_point"),
    __metadata("design:type", TimePoint
    /** Account core token balance */
    )
], AccountObject.prototype, "created", void 0);
__decorate([
    Struct.field("asset?"),
    __metadata("design:type", Asset)
], AccountObject.prototype, "core_liquid_balance", void 0);
__decorate([
    Struct.field("int64"),
    __metadata("design:type", Int64)
], AccountObject.prototype, "ram_quota", void 0);
__decorate([
    Struct.field("int64"),
    __metadata("design:type", Int64)
], AccountObject.prototype, "net_weight", void 0);
__decorate([
    Struct.field("int64"),
    __metadata("design:type", Int64)
], AccountObject.prototype, "cpu_weight", void 0);
__decorate([
    Struct.field(AccountResourceLimit),
    __metadata("design:type", AccountResourceLimit)
], AccountObject.prototype, "net_limit", void 0);
__decorate([
    Struct.field(AccountResourceLimit),
    __metadata("design:type", AccountResourceLimit)
], AccountObject.prototype, "cpu_limit", void 0);
__decorate([
    Struct.field(AccountResourceLimit, { optional: true }),
    __metadata("design:type", AccountResourceLimit)
], AccountObject.prototype, "subjective_cpu_bill_limit", void 0);
__decorate([
    Struct.field("uint64"),
    __metadata("design:type", UInt64)
], AccountObject.prototype, "ram_usage", void 0);
__decorate([
    Struct.field(AccountPermission, { array: true }),
    __metadata("design:type", Array)
], AccountObject.prototype, "permissions", void 0);
__decorate([
    Struct.field(AccountTotalResources, { optional: true }),
    __metadata("design:type", AccountTotalResources)
], AccountObject.prototype, "total_resources", void 0);
__decorate([
    Struct.field(AccountSelfDelegatedBandwidth, { optional: true }),
    __metadata("design:type", AccountSelfDelegatedBandwidth)
], AccountObject.prototype, "self_delegated_bandwidth", void 0);
__decorate([
    Struct.field(AccountRefundRequest, { optional: true }),
    __metadata("design:type", AccountRefundRequest)
], AccountObject.prototype, "refund_request", void 0);
__decorate([
    Struct.field(AccountVoterInfo, { optional: true }),
    __metadata("design:type", AccountVoterInfo)
], AccountObject.prototype, "voter_info", void 0);
__decorate([
    Struct.field(AccountRexInfo, { optional: true }),
    __metadata("design:type", AccountRexInfo)
], AccountObject.prototype, "rex_info", void 0);
AccountObject = __decorate([
    Struct.type("account_object")
], AccountObject);
export { AccountObject };
let AccountByAuthorizersRow = class AccountByAuthorizersRow extends Struct {
};
__decorate([
    Struct.field(Name),
    __metadata("design:type", Name)
], AccountByAuthorizersRow.prototype, "account_name", void 0);
__decorate([
    Struct.field(Name),
    __metadata("design:type", Name)
], AccountByAuthorizersRow.prototype, "permission_name", void 0);
__decorate([
    Struct.field(PublicKey, { optional: true }),
    __metadata("design:type", PublicKey)
], AccountByAuthorizersRow.prototype, "authorizing_key", void 0);
__decorate([
    Struct.field(PermissionLevel, { optional: true }),
    __metadata("design:type", PermissionLevel)
], AccountByAuthorizersRow.prototype, "authorizing_account", void 0);
__decorate([
    Struct.field(Weight),
    __metadata("design:type", Weight)
], AccountByAuthorizersRow.prototype, "weight", void 0);
__decorate([
    Struct.field(UInt32),
    __metadata("design:type", UInt32)
], AccountByAuthorizersRow.prototype, "threshold", void 0);
AccountByAuthorizersRow = __decorate([
    Struct.type("account_by_authorizers_row")
], AccountByAuthorizersRow);
export { AccountByAuthorizersRow };
let AccountsByAuthorizers = class AccountsByAuthorizers extends Struct {
};
__decorate([
    Struct.field(AccountByAuthorizersRow, { array: true }),
    __metadata("design:type", Array)
], AccountsByAuthorizers.prototype, "accounts", void 0);
AccountsByAuthorizers = __decorate([
    Struct.type("account_by_authorizers")
], AccountsByAuthorizers);
export { AccountsByAuthorizers };
let NewProducersEntry = class NewProducersEntry extends Struct {
};
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], NewProducersEntry.prototype, "producer_name", void 0);
__decorate([
    Struct.field("public_key"),
    __metadata("design:type", PublicKey)
], NewProducersEntry.prototype, "block_signing_key", void 0);
NewProducersEntry = __decorate([
    Struct.type("new_producers_entry")
], NewProducersEntry);
export { NewProducersEntry };
let NewProducers = class NewProducers extends Struct {
};
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], NewProducers.prototype, "version", void 0);
__decorate([
    Struct.field(NewProducersEntry, { array: true }),
    __metadata("design:type", NewProducersEntry)
], NewProducers.prototype, "producers", void 0);
NewProducers = __decorate([
    Struct.type("new_producers")
], NewProducers);
export { NewProducers };
let BlockExtension = class BlockExtension extends Struct {
};
__decorate([
    Struct.field("uint16"),
    __metadata("design:type", UInt16)
], BlockExtension.prototype, "type", void 0);
__decorate([
    Struct.field("bytes"),
    __metadata("design:type", Bytes)
], BlockExtension.prototype, "data", void 0);
BlockExtension = __decorate([
    Struct.type("block_extension")
], BlockExtension);
export { BlockExtension };
let HeaderExtension = class HeaderExtension extends Struct {
};
__decorate([
    Struct.field("uint16"),
    __metadata("design:type", UInt16)
], HeaderExtension.prototype, "type", void 0);
__decorate([
    Struct.field("bytes"),
    __metadata("design:type", Bytes)
], HeaderExtension.prototype, "data", void 0);
HeaderExtension = __decorate([
    Struct.type("header_extension")
], HeaderExtension);
export { HeaderExtension };
// fc "mutable variant" returned by get_block api
export class TrxVariant {
    id;
    extra;
    static abiName = "trx_variant";
    static from(data) {
        let id;
        let extra;
        if (typeof data === "string") {
            id = Checksum256.from(data);
            extra = {};
        }
        else {
            id = Checksum256.from(data.id);
            extra = data;
        }
        return new this(id, extra);
    }
    constructor(id, extra) {
        this.id = id;
        this.extra = extra;
    }
    get transaction() {
        if (this.extra.packed_trx) {
            switch (this.extra.compression) {
                case "zlib": {
                    const inflated = pako.inflate(Bytes.from(this.extra.packed_trx, "hex").array);
                    return Serializer.decode({ data: inflated, type: Transaction });
                }
                case "none": {
                    return Serializer.decode({
                        data: this.extra.packed_trx,
                        type: Transaction
                    });
                }
                default: {
                    throw new Error(`Unsupported compression type ${this.extra.compression}`);
                }
            }
        }
    }
    get signatures() {
        if (this.extra.signatures) {
            return this.extra.signatures.map(Signature.from);
        }
    }
    equals(other) {
        return this.id.equals(other.id);
    }
    toJSON() {
        return this.id;
    }
}
let GetBlockResponseTransactionReceipt = class GetBlockResponseTransactionReceipt extends TransactionReceipt {
    // NOTE: Sometimes the trx field is a raw {array: []} object containing the ID.
    // If so, do as follows to convert to hex trxId:
    // const vals = Object.values(trx.array) as number[];
    // const rawBytes = Uint8Array.from(vals);
    // const hexId = Buffer.from(rawBytes).toString('hex')
    get id() {
        return this.trx.id;
    }
};
__decorate([
    Struct.field(TrxVariant),
    __metadata("design:type", TrxVariant
    // NOTE: Sometimes the trx field is a raw {array: []} object containing the ID.
    // If so, do as follows to convert to hex trxId:
    // const vals = Object.values(trx.array) as number[];
    // const rawBytes = Uint8Array.from(vals);
    // const hexId = Buffer.from(rawBytes).toString('hex')
    )
], GetBlockResponseTransactionReceipt.prototype, "trx", void 0);
GetBlockResponseTransactionReceipt = __decorate([
    Struct.type("get_block_response_receipt")
], GetBlockResponseTransactionReceipt);
export { GetBlockResponseTransactionReceipt };
let GetBlockResponse = class GetBlockResponse extends Struct {
    new_producers;
    header_extensions;
    new_protocol_features;
};
__decorate([
    Struct.field("time_point"),
    __metadata("design:type", TimePoint)
], GetBlockResponse.prototype, "timestamp", void 0);
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], GetBlockResponse.prototype, "producer", void 0);
__decorate([
    Struct.field("uint16"),
    __metadata("design:type", UInt16)
], GetBlockResponse.prototype, "confirmed", void 0);
__decorate([
    Struct.field(BlockId),
    __metadata("design:type", BlockId)
], GetBlockResponse.prototype, "previous", void 0);
__decorate([
    Struct.field("checksum256"),
    __metadata("design:type", Checksum256)
], GetBlockResponse.prototype, "transaction_mroot", void 0);
__decorate([
    Struct.field("checksum256"),
    __metadata("design:type", Checksum256)
], GetBlockResponse.prototype, "action_mroot", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], GetBlockResponse.prototype, "schedule_version", void 0);
__decorate([
    Struct.field(NewProducers, { optional: true }),
    __metadata("design:type", NewProducers)
], GetBlockResponse.prototype, "new_producers", void 0);
__decorate([
    Struct.field("header_extension", { optional: true }),
    __metadata("design:type", Array)
], GetBlockResponse.prototype, "header_extensions", void 0);
__decorate([
    Struct.field("any", { optional: true }),
    __metadata("design:type", Object)
], GetBlockResponse.prototype, "new_protocol_features", void 0);
__decorate([
    Struct.field("signature"),
    __metadata("design:type", Signature)
], GetBlockResponse.prototype, "producer_signature", void 0);
__decorate([
    Struct.field(GetBlockResponseTransactionReceipt, { array: true }),
    __metadata("design:type", Array)
], GetBlockResponse.prototype, "transactions", void 0);
__decorate([
    Struct.field("block_extension", { optional: true }),
    __metadata("design:type", Array)
], GetBlockResponse.prototype, "block_extensions", void 0);
__decorate([
    Struct.field(BlockId),
    __metadata("design:type", BlockId)
], GetBlockResponse.prototype, "id", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], GetBlockResponse.prototype, "block_num", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], GetBlockResponse.prototype, "ref_block_prefix", void 0);
GetBlockResponse = __decorate([
    Struct.type("get_block_response")
], GetBlockResponse);
export { GetBlockResponse };
let GetBlockInfoResponse = class GetBlockInfoResponse extends Struct {
};
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], GetBlockInfoResponse.prototype, "block_num", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt16)
], GetBlockInfoResponse.prototype, "ref_block_num", void 0);
__decorate([
    Struct.field(BlockId),
    __metadata("design:type", BlockId)
], GetBlockInfoResponse.prototype, "id", void 0);
__decorate([
    Struct.field("time_point"),
    __metadata("design:type", TimePoint)
], GetBlockInfoResponse.prototype, "timestamp", void 0);
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], GetBlockInfoResponse.prototype, "producer", void 0);
__decorate([
    Struct.field("uint16"),
    __metadata("design:type", UInt16)
], GetBlockInfoResponse.prototype, "confirmed", void 0);
__decorate([
    Struct.field(BlockId),
    __metadata("design:type", BlockId)
], GetBlockInfoResponse.prototype, "previous", void 0);
__decorate([
    Struct.field("checksum256"),
    __metadata("design:type", Checksum256)
], GetBlockInfoResponse.prototype, "transaction_mroot", void 0);
__decorate([
    Struct.field("checksum256"),
    __metadata("design:type", Checksum256)
], GetBlockInfoResponse.prototype, "action_mroot", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], GetBlockInfoResponse.prototype, "schedule_version", void 0);
__decorate([
    Struct.field("signature"),
    __metadata("design:type", Signature)
], GetBlockInfoResponse.prototype, "producer_signature", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], GetBlockInfoResponse.prototype, "ref_block_prefix", void 0);
GetBlockInfoResponse = __decorate([
    Struct.type("get_block_response")
], GetBlockInfoResponse);
export { GetBlockInfoResponse };
let ActiveScheduleProducerAuthority = class ActiveScheduleProducerAuthority extends Struct {
};
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], ActiveScheduleProducerAuthority.prototype, "producer_name", void 0);
__decorate([
    Struct.field("any"),
    __metadata("design:type", Object)
], ActiveScheduleProducerAuthority.prototype, "authority", void 0);
ActiveScheduleProducerAuthority = __decorate([
    Struct.type("active_schedule_producer_authority")
], ActiveScheduleProducerAuthority);
export { ActiveScheduleProducerAuthority };
let ActiveScheduleProducer = class ActiveScheduleProducer extends Struct {
};
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], ActiveScheduleProducer.prototype, "producer_name", void 0);
__decorate([
    Struct.field(ActiveScheduleProducerAuthority),
    __metadata("design:type", ActiveScheduleProducerAuthority)
], ActiveScheduleProducer.prototype, "authority", void 0);
ActiveScheduleProducer = __decorate([
    Struct.type("active_schedule_producer")
], ActiveScheduleProducer);
export { ActiveScheduleProducer };
let ActiveSchedule = class ActiveSchedule extends Struct {
};
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], ActiveSchedule.prototype, "version", void 0);
__decorate([
    Struct.field(ActiveScheduleProducer, { array: true }),
    __metadata("design:type", Array)
], ActiveSchedule.prototype, "producers", void 0);
ActiveSchedule = __decorate([
    Struct.type("active_schedule")
], ActiveSchedule);
export { ActiveSchedule };
let BlockStateHeader = class BlockStateHeader extends Struct {
};
__decorate([
    Struct.field("time_point"),
    __metadata("design:type", TimePoint)
], BlockStateHeader.prototype, "timestamp", void 0);
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], BlockStateHeader.prototype, "producer", void 0);
__decorate([
    Struct.field("uint16"),
    __metadata("design:type", UInt16)
], BlockStateHeader.prototype, "confirmed", void 0);
__decorate([
    Struct.field(BlockId),
    __metadata("design:type", BlockId)
], BlockStateHeader.prototype, "previous", void 0);
__decorate([
    Struct.field("checksum256"),
    __metadata("design:type", Checksum256)
], BlockStateHeader.prototype, "transaction_mroot", void 0);
__decorate([
    Struct.field("checksum256"),
    __metadata("design:type", Checksum256)
], BlockStateHeader.prototype, "action_mroot", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32
    // @Struct.field(HeaderExtension, {array: true, optional: true})
    // header_extensions?: HeaderExtension[];
    )
], BlockStateHeader.prototype, "schedule_version", void 0);
__decorate([
    Struct.field("any", { array: true, optional: true }),
    __metadata("design:type", Array)
], BlockStateHeader.prototype, "header_extensions", void 0);
__decorate([
    Struct.field("signature"),
    __metadata("design:type", Signature)
], BlockStateHeader.prototype, "producer_signature", void 0);
BlockStateHeader = __decorate([
    Struct.type("block_state_header")
], BlockStateHeader);
export { BlockStateHeader };
let GetBlockHeaderStateResponse = class GetBlockHeaderStateResponse extends Struct {
};
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], GetBlockHeaderStateResponse.prototype, "block_num", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], GetBlockHeaderStateResponse.prototype, "dpos_proposed_irreversible_blocknum", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], GetBlockHeaderStateResponse.prototype, "dpos_irreversible_blocknum", void 0);
__decorate([
    Struct.field(BlockId),
    __metadata("design:type", BlockId)
], GetBlockHeaderStateResponse.prototype, "id", void 0);
__decorate([
    Struct.field(BlockStateHeader),
    __metadata("design:type", BlockStateHeader
    /** Unstructured any fields specific to header state calls */
    )
], GetBlockHeaderStateResponse.prototype, "header", void 0);
__decorate([
    Struct.field("any"),
    __metadata("design:type", Object)
], GetBlockHeaderStateResponse.prototype, "active_schedule", void 0);
__decorate([
    Struct.field("any"),
    __metadata("design:type", Object)
], GetBlockHeaderStateResponse.prototype, "blockroot_merkle", void 0);
__decorate([
    Struct.field("any"),
    __metadata("design:type", Object)
], GetBlockHeaderStateResponse.prototype, "producer_to_last_produced", void 0);
__decorate([
    Struct.field("any"),
    __metadata("design:type", Object)
], GetBlockHeaderStateResponse.prototype, "producer_to_last_implied_irb", void 0);
__decorate([
    Struct.field("any"),
    __metadata("design:type", Object)
], GetBlockHeaderStateResponse.prototype, "valid_block_signing_authority", void 0);
__decorate([
    Struct.field("any"),
    __metadata("design:type", Object)
], GetBlockHeaderStateResponse.prototype, "confirm_count", void 0);
__decorate([
    Struct.field("any"),
    __metadata("design:type", Object)
], GetBlockHeaderStateResponse.prototype, "pending_schedule", void 0);
__decorate([
    Struct.field("any"),
    __metadata("design:type", Object)
], GetBlockHeaderStateResponse.prototype, "activated_protocol_features", void 0);
__decorate([
    Struct.field("any"),
    __metadata("design:type", Object)
], GetBlockHeaderStateResponse.prototype, "additional_signatures", void 0);
GetBlockHeaderStateResponse = __decorate([
    Struct.type("get_block_header_state_response")
], GetBlockHeaderStateResponse);
export { GetBlockHeaderStateResponse };
let GetInfoResponse = class GetInfoResponse extends Struct {
    /** String representation of server version. */
    server_version_string;
    /** Sequential block number representing the best known head in the fork database tree. */
    fork_db_head_block_num;
    /** Hash representing the best known head in the fork database tree. */
    fork_db_head_block_id;
    getTransactionHeader(secondsAhead = 120) {
        const expiration = TimePointSec.fromMilliseconds(this.head_block_time.toMilliseconds() + secondsAhead * 1000);
        const id = this.last_irreversible_block_id;
        const prefixArray = id.array.subarray(8, 12);
        const prefix = new Uint32Array(prefixArray.buffer, prefixArray.byteOffset, 1)[0];
        return TransactionHeader.from({
            expiration,
            ref_block_num: Number(this.last_irreversible_block_num) & 0xffff,
            ref_block_prefix: prefix
        });
    }
};
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], GetInfoResponse.prototype, "server_version", void 0);
__decorate([
    Struct.field("checksum256"),
    __metadata("design:type", Checksum256
    /** Highest block number on the chain */
    )
], GetInfoResponse.prototype, "chain_id", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32
    /** Highest block number on the chain that has been irreversibly applied to state. */
    )
], GetInfoResponse.prototype, "head_block_num", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32
    /** Highest block ID on the chain that has been irreversibly applied to state. */
    )
], GetInfoResponse.prototype, "last_irreversible_block_num", void 0);
__decorate([
    Struct.field(BlockId),
    __metadata("design:type", BlockId
    /** Highest block ID on the chain. */
    )
], GetInfoResponse.prototype, "last_irreversible_block_id", void 0);
__decorate([
    Struct.field(BlockId),
    __metadata("design:type", BlockId
    /** Highest block unix timestamp. */
    )
], GetInfoResponse.prototype, "head_block_id", void 0);
__decorate([
    Struct.field("time_point"),
    __metadata("design:type", TimePoint
    /** Producer that signed the highest block (head block). */
    )
], GetInfoResponse.prototype, "head_block_time", void 0);
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name
    /** CPU limit calculated after each block is produced, approximately 1000 times `blockCpuLimit`. */
    )
], GetInfoResponse.prototype, "head_block_producer", void 0);
__decorate([
    Struct.field("uint64"),
    __metadata("design:type", UInt64
    /** NET limit calculated after each block is produced, approximately 1000 times `blockNetLimit`. */
    )
], GetInfoResponse.prototype, "virtual_block_cpu_limit", void 0);
__decorate([
    Struct.field("uint64"),
    __metadata("design:type", UInt64
    /** Actual maximum CPU limit. */
    )
], GetInfoResponse.prototype, "virtual_block_net_limit", void 0);
__decorate([
    Struct.field("uint64"),
    __metadata("design:type", UInt64
    /** Actual maximum NET limit. */
    )
], GetInfoResponse.prototype, "block_cpu_limit", void 0);
__decorate([
    Struct.field("uint64"),
    __metadata("design:type", UInt64
    /** String representation of server version. */
    )
], GetInfoResponse.prototype, "block_net_limit", void 0);
__decorate([
    Struct.field("string?"),
    __metadata("design:type", String)
], GetInfoResponse.prototype, "server_version_string", void 0);
__decorate([
    Struct.field("uint32?"),
    __metadata("design:type", UInt32
    /** Hash representing the best known head in the fork database tree. */
    )
], GetInfoResponse.prototype, "fork_db_head_block_num", void 0);
__decorate([
    Struct.field("block_id_type?"),
    __metadata("design:type", BlockId)
], GetInfoResponse.prototype, "fork_db_head_block_id", void 0);
GetInfoResponse = __decorate([
    Struct.type("get_info_response")
], GetInfoResponse);
export { GetInfoResponse };
let GetTableByScopeResponseRow = class GetTableByScopeResponseRow extends Struct {
};
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], GetTableByScopeResponseRow.prototype, "code", void 0);
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], GetTableByScopeResponseRow.prototype, "scope", void 0);
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], GetTableByScopeResponseRow.prototype, "table", void 0);
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], GetTableByScopeResponseRow.prototype, "payer", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], GetTableByScopeResponseRow.prototype, "count", void 0);
GetTableByScopeResponseRow = __decorate([
    Struct.type("get_table_by_scope_response_row")
], GetTableByScopeResponseRow);
export { GetTableByScopeResponseRow };
let GetTableByScopeResponse = class GetTableByScopeResponse extends Struct {
};
__decorate([
    Struct.field(GetTableByScopeResponseRow, { array: true }),
    __metadata("design:type", Array)
], GetTableByScopeResponse.prototype, "rows", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], GetTableByScopeResponse.prototype, "more", void 0);
GetTableByScopeResponse = __decorate([
    Struct.type("get_table_by_scope_response")
], GetTableByScopeResponse);
export { GetTableByScopeResponse };
let OrderedActionsResult = class OrderedActionsResult extends Struct {
    action_trace;
    irrevirsible;
};
__decorate([
    Struct.field(UInt64),
    __metadata("design:type", UInt64)
], OrderedActionsResult.prototype, "global_action_seq", void 0);
__decorate([
    Struct.field(Int64),
    __metadata("design:type", Int64)
], OrderedActionsResult.prototype, "account_action_seq", void 0);
__decorate([
    Struct.field(UInt32),
    __metadata("design:type", UInt32)
], OrderedActionsResult.prototype, "block_num", void 0);
__decorate([
    Struct.field(BlockTimestamp),
    __metadata("design:type", BlockTimestamp)
], OrderedActionsResult.prototype, "block_time", void 0);
__decorate([
    Struct.field("any"),
    __metadata("design:type", Object)
], OrderedActionsResult.prototype, "action_trace", void 0);
__decorate([
    Struct.field("boolean?"),
    __metadata("design:type", Boolean)
], OrderedActionsResult.prototype, "irrevirsible", void 0);
OrderedActionsResult = __decorate([
    Struct.type("ordered_action_result")
], OrderedActionsResult);
export { OrderedActionsResult };
let GetActionsResponse = class GetActionsResponse extends Struct {
    time_limit_exceeded_error;
};
__decorate([
    Struct.field(OrderedActionsResult, { array: true }),
    __metadata("design:type", Array)
], GetActionsResponse.prototype, "actions", void 0);
__decorate([
    Struct.field(Int32),
    __metadata("design:type", Int32)
], GetActionsResponse.prototype, "last_irreversible_block", void 0);
__decorate([
    Struct.field(Int32),
    __metadata("design:type", Int32)
], GetActionsResponse.prototype, "head_block_num", void 0);
__decorate([
    Struct.field("boolean?"),
    __metadata("design:type", Boolean)
], GetActionsResponse.prototype, "time_limit_exceeded_error", void 0);
GetActionsResponse = __decorate([
    Struct.type("get_actions_response")
], GetActionsResponse);
export { GetActionsResponse };
let TransactionAuthSequence = class TransactionAuthSequence extends Struct {
};
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], TransactionAuthSequence.prototype, "account", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], TransactionAuthSequence.prototype, "sequence", void 0);
TransactionAuthSequence = __decorate([
    Struct.type("transaction_auth_sequence")
], TransactionAuthSequence);
export { TransactionAuthSequence };
let TransactionTraceReceipt = class TransactionTraceReceipt extends Struct {
};
__decorate([
    Struct.field(Checksum256),
    __metadata("design:type", Checksum256)
], TransactionTraceReceipt.prototype, "act_digest", void 0);
__decorate([
    Struct.field(TransactionAuthSequence),
    __metadata("design:type", Array)
], TransactionTraceReceipt.prototype, "auth_sequence", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], TransactionTraceReceipt.prototype, "global_sequence", void 0);
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], TransactionTraceReceipt.prototype, "receiver", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], TransactionTraceReceipt.prototype, "recv_sequence", void 0);
TransactionTraceReceipt = __decorate([
    Struct.type("transaction_trace_receipt")
], TransactionTraceReceipt);
export { TransactionTraceReceipt };
let TransactionTrace = class TransactionTrace extends Struct {
};
__decorate([
    Struct.field("any?"),
    __metadata("design:type", Array)
], TransactionTrace.prototype, "account_ram_deltas", void 0);
__decorate([
    Struct.field(Action),
    __metadata("design:type", Action)
], TransactionTrace.prototype, "act", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], TransactionTrace.prototype, "block_num", void 0);
__decorate([
    Struct.field(BlockTimestamp),
    __metadata("design:type", BlockTimestamp)
], TransactionTrace.prototype, "block_time", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], TransactionTrace.prototype, "console", void 0);
__decorate([
    Struct.field("boolean"),
    __metadata("design:type", Boolean)
], TransactionTrace.prototype, "context_free", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], TransactionTrace.prototype, "elapsed", void 0);
__decorate([
    Struct.field("any?"),
    __metadata("design:type", Object)
], TransactionTrace.prototype, "except", void 0);
__decorate([
    Struct.field(TransactionTrace),
    __metadata("design:type", Array)
], TransactionTrace.prototype, "inline_traces", void 0);
__decorate([
    Struct.field(Checksum256),
    __metadata("design:type", Checksum256)
], TransactionTrace.prototype, "producer_block_id", void 0);
__decorate([
    Struct.field(TransactionTraceReceipt),
    __metadata("design:type", TransactionTraceReceipt)
], TransactionTrace.prototype, "receipt", void 0);
__decorate([
    Struct.field(Checksum256),
    __metadata("design:type", Checksum256)
], TransactionTrace.prototype, "trx_id", void 0);
TransactionTrace = __decorate([
    Struct.type("transaction_trace")
], TransactionTrace);
export { TransactionTrace };
let Trx = class Trx extends Struct {
};
__decorate([
    Struct.field("any"),
    __metadata("design:type", Array)
], Trx.prototype, "actions", void 0);
__decorate([
    Struct.field("any"),
    __metadata("design:type", Array)
], Trx.prototype, "context_free_actions", void 0);
__decorate([
    Struct.field("any"),
    __metadata("design:type", Array)
], Trx.prototype, "context_free_data", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], Trx.prototype, "delay_sec", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], Trx.prototype, "expiration", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], Trx.prototype, "max_cpu_usage_ms", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], Trx.prototype, "max_net_usage_words", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], Trx.prototype, "ref_block_num", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], Trx.prototype, "ref_block_prefix", void 0);
__decorate([
    Struct.field("string", { array: true }),
    __metadata("design:type", Array)
], Trx.prototype, "signatures", void 0);
Trx = __decorate([
    Struct.type("trx")
], Trx);
export { Trx };
let TransactionInfo = class TransactionInfo extends Struct {
};
__decorate([
    Struct.field(TransactionReceipt),
    __metadata("design:type", TransactionReceipt)
], TransactionInfo.prototype, "receipt", void 0);
__decorate([
    Struct.field("trx"),
    __metadata("design:type", Trx)
], TransactionInfo.prototype, "trx", void 0);
TransactionInfo = __decorate([
    Struct.type("transaction_info")
], TransactionInfo);
export { TransactionInfo };
let GetTransactionResponse = class GetTransactionResponse extends Struct {
    traces;
};
__decorate([
    Struct.field(Checksum256),
    __metadata("design:type", Checksum256)
], GetTransactionResponse.prototype, "id", void 0);
__decorate([
    Struct.field(UInt32),
    __metadata("design:type", UInt32)
], GetTransactionResponse.prototype, "block_num", void 0);
__decorate([
    Struct.field(BlockTimestamp),
    __metadata("design:type", BlockTimestamp)
], GetTransactionResponse.prototype, "block_time", void 0);
__decorate([
    Struct.field(UInt32),
    __metadata("design:type", UInt32)
], GetTransactionResponse.prototype, "last_irreversible_block", void 0);
__decorate([
    Struct.field("any?"),
    __metadata("design:type", Array)
], GetTransactionResponse.prototype, "traces", void 0);
__decorate([
    Struct.field("any"),
    __metadata("design:type", TransactionInfo)
], GetTransactionResponse.prototype, "trx", void 0);
GetTransactionResponse = __decorate([
    Struct.type("get_transaction_response")
], GetTransactionResponse);
export { GetTransactionResponse };
let GetKeyAccountsResponse = class GetKeyAccountsResponse extends Struct {
};
__decorate([
    Struct.field("name", { array: true }),
    __metadata("design:type", Array)
], GetKeyAccountsResponse.prototype, "account_names", void 0);
GetKeyAccountsResponse = __decorate([
    Struct.type("get_key_accounts_response")
], GetKeyAccountsResponse);
export { GetKeyAccountsResponse };
let GetCodeResponse = class GetCodeResponse extends Struct {
};
__decorate([
    Struct.field(ABI),
    __metadata("design:type", Object)
], GetCodeResponse.prototype, "abi", void 0);
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], GetCodeResponse.prototype, "account_name", void 0);
__decorate([
    Struct.field("checksum256"),
    __metadata("design:type", Checksum256)
], GetCodeResponse.prototype, "code_hash", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], GetCodeResponse.prototype, "wast", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], GetCodeResponse.prototype, "wasm", void 0);
GetCodeResponse = __decorate([
    Struct.type("get_code_response")
], GetCodeResponse);
export { GetCodeResponse };
let GetControlledAccountsResponse = class GetControlledAccountsResponse extends Struct {
};
__decorate([
    Struct.field("name", { array: true }),
    __metadata("design:type", Array)
], GetControlledAccountsResponse.prototype, "controlled_accounts", void 0);
GetControlledAccountsResponse = __decorate([
    Struct.type("get_controlled_accounts_response")
], GetControlledAccountsResponse);
export { GetControlledAccountsResponse };
let GetCurrencyStatsItemResponse = class GetCurrencyStatsItemResponse extends Struct {
};
__decorate([
    Struct.field("asset"),
    __metadata("design:type", Asset)
], GetCurrencyStatsItemResponse.prototype, "supply", void 0);
__decorate([
    Struct.field("asset"),
    __metadata("design:type", Asset)
], GetCurrencyStatsItemResponse.prototype, "max_supply", void 0);
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], GetCurrencyStatsItemResponse.prototype, "issuer", void 0);
GetCurrencyStatsItemResponse = __decorate([
    Struct.type("get_currency_stats_item_response")
], GetCurrencyStatsItemResponse);
export { GetCurrencyStatsItemResponse };
let GetTransactionStatusResponse = class GetTransactionStatusResponse extends Struct {
};
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], GetTransactionStatusResponse.prototype, "state", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], GetTransactionStatusResponse.prototype, "head_number", void 0);
__decorate([
    Struct.field(BlockId),
    __metadata("design:type", BlockId)
], GetTransactionStatusResponse.prototype, "head_id", void 0);
__decorate([
    Struct.field("time_point"),
    __metadata("design:type", TimePoint)
], GetTransactionStatusResponse.prototype, "head_timestamp", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], GetTransactionStatusResponse.prototype, "irreversible_number", void 0);
__decorate([
    Struct.field(BlockId),
    __metadata("design:type", BlockId)
], GetTransactionStatusResponse.prototype, "irreversible_id", void 0);
__decorate([
    Struct.field("time_point"),
    __metadata("design:type", TimePoint)
], GetTransactionStatusResponse.prototype, "irreversible_timestamp", void 0);
__decorate([
    Struct.field(BlockId),
    __metadata("design:type", BlockId)
], GetTransactionStatusResponse.prototype, "earliest_tracked_block_id", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], GetTransactionStatusResponse.prototype, "earliest_tracked_block_number", void 0);
GetTransactionStatusResponse = __decorate([
    Struct.type("get_transaction_status_response")
], GetTransactionStatusResponse);
export { GetTransactionStatusResponse };
let ProducerAuthority = class ProducerAuthority extends Struct {
};
__decorate([
    Struct.field(UInt32),
    __metadata("design:type", UInt32)
], ProducerAuthority.prototype, "threshold", void 0);
__decorate([
    Struct.field(KeyWeight, { array: true }),
    __metadata("design:type", Array)
], ProducerAuthority.prototype, "keys", void 0);
ProducerAuthority = __decorate([
    Struct.type("producer_authority")
], ProducerAuthority);
export { ProducerAuthority };
let Producer = class Producer extends Struct {
    static from(data) {
        return super.from({
            ...data,
            authority: [data.authority[0], ProducerAuthority.from(data.authority[1])]
        });
    }
};
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], Producer.prototype, "producer_name", void 0);
__decorate([
    Struct.field("any", { array: true }),
    __metadata("design:type", Array)
], Producer.prototype, "authority", void 0);
Producer = __decorate([
    Struct.type("producer")
], Producer);
export { Producer };
let ProducerSchedule = class ProducerSchedule extends Struct {
};
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], ProducerSchedule.prototype, "version", void 0);
__decorate([
    Struct.field(Producer, { array: true }),
    __metadata("design:type", Array)
], ProducerSchedule.prototype, "producers", void 0);
ProducerSchedule = __decorate([
    Struct.type("producer_schedule")
], ProducerSchedule);
export { ProducerSchedule };
let GetProducerScheduleResponse = class GetProducerScheduleResponse extends Struct {
};
__decorate([
    Struct.field(ProducerSchedule, { optional: true }),
    __metadata("design:type", ProducerSchedule)
], GetProducerScheduleResponse.prototype, "active", void 0);
__decorate([
    Struct.field(ProducerSchedule, { optional: true }),
    __metadata("design:type", ProducerSchedule)
], GetProducerScheduleResponse.prototype, "pending", void 0);
__decorate([
    Struct.field(ProducerSchedule, { optional: true }),
    __metadata("design:type", ProducerSchedule)
], GetProducerScheduleResponse.prototype, "proposed", void 0);
GetProducerScheduleResponse = __decorate([
    Struct.type("get_producer_schedule_response")
], GetProducerScheduleResponse);
export { GetProducerScheduleResponse };
let ProtocolFeature = class ProtocolFeature extends Struct {
};
__decorate([
    Struct.field("checksum256"),
    __metadata("design:type", Checksum256)
], ProtocolFeature.prototype, "feature_digest", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], ProtocolFeature.prototype, "activation_ordinal", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], ProtocolFeature.prototype, "activation_block_num", void 0);
__decorate([
    Struct.field("checksum256"),
    __metadata("design:type", Checksum256)
], ProtocolFeature.prototype, "description_digest", void 0);
__decorate([
    Struct.field("string", { array: true }),
    __metadata("design:type", Array)
], ProtocolFeature.prototype, "dependencies", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], ProtocolFeature.prototype, "protocol_feature_type", void 0);
__decorate([
    Struct.field("any", { array: true }),
    __metadata("design:type", Array)
], ProtocolFeature.prototype, "specification", void 0);
ProtocolFeature = __decorate([
    Struct.type("protocol_feature")
], ProtocolFeature);
export { ProtocolFeature };
let GetProtocolFeaturesResponse = class GetProtocolFeaturesResponse extends Struct {
};
__decorate([
    Struct.field(ProtocolFeature, { array: true }),
    __metadata("design:type", Array)
], GetProtocolFeaturesResponse.prototype, "activated_protocol_features", void 0);
__decorate([
    Struct.field("uint32", { optional: true }),
    __metadata("design:type", UInt32)
], GetProtocolFeaturesResponse.prototype, "more", void 0);
GetProtocolFeaturesResponse = __decorate([
    Struct.type("get_protocol_features_response")
], GetProtocolFeaturesResponse);
export { GetProtocolFeaturesResponse };
