// @wireio/sdk-core@1.0.18/lib/resources/powerup/Abstract.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Asset, Float64, Int64, Struct, TimePointSec, UInt128, UInt32, UInt64, UInt8 } from "../../";
import BN from "bn.js";
import bigDecimal from "js-big-decimal";
export class PowerUpStateResource extends Struct {
    default_block_cpu_limit = UInt64.from(200000);
    default_block_net_limit = UInt64.from(1048576000);
    // Get the current number of allocated units (shift from REX -> PowerUp)
    get allocated() {
        return (1 - Number(this.weight_ratio) / Number(this.target_weight_ratio) / 100);
    }
    // Get the current percentage of reserved units
    get reserved() {
        return new BN(String(this.utilization)).div(new BN(String(this.weight)));
    }
    // Get the symbol definition for the token
    get symbol() {
        return this.min_price.symbol;
    }
    // Common casting for typed values to numbers
    cast() {
        return {
            adjusted_utilization: Number(this.adjusted_utilization),
            decay_secs: Number(this.decay_secs.value),
            exponent: Number(this.exponent),
            utilization: Number(this.utilization),
            utilization_timestamp: Number(this.utilization_timestamp.value),
            weight: new BN(String(this.weight)),
            weight_ratio: Number(this.weight_ratio)
        };
    }
    // Mimic: https://github.com/SYSIO/sysio.contracts/blob/d7bc0a5cc8c0c2edd4dc61b0126517d0cb46fd94/contracts/sysio.system/src/powerup.cpp#L358
    utilization_increase(sample, frac) {
        const { weight } = this;
        const frac128 = UInt128.from(frac);
        const resultBN = new BN(weight.value.mul(new BN(frac128.value))).div(new BN("1000000000000000"));
        const resultNumber = resultBN.toNumber();
        return Math.ceil(resultNumber);
    }
    // Mimic: https://github.com/SYSIO/sysio.contracts/blob/d7bc0a5cc8c0c2edd4dc61b0126517d0cb46fd94/contracts/sysio.system/src/powerup.cpp#L284-L298
    price_function(utilization) {
        const { exponent, weight } = this.cast();
        const max_price = this.max_price.value;
        const min_price = this.min_price.value;
        let price = min_price;
        const new_exponent = exponent - 1.0;
        if (new_exponent <= 0.0) {
            return max_price;
        }
        else {
            const util_weight = new BN(utilization).div(weight);
            price +=
                (max_price - min_price) * Math.pow(util_weight.toNumber(), new_exponent);
        }
        return price;
    }
    // Mimic: https://github.com/SYSIO/sysio.contracts/blob/d7bc0a5cc8c0c2edd4dc61b0126517d0cb46fd94/contracts/sysio.system/src/powerup.cpp#L274-L280
    price_integral_delta(start_utilization, end_utilization) {
        const { exponent, weight } = this.cast();
        const max_price = this.max_price.value;
        const min_price = this.min_price.value;
        const coefficient = (max_price - min_price) / exponent;
        const start_u = new BN(start_utilization).div(weight);
        const end_u = new BN(end_utilization).div(weight);
        const delta = min_price * end_u.toNumber() -
            min_price * start_u.toNumber() +
            coefficient * Math.pow(end_u.toNumber(), exponent) -
            coefficient * Math.pow(start_u.toNumber(), exponent);
        return delta;
    }
    // Mimic: https://github.com/SYSIO/sysio.contracts/blob/d7bc0a5cc8c0c2edd4dc61b0126517d0cb46fd94/contracts/sysio.system/src/powerup.cpp#L262-L315
    fee(utilization_increase, adjusted_utilization) {
        const { utilization, weight } = this.cast();
        let start_utilization = utilization;
        const end_utilization = start_utilization + utilization_increase;
        let fee = 0;
        if (start_utilization < adjusted_utilization) {
            const min = Math.min(utilization_increase, adjusted_utilization - start_utilization);
            fee += Number(new bigDecimal(this.price_function(adjusted_utilization) * min)
                .divide(new bigDecimal(weight.toString()))
                .getValue());
            start_utilization = adjusted_utilization;
        }
        if (start_utilization < end_utilization) {
            fee += this.price_integral_delta(start_utilization, end_utilization);
        }
        return fee;
    }
    // Mimic: https://github.com/SYSIO/sysio.contracts/blob/d7bc0a5cc8c0c2edd4dc61b0126517d0cb46fd94/contracts/sysio.system/src/powerup.cpp#L105-L117
    determine_adjusted_utilization(options) {
        // Casting SYSIO types to usable formats for JS calculations
        const { decay_secs, utilization, utilization_timestamp } = this.cast();
        let { adjusted_utilization } = this.cast();
        // If utilization is less than adjusted, calculate real time value
        if (utilization < adjusted_utilization) {
            // Create now & adjust JS timestamp to match SYSIO timestamp values
            const ts = options && options.timestamp ? options.timestamp : new Date();
            const now = TimePointSec.from(ts).toMilliseconds() / 1000;
            const diff = adjusted_utilization - utilization;
            let delta = diff * Math.exp(-(now - utilization_timestamp) / decay_secs);
            delta = Math.min(Math.max(delta, 0), diff); // Clamp the delta
            adjusted_utilization = utilization + delta;
        }
        return adjusted_utilization;
    }
}
__decorate([
    Struct.field("uint8"),
    __metadata("design:type", UInt8)
], PowerUpStateResource.prototype, "version", void 0);
__decorate([
    Struct.field("int64"),
    __metadata("design:type", Int64)
], PowerUpStateResource.prototype, "weight", void 0);
__decorate([
    Struct.field("int64"),
    __metadata("design:type", Int64)
], PowerUpStateResource.prototype, "weight_ratio", void 0);
__decorate([
    Struct.field("int64"),
    __metadata("design:type", Int64)
], PowerUpStateResource.prototype, "assumed_stake_weight", void 0);
__decorate([
    Struct.field("int64"),
    __metadata("design:type", Int64)
], PowerUpStateResource.prototype, "initial_weight_ratio", void 0);
__decorate([
    Struct.field("int64"),
    __metadata("design:type", Int64)
], PowerUpStateResource.prototype, "target_weight_ratio", void 0);
__decorate([
    Struct.field("time_point_sec"),
    __metadata("design:type", TimePointSec)
], PowerUpStateResource.prototype, "initial_timestamp", void 0);
__decorate([
    Struct.field("time_point_sec"),
    __metadata("design:type", TimePointSec)
], PowerUpStateResource.prototype, "target_timestamp", void 0);
__decorate([
    Struct.field("float64"),
    __metadata("design:type", Float64)
], PowerUpStateResource.prototype, "exponent", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], PowerUpStateResource.prototype, "decay_secs", void 0);
__decorate([
    Struct.field("asset"),
    __metadata("design:type", Asset)
], PowerUpStateResource.prototype, "min_price", void 0);
__decorate([
    Struct.field("asset"),
    __metadata("design:type", Asset)
], PowerUpStateResource.prototype, "max_price", void 0);
__decorate([
    Struct.field("int64"),
    __metadata("design:type", Int64)
], PowerUpStateResource.prototype, "utilization", void 0);
__decorate([
    Struct.field("int64"),
    __metadata("design:type", Int64)
], PowerUpStateResource.prototype, "adjusted_utilization", void 0);
__decorate([
    Struct.field("time_point_sec"),
    __metadata("design:type", TimePointSec)
], PowerUpStateResource.prototype, "utilization_timestamp", void 0);
