// @wireio/sdk-core@1.0.18/lib/resources/powerup/Cpu.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { Struct, UInt128 } from "../../";
import { BNPrecision } from "../IndexResources";
import { PowerUpStateResource } from "./Abstract";
import BN from "bn.js";
let PowerUpStateResourceCPU = class PowerUpStateResourceCPU extends PowerUpStateResource {
    // Return smallest units per day, μs (microseconds)
    per_day = (options) => this.us_per_day(options);
    // Return ms (milliseconds) per day
    ms_per_day(options) {
        return this.us_per_day(options) / 1000;
    }
    // Return μs (microseconds) per day
    us_per_day(options) {
        const limit = options && options.virtual_block_cpu_limit
            ? options.virtual_block_cpu_limit
            : this.default_block_cpu_limit;
        return Number(limit) * 2 * 60 * 60 * 24;
    }
    // Convert weight to μs (microseconds)
    weight_to_us(sample, weight) {
        return Math.ceil((weight * Number(sample)) / BNPrecision.toNumber());
    }
    // Convert μs (microseconds) to weight
    us_to_weight(sample, us) {
        return Math.floor((us / Number(sample)) * BNPrecision.toNumber());
    }
    // Default frac generation by smallest unit type
    frac = (usage, us) => this.frac_by_us(usage, us);
    // Frac generation by ms (milliseconds)
    frac_by_ms = (usage, ms) => this.frac_by_us(usage, ms * 1000);
    // Frac generation by μs (microseconds)
    frac_by_us(usage, us) {
        const { weight } = this.cast();
        const frac = new BN(this.us_to_weight(usage.cpu, us)).div(weight);
        return Math.floor(frac.toNumber() * Math.pow(10, 15));
    }
    // Price generation by smallest units, μs (microseconds)
    price_per = (usage, us = 1000, options) => this.price_per_us(usage, us, options);
    // Price generation by ms (milliseconds)
    price_per_ms = (usage, ms = 1, options) => this.price_per_us(usage, ms * 1000, options);
    // Price generation by μs (microseconds)
    price_per_us(usage, us = 1000, options) {
        // Determine the utilization increase by this action
        const frac = UInt128.from(this.frac(usage, us));
        const utilization_increase = this.utilization_increase(usage.cpu, frac);
        // Determine the adjusted utilization if needed
        const adjusted_utilization = this.determine_adjusted_utilization(options);
        // Derive the fee from the increase and utilization
        const fee = this.fee(utilization_increase, adjusted_utilization);
        // Force the fee up to the next highest value of precision
        const precision = Math.pow(10, this.max_price.symbol.precision);
        const value = Math.ceil(fee * precision) / precision;
        // Return the modified fee
        return value;
    }
};
PowerUpStateResourceCPU = __decorate([
    Struct.type("powerupstateresourcecpu")
], PowerUpStateResourceCPU);
export { PowerUpStateResourceCPU };
