// @wireio/sdk-core@1.0.18/lib/resources/powerup/Net.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { Struct, UInt128 } from "../../";
import { BNPrecision } from "../IndexResources";
import { PowerUpStateResource } from "./Abstract";
import BN from "bn.js";
let PowerUpStateResourceNET = class PowerUpStateResourceNET extends PowerUpStateResource {
    // Return smallest units per day, bytes
    per_day = (options) => this.bytes_per_day(options);
    // Return kb per day
    kb_per_day(options) {
        return this.bytes_per_day(options) / 1000;
    }
    // Return bytes per day
    bytes_per_day(options) {
        const limit = options && options.virtual_block_net_limit
            ? options.virtual_block_net_limit
            : this.default_block_net_limit;
        return Number(limit) * 2 * 60 * 60 * 24;
    }
    // Convert weight to bytes
    weight_to_bytes(sample, weight) {
        return Math.ceil((weight * Number(sample)) / BNPrecision.toNumber());
    }
    // Convert bytes to weight
    bytes_to_weight(sample, bytes) {
        return Math.floor((bytes / Number(sample)) * BNPrecision.toNumber());
    }
    // Default frac generation by smallest unit type
    frac = (usage, bytes) => this.frac_by_bytes(usage, bytes);
    // Frac generation by kb
    frac_by_kb = (usage, kilobytes) => this.frac_by_bytes(usage, kilobytes * 1000);
    // Frac generation by bytes
    frac_by_bytes(usage, bytes) {
        const { weight } = this.cast();
        const frac = new BN(this.bytes_to_weight(usage.net, bytes)).div(weight);
        return Math.floor(frac.toNumber() * Math.pow(10, 15));
    }
    // Price generation by smallest units, bytes
    price_per = (usage, bytes = 1000, options) => this.price_per_byte(usage, bytes, options);
    // Price generation by kb
    price_per_kb = (usage, kilobytes = 1, options) => this.price_per_byte(usage, kilobytes * 1000, options);
    // Price generation by bytes
    price_per_byte(usage, bytes = 1000, options) {
        // Determine the utilization increase by this action
        const frac = UInt128.from(this.frac(usage, bytes));
        const utilization_increase = this.utilization_increase(usage.net, frac);
        // Determine the adjusted utilization if needed
        const adjusted_utilization = this.determine_adjusted_utilization(options);
        // Derive the fee from the increase and utilization
        const fee = this.fee(utilization_increase, adjusted_utilization);
        // Force the fee up to the next highest value of precision
        const precision = Math.pow(10, this.max_price.symbol.precision);
        const value = Math.ceil(fee * precision) / precision;
        // Return the modified fee
        return value;
    }
};
PowerUpStateResourceNET = __decorate([
    Struct.type("powerupstateresourcenet")
], PowerUpStateResourceNET);
export { PowerUpStateResourceNET };
