// @wireio/sdk-core@1.0.18/lib/signing/Abi.js
/** SigningRequest ABI and typedefs. */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var RequestFlags_1;
import { Action } from "../chain/Action";
import { Bytes } from "../chain/Bytes";
import { Name } from "../chain/Name";
import { PermissionLevel } from "../chain/PermissionLevel";
import { Signature } from "../chain/Signature";
import { Struct } from "../chain/Struct";
import { Transaction } from "../chain/Transaction";
import { TypeAlias } from "../chain/TypeAlias";
import { UInt8 } from "../chain/Integer";
import { Variant } from "../chain/Variant";
import { ChainIdVariant } from "./ChainId";
let AccountName = class AccountName extends Name {
};
AccountName = __decorate([
    TypeAlias("account_name")
], AccountName);
export { AccountName };
let PermissionName = class PermissionName extends Name {
};
PermissionName = __decorate([
    TypeAlias("permission_name")
], PermissionName);
export { PermissionName };
let IdentityV2 = class IdentityV2 extends Struct {
    permission;
};
__decorate([
    Struct.field(PermissionLevel, { optional: true }),
    __metadata("design:type", PermissionLevel)
], IdentityV2.prototype, "permission", void 0);
IdentityV2 = __decorate([
    Struct.type("identity")
], IdentityV2);
export { IdentityV2 };
let IdentityV3 = class IdentityV3 extends Struct {
    permission;
};
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], IdentityV3.prototype, "scope", void 0);
__decorate([
    Struct.field(PermissionLevel, { optional: true }),
    __metadata("design:type", PermissionLevel)
], IdentityV3.prototype, "permission", void 0);
IdentityV3 = __decorate([
    Struct.type("identity")
], IdentityV3);
export { IdentityV3 };
let RequestVariantV2 = class RequestVariantV2 extends Variant {
};
RequestVariantV2 = __decorate([
    Variant.type("variant_req", [
        Action,
        { type: Action, array: true },
        Transaction,
        IdentityV2
    ])
], RequestVariantV2);
export { RequestVariantV2 };
let RequestVariantV3 = class RequestVariantV3 extends Variant {
};
RequestVariantV3 = __decorate([
    Variant.type("variant_req", [
        Action,
        { type: Action, array: true },
        Transaction,
        IdentityV3
    ])
], RequestVariantV3);
export { RequestVariantV3 };
let RequestFlags = class RequestFlags extends UInt8 {
    static { RequestFlags_1 = this; }
    static broadcast = 1 << 0;
    static background = 1 << 1;
    get broadcast() {
        return (Number(this) & RequestFlags_1.broadcast) !== 0;
    }
    set broadcast(enabled) {
        this.setFlag(RequestFlags_1.broadcast, enabled);
    }
    get background() {
        return (Number(this) & RequestFlags_1.background) !== 0;
    }
    set background(enabled) {
        this.setFlag(RequestFlags_1.background, enabled);
    }
    setFlag(flag, enabled) {
        if (enabled) {
            // TODO: implement bitwise operators in core, bn.js setbit does not work
            this.value = UInt8.from(Number(this) | flag).value;
        }
        else {
            this.value = UInt8.from(Number(this) & ~flag).value;
        }
    }
};
RequestFlags = RequestFlags_1 = __decorate([
    TypeAlias("request_flags")
], RequestFlags);
export { RequestFlags };
let InfoPair = class InfoPair extends Struct {
};
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], InfoPair.prototype, "key", void 0);
__decorate([
    Struct.field("bytes"),
    __metadata("design:type", Bytes)
], InfoPair.prototype, "value", void 0);
InfoPair = __decorate([
    Struct.type("info_pair")
], InfoPair);
export { InfoPair };
let RequestDataV2 = class RequestDataV2 extends Struct {
};
__decorate([
    Struct.field(ChainIdVariant),
    __metadata("design:type", ChainIdVariant)
], RequestDataV2.prototype, "chain_id", void 0);
__decorate([
    Struct.field(RequestVariantV2),
    __metadata("design:type", RequestVariantV2)
], RequestDataV2.prototype, "req", void 0);
__decorate([
    Struct.field(RequestFlags),
    __metadata("design:type", RequestFlags)
], RequestDataV2.prototype, "flags", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], RequestDataV2.prototype, "callback", void 0);
__decorate([
    Struct.field(InfoPair, { array: true }),
    __metadata("design:type", Array)
], RequestDataV2.prototype, "info", void 0);
RequestDataV2 = __decorate([
    Struct.type("signing_request")
], RequestDataV2);
export { RequestDataV2 };
let RequestDataV3 = class RequestDataV3 extends Struct {
};
__decorate([
    Struct.field(ChainIdVariant),
    __metadata("design:type", ChainIdVariant)
], RequestDataV3.prototype, "chain_id", void 0);
__decorate([
    Struct.field(RequestVariantV3),
    __metadata("design:type", RequestVariantV3)
], RequestDataV3.prototype, "req", void 0);
__decorate([
    Struct.field(RequestFlags),
    __metadata("design:type", RequestFlags)
], RequestDataV3.prototype, "flags", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], RequestDataV3.prototype, "callback", void 0);
__decorate([
    Struct.field(InfoPair, { array: true }),
    __metadata("design:type", Array)
], RequestDataV3.prototype, "info", void 0);
RequestDataV3 = __decorate([
    Struct.type("signing_request")
], RequestDataV3);
export { RequestDataV3 };
let RequestSignature = class RequestSignature extends Struct {
};
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], RequestSignature.prototype, "signer", void 0);
__decorate([
    Struct.field("signature"),
    __metadata("design:type", Signature)
], RequestSignature.prototype, "signature", void 0);
RequestSignature = __decorate([
    Struct.type("request_signature")
], RequestSignature);
export { RequestSignature };
