// @wireio/sdk-core@1.0.18/lib/signing/ChainId.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { Checksum256 } from "../chain/Checksum";
import { isInstanceOf } from "../Utils";
import { TypeAlias } from "../chain/TypeAlias";
import { UInt8 } from "../chain/Integer";
import { Variant } from "../chain/Variant";
/** Chain ID aliases. */
export var ChainName;
(function (ChainName) {
    ChainName[ChainName["UNKNOWN"] = 0] = "UNKNOWN";
    ChainName[ChainName["SYS"] = 1] = "SYS";
    ChainName[ChainName["TELOS"] = 2] = "TELOS";
    ChainName[ChainName["JUNGLE"] = 3] = "JUNGLE";
    ChainName[ChainName["KYLIN"] = 4] = "KYLIN";
    ChainName[ChainName["WORBLI"] = 5] = "WORBLI";
    ChainName[ChainName["BOS"] = 6] = "BOS";
    ChainName[ChainName["MEETONE"] = 7] = "MEETONE";
    ChainName[ChainName["INSIGHTS"] = 8] = "INSIGHTS";
    ChainName[ChainName["BEOS"] = 9] = "BEOS";
    ChainName[ChainName["WAX"] = 10] = "WAX";
    ChainName[ChainName["PROTON"] = 11] = "PROTON";
    ChainName[ChainName["FIO"] = 12] = "FIO";
})(ChainName || (ChainName = {}));
let ChainId = class ChainId extends Checksum256 {
    static from(value) {
        if (isInstanceOf(value, this)) {
            return value;
        }
        if (typeof value === "number") {
            value = ChainIdLookup.get(value);
            if (!value) {
                throw new Error("Unknown chain id alias");
            }
        }
        return super.from(value);
    }
    get chainVariant() {
        const name = this.chainName;
        if (name !== ChainName.UNKNOWN) {
            return ChainIdVariant.from(["chain_alias", name]);
        }
        return ChainIdVariant.from(this);
    }
    get chainName() {
        const cid = this.hexString;
        for (const [n, id] of ChainIdLookup) {
            if (id === cid) {
                return n;
            }
        }
        return ChainName.UNKNOWN;
    }
};
ChainId = __decorate([
    TypeAlias("chain_id")
], ChainId);
export { ChainId };
let ChainAlias = class ChainAlias extends UInt8 {
};
ChainAlias = __decorate([
    TypeAlias("chain_alias")
], ChainAlias);
export { ChainAlias };
let ChainIdVariant = class ChainIdVariant extends Variant {
    get chainId() {
        if (isInstanceOf(this.value, ChainId)) {
            return this.value;
        }
        return ChainId.from(Number(this.value.value));
    }
};
ChainIdVariant = __decorate([
    Variant.type("variant_id", [ChainAlias, ChainId])
], ChainIdVariant);
export { ChainIdVariant };
const ChainIdLookup = new Map([
    [
        ChainName.SYS,
        "aca376f206b8fc25a6ed44dbdc66547c36c6c33e3a119ffbeaef943642f0e906"
    ],
    [
        ChainName.TELOS,
        "4667b205c6838ef70ff7988f6e8257e8be0e1284a2f59699054a018f743b1d11"
    ],
    [
        ChainName.JUNGLE,
        "e70aaab8997e1dfce58fbfac80cbbb8fecec7b99cf982a9444273cbc64c41473"
    ],
    [
        ChainName.KYLIN,
        "5fff1dae8dc8e2fc4d5b23b2c7665c97f9e9d8edf2b6485a86ba311c25639191"
    ],
    [
        ChainName.WORBLI,
        "73647cde120091e0a4b85bced2f3cfdb3041e266cbbe95cee59b73235a1b3b6f"
    ],
    [
        ChainName.BOS,
        "d5a3d18fbb3c084e3b1f3fa98c21014b5f3db536cc15d08f9f6479517c6a3d86"
    ],
    [
        ChainName.MEETONE,
        "cfe6486a83bad4962f232d48003b1824ab5665c36778141034d75e57b956e422"
    ],
    [
        ChainName.INSIGHTS,
        "b042025541e25a472bffde2d62edd457b7e70cee943412b1ea0f044f88591664"
    ],
    [
        ChainName.BEOS,
        "b912d19a6abd2b1b05611ae5be473355d64d95aeff0c09bedc8c166cd6468fe4"
    ],
    [
        ChainName.WAX,
        "1064487b3cd1a897ce03ae5b6a865651747e2e152090f99c1d19d44e01aea5a4"
    ],
    [
        ChainName.PROTON,
        "384da888112027f0321850a169f737c33e53b388aad48b5adace4bab97f437e0"
    ],
    [
        ChainName.FIO,
        "21dcae42c0182200e93f954a074011f9048a7624c6fe81d3c9541a614a88bd1c"
    ]
]);
