// @wireio/sdk-core@1.0.18/lib/common/Chains.js
import { Checksum256 } from "../chain/Checksum";
import { Logo } from "./Logo";
/**
 * Holds all information for a single chain.
 */
export class ChainDefinition {
    id;
    name;
    endpoint;
    hyperion;
    websocket;
    watchdawg;
    namespace;
    coreSymbol;
    selected;
    logo;
    constructor(data) {
        this.id = Checksum256.from(data.id);
        this.name = data.name;
        this.endpoint = data.endpoint;
        this.hyperion = data.hyperion;
        this.websocket = data.websocket;
        this.watchdawg = data.watchdawg;
        this.namespace = data.namespace;
        this.coreSymbol = data.coreSymbol;
        this.selected = !!data.selected;
        this.logo = data.logo;
    }
    static from(data) {
        const inst = new ChainDefinition(data);
        if (data.logo) {
            inst.logo = Logo.from(data.logo);
        }
        return inst;
    }
    /**
     * If you passed a `logo`, returns a Logo instance.
     */
    getLogo() {
        return this.logo ? Logo.from(this.logo) : undefined;
    }
    /**
     * Two chains are equal if ID+endpoint match.
     */
    equals(def) {
        const other = ChainDefinition.from(def);
        return this.id.equals(other.id) && this.endpoint === other.endpoint;
    }
}
// ----------------------------------------------------------------------------
// built-in Wire networks
// ----------------------------------------------------------------------------
export var Chains;
(function (Chains) {
    Chains.TESTNET = ChainDefinition.from({
        id: "065dcca2dc758af25bcf3b878260a19dd1b81e4597f2af15a262a0c67f1e0106",
        name: "Wire Testnet",
        endpoint: "https://testnet-00.wire.foundation",
        hyperion: "https://testnet-hyperion.wire.foundation",
        websocket: "ws://testnet-ship.wire.foundation",
        watchdawg: "https://dawg.wire.foundation",
        namespace: "sysio",
        coreSymbol: "SYS",
        selected: true,
        logo: "../assets/logos/wire-testnet.png"
    });
    Chains.DEVNET = ChainDefinition.from({
        id: "260348861b12af1497fd274ca772d1f79978f7acc605c8e766344acd967536dc",
        name: "Wire Devnet",
        endpoint: "https://det-dev.gitgo.app",
        hyperion: "https://dev-hyperion.gitgo.app",
        websocket: "ws://dev-hist.gitgo.app",
        namespace: "sysio",
        coreSymbol: "SYS",
        logo: "../assets/logos/wire-devnet.png"
    });
    Chains.MAINNET_CLASSIC = ChainDefinition.from({
        id: "de9943091e811bfb246ca243144b4d274886b959bbb17dd33d0bc97c745dbbe0",
        name: "Wire Classic",
        endpoint: "https://wire.siliconswamp.info",
        hyperion: "https://hyperwire.airwire.io",
        websocket: "ws://swamprod.airwire.io:8080",
        namespace: "sysio",
        coreSymbol: "SYS",
        logo: "../assets/logos/wire.png"
    });
})(Chains || (Chains = {}));
