// @wireio/sdk-core@1.0.18/lib/common/Connect.js
export var Curve;
(function (Curve) {
    Curve["EC"] = "EC";
    Curve["ED"] = "ED";
})(Curve || (Curve = {}));
export var SolChainID;
(function (SolChainID) {
    SolChainID["Devnet"] = "devnet";
    SolChainID["Testnet"] = "testnet";
    SolChainID["Mainnet"] = "mainnet-beta";
    SolChainID["WireTestnet"] = "wire-testnet";
})(SolChainID || (SolChainID = {}));
export var EvmChainID;
(function (EvmChainID) {
    EvmChainID[EvmChainID["Ethereum"] = 1] = "Ethereum";
    EvmChainID[EvmChainID["Hoodi"] = 560048] = "Hoodi";
    EvmChainID[EvmChainID["Sepolia"] = 11155111] = "Sepolia";
    EvmChainID[EvmChainID["Polygon"] = 137] = "Polygon";
    EvmChainID[EvmChainID["WireTestnet"] = 1122334455] = "WireTestnet";
})(EvmChainID || (EvmChainID = {}));
export var ProviderType;
(function (ProviderType) {
    ProviderType["Metamask"] = "metamask";
    ProviderType["RabbyWallet"] = "rabbywallet";
    ProviderType["BraveWalletEth"] = "bravewalleteth";
    ProviderType["WalletConnect"] = "walletconnect";
    ProviderType["CoinbaseWallet"] = "coinbasewallet";
    ProviderType["BraveWalletSol"] = "bravewalletsol";
    ProviderType["Phantom"] = "phantom";
    ProviderType["Solflare"] = "solflare";
    ProviderType["Backpack"] = "backpack";
    ProviderType["SquadsX"] = "squadsx";
})(ProviderType || (ProviderType = {}));
