// @wireio/opp-solidity-models@1.0.14/sysio/opp/opp.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { ChainSignature } from "./types/types";
import { AttestationType } from "./types/types";
import { EncodingFlags } from "./types/types";
import { ChainId } from "./types/types";
// @generated message type with reflection information, may provide speed optimized methods
class Endpoints$Type extends MessageType {
    constructor() {
        super("sysio.opp.Endpoints", [
            { no: 1, name: "start", kind: "message", T: () => ChainId },
            { no: 2, name: "end", kind: "message", T: () => ChainId }
        ]);
    }
    create(value) {
        const message = globalThis.Object.create((this.messagePrototype));
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* sysio.opp.types.ChainId start */ 1:
                    message.start = ChainId.internalBinaryRead(reader, reader.uint32(), options, message.start);
                    break;
                case /* sysio.opp.types.ChainId end */ 2:
                    message.end = ChainId.internalBinaryRead(reader, reader.uint32(), options, message.end);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* sysio.opp.types.ChainId start = 1; */
        if (message.start)
            ChainId.internalBinaryWrite(message.start, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
        /* sysio.opp.types.ChainId end = 2; */
        if (message.end)
            ChainId.internalBinaryWrite(message.end, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message sysio.opp.Endpoints
 */
export const Endpoints = new Endpoints$Type();
// @generated message type with reflection information, may provide speed optimized methods
class MessageHeader$Type extends MessageType {
    constructor() {
        super("sysio.opp.MessageHeader", [
            { no: 1, name: "endpoints", kind: "message", T: () => Endpoints },
            { no: 2, name: "message_id", kind: "scalar", T: 12 /*ScalarType.BYTES*/ },
            { no: 3, name: "previous_message_id", kind: "scalar", T: 12 /*ScalarType.BYTES*/ },
            { no: 4, name: "encoding_flags", kind: "message", T: () => EncodingFlags },
            { no: 5, name: "payload_size", kind: "scalar", T: 13 /*ScalarType.UINT32*/ },
            { no: 6, name: "payload_checksum", kind: "scalar", T: 12 /*ScalarType.BYTES*/ },
            { no: 7, name: "timestamp", kind: "scalar", T: 4 /*ScalarType.UINT64*/, L: 0 /*LongType.BIGINT*/ },
            { no: 8, name: "header_checksum", kind: "scalar", T: 12 /*ScalarType.BYTES*/ }
        ]);
    }
    create(value) {
        const message = globalThis.Object.create((this.messagePrototype));
        message.messageId = new Uint8Array(0);
        message.previousMessageId = new Uint8Array(0);
        message.payloadSize = 0;
        message.payloadChecksum = new Uint8Array(0);
        message.timestamp = 0n;
        message.headerChecksum = new Uint8Array(0);
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* sysio.opp.Endpoints endpoints */ 1:
                    message.endpoints = Endpoints.internalBinaryRead(reader, reader.uint32(), options, message.endpoints);
                    break;
                case /* bytes message_id */ 2:
                    message.messageId = reader.bytes();
                    break;
                case /* bytes previous_message_id */ 3:
                    message.previousMessageId = reader.bytes();
                    break;
                case /* sysio.opp.types.EncodingFlags encoding_flags */ 4:
                    message.encodingFlags = EncodingFlags.internalBinaryRead(reader, reader.uint32(), options, message.encodingFlags);
                    break;
                case /* uint32 payload_size */ 5:
                    message.payloadSize = reader.uint32();
                    break;
                case /* bytes payload_checksum */ 6:
                    message.payloadChecksum = reader.bytes();
                    break;
                case /* uint64 timestamp */ 7:
                    message.timestamp = reader.uint64().toBigInt();
                    break;
                case /* bytes header_checksum */ 8:
                    message.headerChecksum = reader.bytes();
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* sysio.opp.Endpoints endpoints = 1; */
        if (message.endpoints)
            Endpoints.internalBinaryWrite(message.endpoints, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
        /* bytes message_id = 2; */
        if (message.messageId.length)
            writer.tag(2, WireType.LengthDelimited).bytes(message.messageId);
        /* bytes previous_message_id = 3; */
        if (message.previousMessageId.length)
            writer.tag(3, WireType.LengthDelimited).bytes(message.previousMessageId);
        /* sysio.opp.types.EncodingFlags encoding_flags = 4; */
        if (message.encodingFlags)
            EncodingFlags.internalBinaryWrite(message.encodingFlags, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
        /* uint32 payload_size = 5; */
        if (message.payloadSize !== 0)
            writer.tag(5, WireType.Varint).uint32(message.payloadSize);
        /* bytes payload_checksum = 6; */
        if (message.payloadChecksum.length)
            writer.tag(6, WireType.LengthDelimited).bytes(message.payloadChecksum);
        /* uint64 timestamp = 7; */
        if (message.timestamp !== 0n)
            writer.tag(7, WireType.Varint).uint64(message.timestamp);
        /* bytes header_checksum = 8; */
        if (message.headerChecksum.length)
            writer.tag(8, WireType.LengthDelimited).bytes(message.headerChecksum);
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message sysio.opp.MessageHeader
 */
export const MessageHeader = new MessageHeader$Type();
// @generated message type with reflection information, may provide speed optimized methods
class MessagePayload$Type extends MessageType {
    constructor() {
        super("sysio.opp.MessagePayload", [
            { no: 1, name: "version", kind: "scalar", T: 13 /*ScalarType.UINT32*/ },
            { no: 2, name: "attestations", kind: "message", repeat: 2 /*RepeatType.UNPACKED*/, T: () => AttestationEntry }
        ]);
    }
    create(value) {
        const message = globalThis.Object.create((this.messagePrototype));
        message.version = 0;
        message.attestations = [];
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* uint32 version */ 1:
                    message.version = reader.uint32();
                    break;
                case /* repeated sysio.opp.AttestationEntry attestations */ 2:
                    message.attestations.push(AttestationEntry.internalBinaryRead(reader, reader.uint32(), options));
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* uint32 version = 1; */
        if (message.version !== 0)
            writer.tag(1, WireType.Varint).uint32(message.version);
        /* repeated sysio.opp.AttestationEntry attestations = 2; */
        for (let i = 0; i < message.attestations.length; i++)
            AttestationEntry.internalBinaryWrite(message.attestations[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message sysio.opp.MessagePayload
 */
export const MessagePayload = new MessagePayload$Type();
// @generated message type with reflection information, may provide speed optimized methods
class AttestationEntry$Type extends MessageType {
    constructor() {
        super("sysio.opp.AttestationEntry", [
            { no: 1, name: "type", kind: "enum", T: () => ["sysio.opp.types.AttestationType", AttestationType, "ATTESTATION_TYPE_"] },
            { no: 2, name: "data_size", kind: "scalar", T: 13 /*ScalarType.UINT32*/ },
            { no: 3, name: "data", kind: "scalar", T: 12 /*ScalarType.BYTES*/ }
        ]);
    }
    create(value) {
        const message = globalThis.Object.create((this.messagePrototype));
        message.type = 0;
        message.dataSize = 0;
        message.data = new Uint8Array(0);
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* sysio.opp.types.AttestationType type */ 1:
                    message.type = reader.int32();
                    break;
                case /* uint32 data_size */ 2:
                    message.dataSize = reader.uint32();
                    break;
                case /* bytes data */ 3:
                    message.data = reader.bytes();
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* sysio.opp.types.AttestationType type = 1; */
        if (message.type !== 0)
            writer.tag(1, WireType.Varint).int32(message.type);
        /* uint32 data_size = 2; */
        if (message.dataSize !== 0)
            writer.tag(2, WireType.Varint).uint32(message.dataSize);
        /* bytes data = 3; */
        if (message.data.length)
            writer.tag(3, WireType.LengthDelimited).bytes(message.data);
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message sysio.opp.AttestationEntry
 */
export const AttestationEntry = new AttestationEntry$Type();
// @generated message type with reflection information, may provide speed optimized methods
class Message$Type extends MessageType {
    constructor() {
        super("sysio.opp.Message", [
            { no: 1, name: "header", kind: "message", T: () => MessageHeader },
            { no: 2, name: "payload", kind: "message", T: () => MessagePayload }
        ]);
    }
    create(value) {
        const message = globalThis.Object.create((this.messagePrototype));
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* sysio.opp.MessageHeader header */ 1:
                    message.header = MessageHeader.internalBinaryRead(reader, reader.uint32(), options, message.header);
                    break;
                case /* sysio.opp.MessagePayload payload */ 2:
                    message.payload = MessagePayload.internalBinaryRead(reader, reader.uint32(), options, message.payload);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* sysio.opp.MessageHeader header = 1; */
        if (message.header)
            MessageHeader.internalBinaryWrite(message.header, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
        /* sysio.opp.MessagePayload payload = 2; */
        if (message.payload)
            MessagePayload.internalBinaryWrite(message.payload, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message sysio.opp.Message
 */
export const Message = new Message$Type();
// @generated message type with reflection information, may provide speed optimized methods
class Envelope$Type extends MessageType {
    constructor() {
        super("sysio.opp.Envelope", [
            { no: 1, name: "envelope_hash", kind: "scalar", T: 12 /*ScalarType.BYTES*/ },
            { no: 2, name: "endpoints", kind: "message", T: () => Endpoints },
            { no: 5, name: "epoch_timestamp", kind: "scalar", T: 4 /*ScalarType.UINT64*/, L: 0 /*LongType.BIGINT*/ },
            { no: 6, name: "epoch_index", kind: "scalar", T: 13 /*ScalarType.UINT32*/ },
            { no: 7, name: "epoch_envelope_index", kind: "scalar", T: 13 /*ScalarType.UINT32*/ },
            { no: 15, name: "merkle", kind: "scalar", T: 12 /*ScalarType.BYTES*/ },
            { no: 20, name: "previous_envelope_hash", kind: "scalar", T: 12 /*ScalarType.BYTES*/ },
            { no: 30, name: "start_message_id", kind: "scalar", T: 12 /*ScalarType.BYTES*/ },
            { no: 31, name: "end_message_id", kind: "scalar", T: 12 /*ScalarType.BYTES*/ },
            { no: 50, name: "signatures", kind: "message", repeat: 2 /*RepeatType.UNPACKED*/, T: () => ChainSignature }
        ]);
    }
    create(value) {
        const message = globalThis.Object.create((this.messagePrototype));
        message.envelopeHash = new Uint8Array(0);
        message.epochTimestamp = 0n;
        message.epochIndex = 0;
        message.epochEnvelopeIndex = 0;
        message.merkle = new Uint8Array(0);
        message.previousEnvelopeHash = new Uint8Array(0);
        message.startMessageId = new Uint8Array(0);
        message.endMessageId = new Uint8Array(0);
        message.signatures = [];
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* bytes envelope_hash */ 1:
                    message.envelopeHash = reader.bytes();
                    break;
                case /* sysio.opp.Endpoints endpoints */ 2:
                    message.endpoints = Endpoints.internalBinaryRead(reader, reader.uint32(), options, message.endpoints);
                    break;
                case /* uint64 epoch_timestamp */ 5:
                    message.epochTimestamp = reader.uint64().toBigInt();
                    break;
                case /* uint32 epoch_index */ 6:
                    message.epochIndex = reader.uint32();
                    break;
                case /* uint32 epoch_envelope_index */ 7:
                    message.epochEnvelopeIndex = reader.uint32();
                    break;
                case /* bytes merkle */ 15:
                    message.merkle = reader.bytes();
                    break;
                case /* bytes previous_envelope_hash */ 20:
                    message.previousEnvelopeHash = reader.bytes();
                    break;
                case /* bytes start_message_id */ 30:
                    message.startMessageId = reader.bytes();
                    break;
                case /* bytes end_message_id */ 31:
                    message.endMessageId = reader.bytes();
                    break;
                case /* repeated sysio.opp.types.ChainSignature signatures */ 50:
                    message.signatures.push(ChainSignature.internalBinaryRead(reader, reader.uint32(), options));
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* bytes envelope_hash = 1; */
        if (message.envelopeHash.length)
            writer.tag(1, WireType.LengthDelimited).bytes(message.envelopeHash);
        /* sysio.opp.Endpoints endpoints = 2; */
        if (message.endpoints)
            Endpoints.internalBinaryWrite(message.endpoints, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
        /* uint64 epoch_timestamp = 5; */
        if (message.epochTimestamp !== 0n)
            writer.tag(5, WireType.Varint).uint64(message.epochTimestamp);
        /* uint32 epoch_index = 6; */
        if (message.epochIndex !== 0)
            writer.tag(6, WireType.Varint).uint32(message.epochIndex);
        /* uint32 epoch_envelope_index = 7; */
        if (message.epochEnvelopeIndex !== 0)
            writer.tag(7, WireType.Varint).uint32(message.epochEnvelopeIndex);
        /* bytes merkle = 15; */
        if (message.merkle.length)
            writer.tag(15, WireType.LengthDelimited).bytes(message.merkle);
        /* bytes previous_envelope_hash = 20; */
        if (message.previousEnvelopeHash.length)
            writer.tag(20, WireType.LengthDelimited).bytes(message.previousEnvelopeHash);
        /* bytes start_message_id = 30; */
        if (message.startMessageId.length)
            writer.tag(30, WireType.LengthDelimited).bytes(message.startMessageId);
        /* bytes end_message_id = 31; */
        if (message.endMessageId.length)
            writer.tag(31, WireType.LengthDelimited).bytes(message.endMessageId);
        /* repeated sysio.opp.types.ChainSignature signatures = 50; */
        for (let i = 0; i < message.signatures.length; i++)
            ChainSignature.internalBinaryWrite(message.signatures[i], writer.tag(50, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message sysio.opp.Envelope
 */
export const Envelope = new Envelope$Type();
