// @wireio/sdk-core@1.0.18/lib/serializer/Builtins.js
import { Asset, ExtendedAsset } from "../chain/Asset";
import { BlockTimestamp, TimePoint, TimePointSec } from "../chain/Time";
import { Bytes } from "../chain/Bytes";
import { Checksum160, Checksum256, Checksum512 } from "../chain/Checksum";
import { Float128, Float32, Float64 } from "../chain/Float";
import { Int128, Int16, Int32, Int64, Int8, UInt128, UInt16, UInt32, UInt64, UInt8, VarInt, VarUInt } from "../chain/Integer";
import { Name } from "../chain/Name";
import { PublicKey } from "../chain/PublicKey";
import { Signature } from "../chain/Signature";
import { Struct } from "../chain/Struct";
const StringType = {
    abiName: "string",
    abiDefault: () => "",
    fromABI: (decoder) => {
        return decoder.readString();
    },
    from: (string) => string,
    toABI: (string, encoder) => {
        encoder.writeString(string);
    }
};
const BoolType = {
    abiName: "bool",
    abiDefault: () => false,
    fromABI: (decoder) => {
        return decoder.readByte() === 1;
    },
    from: (value) => value,
    toABI: (value, encoder) => {
        encoder.writeByte(value === true ? 1 : 0);
    }
};
function getBuiltins() {
    return [
        // types represented by JavaScript builtins
        BoolType,
        StringType,
        // types represented by Classes
        Asset,
        Asset.Symbol,
        Asset.SymbolCode,
        BlockTimestamp,
        Bytes,
        Checksum160,
        Checksum256,
        Checksum512,
        ExtendedAsset,
        Float128,
        Float32,
        Float64,
        Int128,
        Int16,
        Int32,
        Int64,
        Int8,
        Name,
        PublicKey,
        Signature,
        TimePoint,
        TimePointSec,
        UInt128,
        UInt16,
        UInt32,
        UInt64,
        UInt8,
        VarInt,
        VarUInt
    ];
}
export function buildTypeLookup(additional = []) {
    const rv = {};
    const builtins = getBuiltins();
    for (const type of builtins) {
        rv[type.abiName] = type;
    }
    for (const type of additional) {
        if (!type.abiName) {
            throw new Error("Invalid type");
        }
        rv[type.abiName] = type;
    }
    return rv;
}
export function getTypeName(object) {
    if (object.constructor && object.constructor.abiName !== undefined) {
        return object.constructor.abiName;
    }
    if (Array.isArray(object)) {
        const types = object.map(getTypeName);
        const type = types[0];
        if (!type || !types.every(t => t === type)) {
            return;
        }
        return type + "[]";
    }
    switch (typeof object) {
        case "boolean":
            return "bool";
        case "string":
            return "string";
    }
}
export function getType(object, name = "jsobj") {
    if (object.constructor && object.constructor.abiName !== undefined) {
        return object.constructor;
    }
    if (Array.isArray(object)) {
        // check for array of all ABISerializableType with same type name
        const types = object.map(v => {
            return getType(v, name);
        });
        const type = types[0];
        if (!type) {
            return; // some type not known
        }
        if (!types.every(t => t && t.abiName === type.abiName)) {
            return; // not all types are the same
        }
        return type;
    }
    const objectType = typeof object;
    if (objectType === "object" && object !== null) {
        const fields = Object.keys(object).map(key => {
            return { name: key, type: getType(object[key], name + "_nested") };
        });
        if (fields.find(field => !field.type)) {
            return; // encountered unknown type
        }
        return class extends Struct {
            static abiName = name;
            static abiFields = fields;
        };
    }
    switch (objectType) {
        case "boolean":
            return BoolType;
        case "string":
            return StringType;
    }
}
