// @wireio/sdk-core@1.0.18/lib/serializer/index.js
import { abiEncode } from "./Encoder";
import { abiDecode } from "./Decoder";
import { synthesizeABI } from "./Serializable";
export { ABIEncoder } from "./Encoder";
export { ABIDecoder } from "./Decoder";
export { ManualWriter } from "./Writer";
export var Serializer;
(function (Serializer) {
    Serializer.encode = abiEncode;
    Serializer.decode = abiDecode;
    /** Create an Antelope/SYSIO ABI definition for given core type. */
    function synthesize(type) {
        return synthesizeABI(type).abi;
    }
    Serializer.synthesize = synthesize;
    /** Create JSON representation of a core object. */
    function stringify(object) {
        return JSON.stringify(object);
    }
    Serializer.stringify = stringify;
    /** Create a vanilla js representation of a core object. */
    function objectify(object) {
        const walk = (v) => {
            switch (typeof v) {
                case "boolean":
                case "number":
                case "string":
                    return v;
                case "object": {
                    if (v === null) {
                        return v;
                    }
                    if (typeof v.toJSON === "function") {
                        return walk(v.toJSON());
                    }
                    if (Array.isArray(v)) {
                        return v.map(walk);
                    }
                    const rv = {};
                    for (const key of Object.keys(v)) {
                        rv[key] = walk(v[key]);
                    }
                    return rv;
                }
            }
        };
        return walk(object);
    }
    Serializer.objectify = objectify;
})(Serializer || (Serializer = {}));
