// @wireio/sdk-core@1.0.18/lib/serializer/Serializable.js
import { ABI } from "../chain/Abi";
/** Return a ABI definition for given ABISerializableType. */
export function synthesizeABI(type) {
    const structs = [];
    const variants = [];
    const aliases = [];
    const seen = new Set();
    const resolveAbiType = (t) => {
        let typeName;
        if (typeof t.type !== "string") {
            typeName = resolve(t.type);
        }
        else {
            typeName = t.type;
        }
        if (t.array === true) {
            typeName += "[]";
        }
        if (t.optional === true) {
            typeName += "?";
        }
        if (t.extension === true) {
            typeName += "$";
        }
        return typeName;
    };
    const resolve = (t) => {
        if (!t.abiName) {
            throw new Error("Encountered non-conforming type");
        }
        else if (t.abiName === "__struct") {
            throw new Error("Misconfigured Struct subclass, did you forget @Struct.type?");
        }
        if (seen.has(t)) {
            return t.abiName;
        }
        seen.add(t);
        if (t.abiAlias) {
            aliases.push({
                new_type_name: t.abiName,
                type: resolveAbiType(t.abiAlias)
            });
        }
        else if (t.abiFields) {
            const fields = t.abiFields.map(field => {
                return {
                    name: field.name,
                    type: resolveAbiType(field)
                };
            });
            const struct = {
                base: t.abiBase ? resolve(t.abiBase) : "",
                name: t.abiName,
                fields
            };
            structs.push(struct);
        }
        else if (t.abiVariant) {
            const variant = {
                name: t.abiName,
                types: t.abiVariant.map(resolveAbiType)
            };
            variants.push(variant);
        }
        return t.abiName;
    };
    const root = resolve(type);
    return {
        abi: ABI.from({ structs, variants, types: aliases }),
        types: Array.from(seen),
        root
    };
}
export function abiTypeString(type) {
    let typeName = typeof type.type === "string" ? type.type : type.type.abiName;
    if (type.array === true) {
        typeName += "[]";
    }
    if (type.optional === true) {
        typeName += "?";
    }
    if (type.extension === true) {
        typeName += "$";
    }
    return typeName;
}
export function isTypeDescriptor(type) {
    return (typeof type !== "string" &&
        type.abiName === undefined &&
        type.type !== undefined);
}
export function toTypeDescriptor(type) {
    if (typeof type === "string") {
        return { type };
    }
    if (typeof type.abiName !== "undefined") {
        return { type: type };
    }
    return type;
}
/** Returns true if the given value conforms to ABISerializableObject. */
export function isABISerializableObject(value) {
    return (value && value.constructor && typeof value.constructor.abiName === "string");
}
