// @wireio/sdk-core@1.0.18/lib/serializer/Writer.js
/** A tiny writer to do SYSIO packing exactly like eosjs.SerialBuffer */
export class ManualWriter {
    buf = [];
    writeUint32(v) {
        for (let i = 0; i < 4; i++)
            this.buf.push((v >>> (i * 8)) & 0xff);
    }
    writeUint64(v) {
        // Handle as two 32-bit parts for compatibility without BigInt
        const low = v >>> 0;
        const high = Math.floor(v / 0x100000000) >>> 0;
        // Write low 32 bits
        for (let i = 0; i < 4; i++)
            this.buf.push((low >>> (i * 8)) & 0xff);
        // Write high 32 bits
        for (let i = 0; i < 4; i++)
            this.buf.push((high >>> (i * 8)) & 0xff);
    }
    writeUint16(v) {
        this.buf.push(v & 0xff, (v >>> 8) & 0xff);
    }
    writeVarUint32(v) {
        while (v >= 0x80) {
            this.buf.push((v & 0x7f) | 0x80);
            v >>>= 7;
        }
        this.buf.push(v);
    }
    writeBytes(data) {
        for (const b of data)
            this.buf.push(b);
    }
    toUint8Array() {
        return Uint8Array.from(this.buf);
    }
}
