// @wireio/opp-solidity-models@1.0.14/sysio/opp/attestations/attestations.ts
// @generated by protobuf-ts 2.11.1
// @generated from protobuf file "sysio/opp/attestations/attestations.proto" (package "sysio.opp.attestations", syntax proto3)
// tslint:disable
import type { BinaryWriteOptions } from "@protobuf-ts/runtime";
import type { IBinaryWriter } from "@protobuf-ts/runtime";
import { WireType } from "@protobuf-ts/runtime";
import type { BinaryReadOptions } from "@protobuf-ts/runtime";
import type { IBinaryReader } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import type { PartialMessage } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Message } from "../opp";
import { ChainId } from "../types/types";
import { ChainSignature } from "../types/types";
import { OperatorStatus } from "../types/types";
import { OperatorType } from "../types/types";
import { WireAccount } from "../types/types";
import { StakeStatus } from "../types/types";
import { ChainAddress } from "../types/types";
import { TokenAmount } from "../types/types";
import { ChainKind } from "../types/types";
// ---------------------------------------------------------------------------
//  Common (Pre & Post Launch) Attestations
// ---------------------------------------------------------------------------

/**
 * Reserve balance snapshot for a given chain.
 *
 * @generated from protobuf message sysio.opp.attestations.ChainReserveBalanceSheet
 */
export interface ChainReserveBalanceSheet {
    /**
     * @generated from protobuf field: sysio.opp.types.ChainKind kind = 1
     */
    kind: ChainKind;
    /**
     * @generated from protobuf field: repeated sysio.opp.types.TokenAmount amounts = 2
     */
    amounts: TokenAmount[];
}
// ---------------------------------------------------------------------------
//  Pre-launch specific attestations
// ---------------------------------------------------------------------------

/**
 * Pre-token stake attestation.
 *
 * @generated from protobuf message sysio.opp.attestations.PretokenStakeChange
 */
export interface PretokenStakeChange {
    /**
     * @generated from protobuf field: sysio.opp.types.ChainAddress actor = 1
     */
    actor?: ChainAddress;
    /**
     * @generated from protobuf field: sysio.opp.types.TokenAmount amount = 2
     */
    amount?: TokenAmount;
    /**
     * @generated from protobuf field: int64 index_at_mint = 10
     */
    indexAtMint: bigint; // -1 == unstake action
    /**
     * @generated from protobuf field: int64 index_at_burn = 11
     */
    indexAtBurn: bigint; // -1 == stake action
}
/**
 * Pre-token purchase attestation.
 *
 * @generated from protobuf message sysio.opp.attestations.PretokenPurchase
 */
export interface PretokenPurchase {
    /**
     * @generated from protobuf field: sysio.opp.types.ChainAddress actor = 1
     */
    actor?: ChainAddress;
    /**
     * @generated from protobuf field: sysio.opp.types.TokenAmount amount = 2
     */
    amount?: TokenAmount;
    /**
     * @generated from protobuf field: int64 pretoken_count = 3
     */
    pretokenCount: bigint;
    /**
     * @generated from protobuf field: int64 index_at_mint = 10
     */
    indexAtMint: bigint;
}
/**
 * Pre-token yield distribution.
 *
 * @generated from protobuf message sysio.opp.attestations.PretokenYield
 */
export interface PretokenYield {
    /**
     * @generated from protobuf field: sysio.opp.types.ChainAddress actor = 1
     */
    actor?: ChainAddress;
    /**
     * @generated from protobuf field: sysio.opp.types.TokenAmount amount = 2
     */
    amount?: TokenAmount;
    /**
     * @generated from protobuf field: int64 index_at_mint = 3
     */
    indexAtMint: bigint;
}
// ---------------------------------------------------------------------------
//  Post-launch attestations
// ---------------------------------------------------------------------------

/**
 * Stake status update.
 *
 * @generated from protobuf message sysio.opp.attestations.StakeUpdate
 */
export interface StakeUpdate {
    /**
     * @generated from protobuf field: sysio.opp.types.ChainAddress actor = 1
     */
    actor?: ChainAddress;
    /**
     * @generated from protobuf field: sysio.opp.types.StakeStatus status = 2
     */
    status: StakeStatus;
    /**
     * @generated from protobuf field: sysio.opp.types.TokenAmount amount = 3
     */
    amount?: TokenAmount;
}
/**
 * Wire token purchase.
 *
 * @generated from protobuf message sysio.opp.attestations.WireTokenPurchase
 */
export interface WireTokenPurchase {
    /**
     * @generated from protobuf field: sysio.opp.types.ChainAddress actor = 1
     */
    actor?: ChainAddress;
    /**
     * @generated from protobuf field: repeated sysio.opp.types.TokenAmount amounts = 2
     */
    amounts: TokenAmount[];
}
/**
 * Operator registration or deregistration.
 *
 * @generated from protobuf message sysio.opp.attestations.OperatorAction
 */
export interface OperatorAction {
    /**
     * @generated from protobuf field: sysio.opp.attestations.OperatorAction.ActionType action_type = 1
     */
    actionType: OperatorAction_ActionType;
    /**
     * @generated from protobuf field: sysio.opp.types.ChainAddress actor = 2
     */
    actor?: ChainAddress;
    /**
     * @generated from protobuf field: sysio.opp.types.WireAccount wire_account = 3
     */
    wireAccount?: WireAccount;
    /**
     * @generated from protobuf field: sysio.opp.types.OperatorType type = 5
     */
    type: OperatorType;
    /**
     * @generated from protobuf field: sysio.opp.types.OperatorStatus status = 10
     */
    status: OperatorStatus;
    /**
     * @generated from protobuf field: sysio.opp.types.TokenAmount amount = 20
     */
    amount?: TokenAmount;
}
/**
 * @generated from protobuf enum sysio.opp.attestations.OperatorAction.ActionType
 */
export enum OperatorAction_ActionType {
    /**
     * @generated from protobuf enum value: ACTION_TYPE_UNKNOWN = 0;
     */
    UNKNOWN = 0,
    /**
     * @generated from protobuf enum value: ACTION_TYPE_DEPOSIT = 1;
     */
    DEPOSIT = 1,
    /**
     * @generated from protobuf enum value: ACTION_TYPE_WITHDRAW = 2;
     */
    WITHDRAW = 2
}
/**
 * Reserve fund disbursement.
 *
 * @generated from protobuf message sysio.opp.attestations.ReserveDisbursement
 */
export interface ReserveDisbursement {
    /**
     * @generated from protobuf field: sysio.opp.types.ChainAddress actor = 1
     */
    actor?: ChainAddress;
    /**
     * @generated from protobuf field: sysio.opp.types.TokenAmount amount = 2
     */
    amount?: TokenAmount;
    /**
     * @generated from protobuf field: repeated sysio.opp.types.ChainSignature signature = 3
     */
    signature: ChainSignature[];
}
// ---------------------------------------------------------------------------
//  Protocol state tracking
// ---------------------------------------------------------------------------

/**
 * Protocol state tracking for a message chain on a given chain.
 *
 * @generated from protobuf message sysio.opp.attestations.ProtocolState
 */
export interface ProtocolState {
    /**
     * @generated from protobuf field: sysio.opp.types.ChainId chain_id = 1
     */
    chainId?: ChainId;
    /**
     * @generated from protobuf field: bytes current_message_id = 2
     */
    currentMessageId: Uint8Array; // 32 bytes
    /**
     * @generated from protobuf field: bytes processed_message_id = 3
     */
    processedMessageId: Uint8Array; // 32 bytes
    /**
     * @generated from protobuf field: repeated sysio.opp.Message incoming_messages = 10
     */
    incomingMessages: Message[];
    /**
     * @generated from protobuf field: repeated sysio.opp.Message outgoing_messages = 11
     */
    outgoingMessages: Message[];
}
// @generated message type with reflection information, may provide speed optimized methods
class ChainReserveBalanceSheet$Type extends MessageType<ChainReserveBalanceSheet> {
    constructor() {
        super("sysio.opp.attestations.ChainReserveBalanceSheet", [
            { no: 1, name: "kind", kind: "enum", T: () => ["sysio.opp.types.ChainKind", ChainKind, "CHAIN_KIND_"] },
            { no: 2, name: "amounts", kind: "message", repeat: 2 /*RepeatType.UNPACKED*/, T: () => TokenAmount }
        ]);
    }
    create(value?: PartialMessage<ChainReserveBalanceSheet>): ChainReserveBalanceSheet {
        const message = globalThis.Object.create((this.messagePrototype!));
        message.kind = 0;
        message.amounts = [];
        if (value !== undefined)
            reflectionMergePartial<ChainReserveBalanceSheet>(this, message, value);
        return message;
    }
    internalBinaryRead(reader: IBinaryReader, length: number, options: BinaryReadOptions, target?: ChainReserveBalanceSheet): ChainReserveBalanceSheet {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* sysio.opp.types.ChainKind kind */ 1:
                    message.kind = reader.int32();
                    break;
                case /* repeated sysio.opp.types.TokenAmount amounts */ 2:
                    message.amounts.push(TokenAmount.internalBinaryRead(reader, reader.uint32(), options));
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message: ChainReserveBalanceSheet, writer: IBinaryWriter, options: BinaryWriteOptions): IBinaryWriter {
        /* sysio.opp.types.ChainKind kind = 1; */
        if (message.kind !== 0)
            writer.tag(1, WireType.Varint).int32(message.kind);
        /* repeated sysio.opp.types.TokenAmount amounts = 2; */
        for (let i = 0; i < message.amounts.length; i++)
            TokenAmount.internalBinaryWrite(message.amounts[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message sysio.opp.attestations.ChainReserveBalanceSheet
 */
export const ChainReserveBalanceSheet = new ChainReserveBalanceSheet$Type();
// @generated message type with reflection information, may provide speed optimized methods
class PretokenStakeChange$Type extends MessageType<PretokenStakeChange> {
    constructor() {
        super("sysio.opp.attestations.PretokenStakeChange", [
            { no: 1, name: "actor", kind: "message", T: () => ChainAddress },
            { no: 2, name: "amount", kind: "message", T: () => TokenAmount },
            { no: 10, name: "index_at_mint", kind: "scalar", T: 3 /*ScalarType.INT64*/, L: 0 /*LongType.BIGINT*/ },
            { no: 11, name: "index_at_burn", kind: "scalar", T: 3 /*ScalarType.INT64*/, L: 0 /*LongType.BIGINT*/ }
        ]);
    }
    create(value?: PartialMessage<PretokenStakeChange>): PretokenStakeChange {
        const message = globalThis.Object.create((this.messagePrototype!));
        message.indexAtMint = 0n;
        message.indexAtBurn = 0n;
        if (value !== undefined)
            reflectionMergePartial<PretokenStakeChange>(this, message, value);
        return message;
    }
    internalBinaryRead(reader: IBinaryReader, length: number, options: BinaryReadOptions, target?: PretokenStakeChange): PretokenStakeChange {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* sysio.opp.types.ChainAddress actor */ 1:
                    message.actor = ChainAddress.internalBinaryRead(reader, reader.uint32(), options, message.actor);
                    break;
                case /* sysio.opp.types.TokenAmount amount */ 2:
                    message.amount = TokenAmount.internalBinaryRead(reader, reader.uint32(), options, message.amount);
                    break;
                case /* int64 index_at_mint */ 10:
                    message.indexAtMint = reader.int64().toBigInt();
                    break;
                case /* int64 index_at_burn */ 11:
                    message.indexAtBurn = reader.int64().toBigInt();
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message: PretokenStakeChange, writer: IBinaryWriter, options: BinaryWriteOptions): IBinaryWriter {
        /* sysio.opp.types.ChainAddress actor = 1; */
        if (message.actor)
            ChainAddress.internalBinaryWrite(message.actor, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
        /* sysio.opp.types.TokenAmount amount = 2; */
        if (message.amount)
            TokenAmount.internalBinaryWrite(message.amount, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
        /* int64 index_at_mint = 10; */
        if (message.indexAtMint !== 0n)
            writer.tag(10, WireType.Varint).int64(message.indexAtMint);
        /* int64 index_at_burn = 11; */
        if (message.indexAtBurn !== 0n)
            writer.tag(11, WireType.Varint).int64(message.indexAtBurn);
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message sysio.opp.attestations.PretokenStakeChange
 */
export const PretokenStakeChange = new PretokenStakeChange$Type();
// @generated message type with reflection information, may provide speed optimized methods
class PretokenPurchase$Type extends MessageType<PretokenPurchase> {
    constructor() {
        super("sysio.opp.attestations.PretokenPurchase", [
            { no: 1, name: "actor", kind: "message", T: () => ChainAddress },
            { no: 2, name: "amount", kind: "message", T: () => TokenAmount },
            { no: 3, name: "pretoken_count", kind: "scalar", T: 3 /*ScalarType.INT64*/, L: 0 /*LongType.BIGINT*/ },
            { no: 10, name: "index_at_mint", kind: "scalar", T: 3 /*ScalarType.INT64*/, L: 0 /*LongType.BIGINT*/ }
        ]);
    }
    create(value?: PartialMessage<PretokenPurchase>): PretokenPurchase {
        const message = globalThis.Object.create((this.messagePrototype!));
        message.pretokenCount = 0n;
        message.indexAtMint = 0n;
        if (value !== undefined)
            reflectionMergePartial<PretokenPurchase>(this, message, value);
        return message;
    }
    internalBinaryRead(reader: IBinaryReader, length: number, options: BinaryReadOptions, target?: PretokenPurchase): PretokenPurchase {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* sysio.opp.types.ChainAddress actor */ 1:
                    message.actor = ChainAddress.internalBinaryRead(reader, reader.uint32(), options, message.actor);
                    break;
                case /* sysio.opp.types.TokenAmount amount */ 2:
                    message.amount = TokenAmount.internalBinaryRead(reader, reader.uint32(), options, message.amount);
                    break;
                case /* int64 pretoken_count */ 3:
                    message.pretokenCount = reader.int64().toBigInt();
                    break;
                case /* int64 index_at_mint */ 10:
                    message.indexAtMint = reader.int64().toBigInt();
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message: PretokenPurchase, writer: IBinaryWriter, options: BinaryWriteOptions): IBinaryWriter {
        /* sysio.opp.types.ChainAddress actor = 1; */
        if (message.actor)
            ChainAddress.internalBinaryWrite(message.actor, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
        /* sysio.opp.types.TokenAmount amount = 2; */
        if (message.amount)
            TokenAmount.internalBinaryWrite(message.amount, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
        /* int64 pretoken_count = 3; */
        if (message.pretokenCount !== 0n)
            writer.tag(3, WireType.Varint).int64(message.pretokenCount);
        /* int64 index_at_mint = 10; */
        if (message.indexAtMint !== 0n)
            writer.tag(10, WireType.Varint).int64(message.indexAtMint);
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message sysio.opp.attestations.PretokenPurchase
 */
export const PretokenPurchase = new PretokenPurchase$Type();
// @generated message type with reflection information, may provide speed optimized methods
class PretokenYield$Type extends MessageType<PretokenYield> {
    constructor() {
        super("sysio.opp.attestations.PretokenYield", [
            { no: 1, name: "actor", kind: "message", T: () => ChainAddress },
            { no: 2, name: "amount", kind: "message", T: () => TokenAmount },
            { no: 3, name: "index_at_mint", kind: "scalar", T: 3 /*ScalarType.INT64*/, L: 0 /*LongType.BIGINT*/ }
        ]);
    }
    create(value?: PartialMessage<PretokenYield>): PretokenYield {
        const message = globalThis.Object.create((this.messagePrototype!));
        message.indexAtMint = 0n;
        if (value !== undefined)
            reflectionMergePartial<PretokenYield>(this, message, value);
        return message;
    }
    internalBinaryRead(reader: IBinaryReader, length: number, options: BinaryReadOptions, target?: PretokenYield): PretokenYield {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* sysio.opp.types.ChainAddress actor */ 1:
                    message.actor = ChainAddress.internalBinaryRead(reader, reader.uint32(), options, message.actor);
                    break;
                case /* sysio.opp.types.TokenAmount amount */ 2:
                    message.amount = TokenAmount.internalBinaryRead(reader, reader.uint32(), options, message.amount);
                    break;
                case /* int64 index_at_mint */ 3:
                    message.indexAtMint = reader.int64().toBigInt();
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message: PretokenYield, writer: IBinaryWriter, options: BinaryWriteOptions): IBinaryWriter {
        /* sysio.opp.types.ChainAddress actor = 1; */
        if (message.actor)
            ChainAddress.internalBinaryWrite(message.actor, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
        /* sysio.opp.types.TokenAmount amount = 2; */
        if (message.amount)
            TokenAmount.internalBinaryWrite(message.amount, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
        /* int64 index_at_mint = 3; */
        if (message.indexAtMint !== 0n)
            writer.tag(3, WireType.Varint).int64(message.indexAtMint);
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message sysio.opp.attestations.PretokenYield
 */
export const PretokenYield = new PretokenYield$Type();
// @generated message type with reflection information, may provide speed optimized methods
class StakeUpdate$Type extends MessageType<StakeUpdate> {
    constructor() {
        super("sysio.opp.attestations.StakeUpdate", [
            { no: 1, name: "actor", kind: "message", T: () => ChainAddress },
            { no: 2, name: "status", kind: "enum", T: () => ["sysio.opp.types.StakeStatus", StakeStatus, "STAKE_STATUS_"] },
            { no: 3, name: "amount", kind: "message", T: () => TokenAmount }
        ]);
    }
    create(value?: PartialMessage<StakeUpdate>): StakeUpdate {
        const message = globalThis.Object.create((this.messagePrototype!));
        message.status = 0;
        if (value !== undefined)
            reflectionMergePartial<StakeUpdate>(this, message, value);
        return message;
    }
    internalBinaryRead(reader: IBinaryReader, length: number, options: BinaryReadOptions, target?: StakeUpdate): StakeUpdate {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* sysio.opp.types.ChainAddress actor */ 1:
                    message.actor = ChainAddress.internalBinaryRead(reader, reader.uint32(), options, message.actor);
                    break;
                case /* sysio.opp.types.StakeStatus status */ 2:
                    message.status = reader.int32();
                    break;
                case /* sysio.opp.types.TokenAmount amount */ 3:
                    message.amount = TokenAmount.internalBinaryRead(reader, reader.uint32(), options, message.amount);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message: StakeUpdate, writer: IBinaryWriter, options: BinaryWriteOptions): IBinaryWriter {
        /* sysio.opp.types.ChainAddress actor = 1; */
        if (message.actor)
            ChainAddress.internalBinaryWrite(message.actor, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
        /* sysio.opp.types.StakeStatus status = 2; */
        if (message.status !== 0)
            writer.tag(2, WireType.Varint).int32(message.status);
        /* sysio.opp.types.TokenAmount amount = 3; */
        if (message.amount)
            TokenAmount.internalBinaryWrite(message.amount, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message sysio.opp.attestations.StakeUpdate
 */
export const StakeUpdate = new StakeUpdate$Type();
// @generated message type with reflection information, may provide speed optimized methods
class WireTokenPurchase$Type extends MessageType<WireTokenPurchase> {
    constructor() {
        super("sysio.opp.attestations.WireTokenPurchase", [
            { no: 1, name: "actor", kind: "message", T: () => ChainAddress },
            { no: 2, name: "amounts", kind: "message", repeat: 2 /*RepeatType.UNPACKED*/, T: () => TokenAmount }
        ]);
    }
    create(value?: PartialMessage<WireTokenPurchase>): WireTokenPurchase {
        const message = globalThis.Object.create((this.messagePrototype!));
        message.amounts = [];
        if (value !== undefined)
            reflectionMergePartial<WireTokenPurchase>(this, message, value);
        return message;
    }
    internalBinaryRead(reader: IBinaryReader, length: number, options: BinaryReadOptions, target?: WireTokenPurchase): WireTokenPurchase {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* sysio.opp.types.ChainAddress actor */ 1:
                    message.actor = ChainAddress.internalBinaryRead(reader, reader.uint32(), options, message.actor);
                    break;
                case /* repeated sysio.opp.types.TokenAmount amounts */ 2:
                    message.amounts.push(TokenAmount.internalBinaryRead(reader, reader.uint32(), options));
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message: WireTokenPurchase, writer: IBinaryWriter, options: BinaryWriteOptions): IBinaryWriter {
        /* sysio.opp.types.ChainAddress actor = 1; */
        if (message.actor)
            ChainAddress.internalBinaryWrite(message.actor, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
        /* repeated sysio.opp.types.TokenAmount amounts = 2; */
        for (let i = 0; i < message.amounts.length; i++)
            TokenAmount.internalBinaryWrite(message.amounts[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message sysio.opp.attestations.WireTokenPurchase
 */
export const WireTokenPurchase = new WireTokenPurchase$Type();
// @generated message type with reflection information, may provide speed optimized methods
class OperatorAction$Type extends MessageType<OperatorAction> {
    constructor() {
        super("sysio.opp.attestations.OperatorAction", [
            { no: 1, name: "action_type", kind: "enum", T: () => ["sysio.opp.attestations.OperatorAction.ActionType", OperatorAction_ActionType, "ACTION_TYPE_"] },
            { no: 2, name: "actor", kind: "message", T: () => ChainAddress },
            { no: 3, name: "wire_account", kind: "message", T: () => WireAccount },
            { no: 5, name: "type", kind: "enum", T: () => ["sysio.opp.types.OperatorType", OperatorType, "OPERATOR_TYPE_"] },
            { no: 10, name: "status", kind: "enum", T: () => ["sysio.opp.types.OperatorStatus", OperatorStatus, "OPERATOR_STATUS_"] },
            { no: 20, name: "amount", kind: "message", T: () => TokenAmount }
        ]);
    }
    create(value?: PartialMessage<OperatorAction>): OperatorAction {
        const message = globalThis.Object.create((this.messagePrototype!));
        message.actionType = 0;
        message.type = 0;
        message.status = 0;
        if (value !== undefined)
            reflectionMergePartial<OperatorAction>(this, message, value);
        return message;
    }
    internalBinaryRead(reader: IBinaryReader, length: number, options: BinaryReadOptions, target?: OperatorAction): OperatorAction {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* sysio.opp.attestations.OperatorAction.ActionType action_type */ 1:
                    message.actionType = reader.int32();
                    break;
                case /* sysio.opp.types.ChainAddress actor */ 2:
                    message.actor = ChainAddress.internalBinaryRead(reader, reader.uint32(), options, message.actor);
                    break;
                case /* sysio.opp.types.WireAccount wire_account */ 3:
                    message.wireAccount = WireAccount.internalBinaryRead(reader, reader.uint32(), options, message.wireAccount);
                    break;
                case /* sysio.opp.types.OperatorType type */ 5:
                    message.type = reader.int32();
                    break;
                case /* sysio.opp.types.OperatorStatus status */ 10:
                    message.status = reader.int32();
                    break;
                case /* sysio.opp.types.TokenAmount amount */ 20:
                    message.amount = TokenAmount.internalBinaryRead(reader, reader.uint32(), options, message.amount);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message: OperatorAction, writer: IBinaryWriter, options: BinaryWriteOptions): IBinaryWriter {
        /* sysio.opp.attestations.OperatorAction.ActionType action_type = 1; */
        if (message.actionType !== 0)
            writer.tag(1, WireType.Varint).int32(message.actionType);
        /* sysio.opp.types.ChainAddress actor = 2; */
        if (message.actor)
            ChainAddress.internalBinaryWrite(message.actor, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
        /* sysio.opp.types.WireAccount wire_account = 3; */
        if (message.wireAccount)
            WireAccount.internalBinaryWrite(message.wireAccount, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
        /* sysio.opp.types.OperatorType type = 5; */
        if (message.type !== 0)
            writer.tag(5, WireType.Varint).int32(message.type);
        /* sysio.opp.types.OperatorStatus status = 10; */
        if (message.status !== 0)
            writer.tag(10, WireType.Varint).int32(message.status);
        /* sysio.opp.types.TokenAmount amount = 20; */
        if (message.amount)
            TokenAmount.internalBinaryWrite(message.amount, writer.tag(20, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message sysio.opp.attestations.OperatorAction
 */
export const OperatorAction = new OperatorAction$Type();
// @generated message type with reflection information, may provide speed optimized methods
class ReserveDisbursement$Type extends MessageType<ReserveDisbursement> {
    constructor() {
        super("sysio.opp.attestations.ReserveDisbursement", [
            { no: 1, name: "actor", kind: "message", T: () => ChainAddress },
            { no: 2, name: "amount", kind: "message", T: () => TokenAmount },
            { no: 3, name: "signature", kind: "message", repeat: 2 /*RepeatType.UNPACKED*/, T: () => ChainSignature }
        ]);
    }
    create(value?: PartialMessage<ReserveDisbursement>): ReserveDisbursement {
        const message = globalThis.Object.create((this.messagePrototype!));
        message.signature = [];
        if (value !== undefined)
            reflectionMergePartial<ReserveDisbursement>(this, message, value);
        return message;
    }
    internalBinaryRead(reader: IBinaryReader, length: number, options: BinaryReadOptions, target?: ReserveDisbursement): ReserveDisbursement {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* sysio.opp.types.ChainAddress actor */ 1:
                    message.actor = ChainAddress.internalBinaryRead(reader, reader.uint32(), options, message.actor);
                    break;
                case /* sysio.opp.types.TokenAmount amount */ 2:
                    message.amount = TokenAmount.internalBinaryRead(reader, reader.uint32(), options, message.amount);
                    break;
                case /* repeated sysio.opp.types.ChainSignature signature */ 3:
                    message.signature.push(ChainSignature.internalBinaryRead(reader, reader.uint32(), options));
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message: ReserveDisbursement, writer: IBinaryWriter, options: BinaryWriteOptions): IBinaryWriter {
        /* sysio.opp.types.ChainAddress actor = 1; */
        if (message.actor)
            ChainAddress.internalBinaryWrite(message.actor, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
        /* sysio.opp.types.TokenAmount amount = 2; */
        if (message.amount)
            TokenAmount.internalBinaryWrite(message.amount, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
        /* repeated sysio.opp.types.ChainSignature signature = 3; */
        for (let i = 0; i < message.signature.length; i++)
            ChainSignature.internalBinaryWrite(message.signature[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message sysio.opp.attestations.ReserveDisbursement
 */
export const ReserveDisbursement = new ReserveDisbursement$Type();
// @generated message type with reflection information, may provide speed optimized methods
class ProtocolState$Type extends MessageType<ProtocolState> {
    constructor() {
        super("sysio.opp.attestations.ProtocolState", [
            { no: 1, name: "chain_id", kind: "message", T: () => ChainId },
            { no: 2, name: "current_message_id", kind: "scalar", T: 12 /*ScalarType.BYTES*/ },
            { no: 3, name: "processed_message_id", kind: "scalar", T: 12 /*ScalarType.BYTES*/ },
            { no: 10, name: "incoming_messages", kind: "message", repeat: 2 /*RepeatType.UNPACKED*/, T: () => Message },
            { no: 11, name: "outgoing_messages", kind: "message", repeat: 2 /*RepeatType.UNPACKED*/, T: () => Message }
        ]);
    }
    create(value?: PartialMessage<ProtocolState>): ProtocolState {
        const message = globalThis.Object.create((this.messagePrototype!));
        message.currentMessageId = new Uint8Array(0);
        message.processedMessageId = new Uint8Array(0);
        message.incomingMessages = [];
        message.outgoingMessages = [];
        if (value !== undefined)
            reflectionMergePartial<ProtocolState>(this, message, value);
        return message;
    }
    internalBinaryRead(reader: IBinaryReader, length: number, options: BinaryReadOptions, target?: ProtocolState): ProtocolState {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* sysio.opp.types.ChainId chain_id */ 1:
                    message.chainId = ChainId.internalBinaryRead(reader, reader.uint32(), options, message.chainId);
                    break;
                case /* bytes current_message_id */ 2:
                    message.currentMessageId = reader.bytes();
                    break;
                case /* bytes processed_message_id */ 3:
                    message.processedMessageId = reader.bytes();
                    break;
                case /* repeated sysio.opp.Message incoming_messages */ 10:
                    message.incomingMessages.push(Message.internalBinaryRead(reader, reader.uint32(), options));
                    break;
                case /* repeated sysio.opp.Message outgoing_messages */ 11:
                    message.outgoingMessages.push(Message.internalBinaryRead(reader, reader.uint32(), options));
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message: ProtocolState, writer: IBinaryWriter, options: BinaryWriteOptions): IBinaryWriter {
        /* sysio.opp.types.ChainId chain_id = 1; */
        if (message.chainId)
            ChainId.internalBinaryWrite(message.chainId, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
        /* bytes current_message_id = 2; */
        if (message.currentMessageId.length)
            writer.tag(2, WireType.LengthDelimited).bytes(message.currentMessageId);
        /* bytes processed_message_id = 3; */
        if (message.processedMessageId.length)
            writer.tag(3, WireType.LengthDelimited).bytes(message.processedMessageId);
        /* repeated sysio.opp.Message incoming_messages = 10; */
        for (let i = 0; i < message.incomingMessages.length; i++)
            Message.internalBinaryWrite(message.incomingMessages[i], writer.tag(10, WireType.LengthDelimited).fork(), options).join();
        /* repeated sysio.opp.Message outgoing_messages = 11; */
        for (let i = 0; i < message.outgoingMessages.length; i++)
            Message.internalBinaryWrite(message.outgoingMessages[i], writer.tag(11, WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message sysio.opp.attestations.ProtocolState
 */
export const ProtocolState = new ProtocolState$Type();
