// @wireio/sdk-core@1.0.18/lib/chain/Abi.js
import { isInstanceOf } from "../Utils";
import { abiDecode } from "../serializer/Decoder";
import { abiEncode } from "../serializer/Encoder";
import { Blob } from "./Blob";
import { Name } from "./Name";
export class ABI {
    static abiName = "abi";
    static version = "sysio::abi/1.1";
    version;
    /// List of type aliases.
    types;
    /// List of variant types.
    variants;
    /// List of struct types.
    structs;
    /// List of contract actions.
    actions;
    /// List of contract tables.
    tables;
    /// Ricardian contracts.
    ricardian_clauses;
    /// Action Results
    action_results;
    /// List of enum types.
    enums;
    constructor(args) {
        this.version = args.version || ABI.version;
        this.types = args.types || [];
        this.variants = args.variants || [];
        this.structs = args.structs || [];
        this.actions = args.actions || [];
        this.tables = args.tables || [];
        this.ricardian_clauses = args.ricardian_clauses || [];
        this.action_results = args.action_results || [];
        this.enums = args.enums || [];
    }
    static from(value) {
        if (isInstanceOf(value, ABI)) {
            return value;
        }
        if (isInstanceOf(value, Blob)) {
            return abiDecode({
                data: value.array,
                type: this
            });
        }
        if (typeof value === "string") {
            return new ABI(JSON.parse(value));
        }
        return new ABI(value);
    }
    static fromABI(decoder) {
        const version = decoder.readString();
        const types = [];
        const numTypes = decoder.readVaruint32();
        for (let i = 0; i < numTypes; i++) {
            types.push({
                new_type_name: decoder.readString(),
                type: decoder.readString()
            });
        }
        const structs = [];
        const numStructs = decoder.readVaruint32();
        for (let i = 0; i < numStructs; i++) {
            const name = decoder.readString();
            const base = decoder.readString();
            const numFields = decoder.readVaruint32();
            const fields = [];
            for (let j = 0; j < numFields; j++) {
                fields.push({ name: decoder.readString(), type: decoder.readString() });
            }
            structs.push({ base, name, fields });
        }
        const actions = [];
        const numActions = decoder.readVaruint32();
        for (let i = 0; i < numActions; i++) {
            const name = Name.fromABI(decoder);
            const type = decoder.readString();
            const ricardian_contract = decoder.readString();
            actions.push({ name, type, ricardian_contract });
        }
        const tables = [];
        const numTables = decoder.readVaruint32();
        for (let i = 0; i < numTables; i++) {
            const name = Name.fromABI(decoder);
            const index_type = decoder.readString();
            const key_names = [];
            const numKeyNames = decoder.readVaruint32();
            for (let j = 0; j < numKeyNames; j++) {
                key_names.push(decoder.readString());
            }
            const key_types = [];
            const numKeyTypes = decoder.readVaruint32();
            for (let j = 0; j < numKeyTypes; j++) {
                key_types.push(decoder.readString());
            }
            const type = decoder.readString();
            tables.push({ name, index_type, key_names, key_types, type });
        }
        const ricardian_clauses = [];
        const numClauses = decoder.readVaruint32();
        for (let i = 0; i < numClauses; i++) {
            const id = decoder.readString();
            const body = decoder.readString();
            ricardian_clauses.push({ id, body });
        }
        // error_messages, never used?
        const numErrors = decoder.readVaruint32();
        for (let i = 0; i < numErrors; i++) {
            decoder.advance(8); // uint64 error_code
            decoder.advance(decoder.readVaruint32()); // string error_msgr
        }
        // extensions, not used
        const numExtensions = decoder.readVaruint32();
        for (let i = 0; i < numExtensions; i++) {
            decoder.advance(2); // uint16 type
            decoder.advance(decoder.readVaruint32()); // bytes data
        }
        // variants is a binary extension for some reason even though extensions are defined on the type
        const variants = [];
        if (decoder.canRead()) {
            const numVariants = decoder.readVaruint32();
            for (let i = 0; i < numVariants; i++) {
                const name = decoder.readString();
                const types = [];
                const numTypes = decoder.readVaruint32();
                for (let j = 0; j < numTypes; j++) {
                    types.push(decoder.readString());
                }
                variants.push({ name, types });
            }
        }
        const action_results = [];
        if (decoder.canRead()) {
            const numActionResults = decoder.readVaruint32();
            for (let i = 0; i < numActionResults; i++) {
                const name = Name.fromABI(decoder);
                const result_type = decoder.readString();
                action_results.push({ name, result_type });
            }
        }
        const enums = [];
        if (decoder.canRead()) {
            const numEnums = decoder.readVaruint32();
            for (let i = 0; i < numEnums; i++) {
                const name = decoder.readString();
                const type = decoder.readString();
                const values = [];
                const numValues = decoder.readVaruint32();
                for (let j = 0; j < numValues; j++) {
                    const vname = decoder.readString();
                    // int64 value - read as two uint32s (little-endian), reconstruct as signed
                    const lo = decoder.readArray(4);
                    const hi = decoder.readArray(4);
                    const loVal = lo[0] | (lo[1] << 8) | (lo[2] << 16) | ((lo[3] << 24) >>> 0);
                    const hiVal = hi[0] | (hi[1] << 8) | (hi[2] << 16) | ((hi[3] << 24) >>> 0);
                    // Reconstruct as signed: if high bit set, value is negative
                    let value;
                    if (hiVal >= 0x80000000) {
                        // Negative: compute -(2^64 - raw)
                        value = -(0x100000000 * (0x100000000 - hiVal) - loVal);
                    }
                    else {
                        value = hiVal * 0x100000000 + loVal;
                    }
                    values.push({ name: vname, value });
                }
                enums.push({ name, type, values });
            }
        }
        return new ABI({
            version,
            types,
            structs,
            actions,
            tables,
            ricardian_clauses,
            variants,
            action_results,
            enums
        });
    }
    toABI(encoder) {
        encoder.writeString(this.version);
        encoder.writeVaruint32(this.types.length);
        for (const type of this.types) {
            encoder.writeString(type.new_type_name);
            encoder.writeString(type.type);
        }
        encoder.writeVaruint32(this.structs.length);
        for (const struct of this.structs) {
            encoder.writeString(struct.name);
            encoder.writeString(struct.base);
            encoder.writeVaruint32(struct.fields.length);
            for (const field of struct.fields) {
                encoder.writeString(field.name);
                encoder.writeString(field.type);
            }
        }
        encoder.writeVaruint32(this.actions.length);
        for (const action of this.actions) {
            Name.from(action.name).toABI(encoder);
            encoder.writeString(action.type);
            encoder.writeString(action.ricardian_contract);
        }
        encoder.writeVaruint32(this.tables.length);
        for (const table of this.tables) {
            Name.from(table.name).toABI(encoder);
            encoder.writeString(table.index_type);
            encoder.writeVaruint32(table.key_names.length);
            for (const key of table.key_names) {
                encoder.writeString(key);
            }
            encoder.writeVaruint32(table.key_types.length);
            for (const key of table.key_types) {
                encoder.writeString(key);
            }
            encoder.writeString(table.type);
        }
        encoder.writeVaruint32(this.ricardian_clauses.length);
        for (const clause of this.ricardian_clauses) {
            encoder.writeString(clause.id);
            encoder.writeString(clause.body);
        }
        encoder.writeVaruint32(0); // error_messages
        encoder.writeVaruint32(0); // extensions
        encoder.writeVaruint32(this.variants.length);
        for (const variant of this.variants) {
            encoder.writeString(variant.name);
            encoder.writeVaruint32(variant.types.length);
            for (const type of variant.types) {
                encoder.writeString(type);
            }
        }
        encoder.writeVaruint32(this.action_results.length);
        for (const result of this.action_results) {
            Name.from(result.name).toABI(encoder);
            encoder.writeString(result.result_type);
        }
        encoder.writeVaruint32(this.enums.length);
        for (const enumDef of this.enums) {
            encoder.writeString(enumDef.name);
            encoder.writeString(enumDef.type);
            encoder.writeVaruint32(enumDef.values.length);
            for (const ev of enumDef.values) {
                encoder.writeString(ev.name);
                // int64 value as two uint32 little-endian
                const raw = ev.value < 0 ? ev.value + 0x10000000000000000 : ev.value;
                const lo = raw >>> 0;
                const hi = (raw / 0x100000000) >>> 0;
                const buf = new Uint8Array(8);
                buf[0] = lo & 0xff;
                buf[1] = (lo >> 8) & 0xff;
                buf[2] = (lo >> 16) & 0xff;
                buf[3] = (lo >> 24) & 0xff;
                buf[4] = hi & 0xff;
                buf[5] = (hi >> 8) & 0xff;
                buf[6] = (hi >> 16) & 0xff;
                buf[7] = (hi >> 24) & 0xff;
                encoder.writeArray(buf);
            }
        }
    }
    resolveType(name) {
        const types = {};
        return this.resolve({ name, types }, { id: 0 });
    }
    resolveAll() {
        const types = {};
        const ctx = { id: 0 };
        return {
            types: this.types.map(t => this.resolve({ name: t.new_type_name, types }, ctx)),
            variants: this.variants.map(t => this.resolve({ name: t.name, types }, ctx)),
            structs: this.structs.map(t => this.resolve({ name: t.name, types }, ctx)),
            enums: this.enums.map(t => this.resolve({ name: t.name, types }, ctx))
        };
    }
    resolve({ name, types }, ctx) {
        const existing = types[name];
        if (existing) {
            return existing;
        }
        const type = new ABI.ResolvedType(name, ++ctx.id);
        types[type.typeName] = type;
        const alias = this.types.find(typeDef => typeDef.new_type_name == type.name);
        if (alias) {
            type.ref = this.resolve({ name: alias.type, types }, ctx);
            return type;
        }
        const struct = this.getStruct(type.name);
        if (struct) {
            if (struct.base) {
                type.base = this.resolve({ name: struct.base, types }, ctx);
            }
            type.fields = struct.fields.map(field => {
                return {
                    name: field.name,
                    type: this.resolve({ name: field.type, types }, ctx)
                };
            });
            return type;
        }
        const variant = this.getVariant(type.name);
        if (variant) {
            type.variant = variant.types.map(name => this.resolve({ name, types }, ctx));
            return type;
        }
        const enumDef = this.getEnum(type.name);
        if (enumDef) {
            type.enum = {
                type: enumDef.type,
                values: enumDef.values
            };
            return type;
        }
        // builtin or unknown type
        return type;
    }
    getStruct(name) {
        return this.structs.find(struct => struct.name == name);
    }
    getVariant(name) {
        return this.variants.find(variant => variant.name == name);
    }
    getEnum(name) {
        return this.enums.find(e => e.name == name);
    }
    /** Return arguments type of an action in this ABI. */
    getActionType(actionName) {
        const name = Name.from(actionName).toString();
        const action = this.actions.find(a => a.name.toString() === name);
        if (action) {
            return action.type;
        }
    }
    equals(other) {
        const o = ABI.from(other);
        if (this.version != o.version ||
            this.types.length != o.types.length ||
            this.structs.length != o.structs.length ||
            this.actions.length != o.actions.length ||
            this.tables.length != o.tables.length ||
            this.ricardian_clauses.length != o.ricardian_clauses.length ||
            this.variants.length != o.variants.length ||
            this.action_results.length != o.action_results.length ||
            this.enums.length != o.enums.length) {
            return false;
        }
        return abiEncode({ object: this }).equals(abiEncode({ object: o }));
    }
    toJSON() {
        return {
            version: this.version,
            types: this.types,
            structs: this.structs,
            actions: this.actions,
            tables: this.tables,
            ricardian_clauses: this.ricardian_clauses,
            error_messages: [],
            abi_extensions: [],
            variants: this.variants,
            action_results: this.action_results,
            enums: this.enums
        };
    }
}
(function (ABI) {
    class ResolvedType {
        name;
        id;
        isArray;
        isOptional;
        isExtension;
        base;
        fields;
        variant;
        ref;
        enum;
        constructor(fullName, id = 0) {
            let name = fullName;
            if (name.endsWith("$")) {
                name = name.slice(0, -1);
                this.isExtension = true;
            }
            else {
                this.isExtension = false;
            }
            if (name.endsWith("?")) {
                name = name.slice(0, -1);
                this.isOptional = true;
            }
            else {
                this.isOptional = false;
            }
            if (name.endsWith("[]")) {
                name = name.slice(0, -2);
                this.isArray = true;
            }
            else {
                this.isArray = false;
            }
            this.id = id;
            this.name = name;
        }
        /**
         * Type name including suffixes: [] array, ? optional, $ binary ext
         */
        get typeName() {
            let rv = this.name;
            if (this.isArray) {
                rv += "[]";
            }
            if (this.isOptional) {
                rv += "?";
            }
            if (this.isExtension) {
                rv += "$";
            }
            return rv;
        }
        /** All fields including base struct(s), undefined if not a struct type. */
        get allFields() {
            // eslint-disable-next-line @typescript-eslint/no-this-alias
            let current = this;
            const rv = [];
            const seen = new Set();
            do {
                if (!current.fields) {
                    return; // invalid struct
                }
                if (seen.has(current.name)) {
                    return; // circular ref
                }
                for (let i = current.fields.length - 1; i >= 0; i--) {
                    rv.unshift(current.fields[i]);
                }
                seen.add(current.name);
                current = current.base;
            } while (current !== undefined);
            return rv;
        }
    }
    ABI.ResolvedType = ResolvedType;
})(ABI || (ABI = {}));
