// @wireio/sdk-core@1.0.18/lib/chain/Action.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var Action_1;
import { abiEncode } from "../serializer/Encoder";
import { abiDecode } from "../serializer/Decoder";
import { synthesizeABI } from "../serializer/Serializable";
import { arrayEquatableEquals } from "../Utils";
import { getType } from "../serializer/Builtins";
import { ABI } from "./Abi";
import { Bytes } from "./Bytes";
import { Name } from "./Name";
import { PermissionLevel } from "./PermissionLevel";
import { Struct } from "./Struct";
let Action = Action_1 = class Action extends Struct {
    abi;
    static from(anyAction, abi) {
        let object = { ...anyAction };
        const data = object.data;
        if (!Bytes.isBytes(data)) {
            let type;
            if (abi) {
                type = ABI.from(abi).getActionType(object.name);
            }
            else if (!data.constructor || data.constructor.abiName === undefined) {
                throw new Error("Missing ABI definition when creating action with untyped action data");
            }
            object = {
                ...object,
                data: abiEncode({ object: data, type, abi })
            };
        }
        const action = super.from(object);
        if (abi) {
            action.abi = ABI.from(abi);
        }
        else {
            const type = getType(data);
            if (type) {
                action.abi = ABI.from({
                    ...synthesizeABI(type).abi,
                    actions: [
                        {
                            name: action.name,
                            type: type.abiName,
                            ricardian_contract: ""
                        }
                    ]
                });
            }
        }
        return action;
    }
    /** Return true if this Action is equal to given action. */
    equals(other) {
        const otherAction = Action_1.from(other, this.abi);
        return (this.account.equals(otherAction.account) &&
            this.name.equals(otherAction.name) &&
            arrayEquatableEquals(this.authorization, otherAction.authorization) &&
            this.data.equals(otherAction.data));
    }
    decodeData(typeOrAbi) {
        if (typeof typeOrAbi === "string" ||
            typeOrAbi.abiName) {
            return abiDecode({
                data: this.data,
                type: typeOrAbi
            });
        }
        else {
            const abi = ABI.from(typeOrAbi);
            const type = abi.getActionType(this.name);
            if (!type) {
                throw new Error(`Action ${this.name} does not exist in provided ABI`);
            }
            return abiDecode({ data: this.data, type, abi });
        }
    }
    get decoded() {
        if (!this.abi) {
            throw new Error("Missing ABI definition when decoding action data");
        }
        return {
            ...this.toJSON(),
            data: this.decodeData(this.abi)
        };
    }
};
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name
    /** The name of the action. */
    )
], Action.prototype, "account", void 0);
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name
    /** The permissions authorizing the action. */
    )
], Action.prototype, "name", void 0);
__decorate([
    Struct.field(PermissionLevel, { array: true }),
    __metadata("design:type", Array)
], Action.prototype, "authorization", void 0);
__decorate([
    Struct.field("bytes"),
    __metadata("design:type", Bytes)
], Action.prototype, "data", void 0);
Action = Action_1 = __decorate([
    Struct.type("action")
], Action);
export { Action };
