// @wireio/sdk-core@1.0.18/lib/chain/Asset.js
import { isInstanceOf } from "../Utils";
import { Int64, UInt64 } from "./Integer";
import { Name } from "./Name";
export class Asset {
    static abiName = "asset";
    units;
    symbol;
    static from(value, symbol) {
        if (isInstanceOf(value, Asset)) {
            return value;
        }
        switch (typeof value) {
            case "number":
                if (!symbol) {
                    throw new Error("Symbol is required when creating Asset from number");
                }
                return this.fromFloat(value, symbol);
            case "string":
                return this.fromString(value);
            default:
                throw new Error("Invalid asset");
        }
    }
    static fromString(value) {
        const parts = (typeof value === "string" ? value : "").split(" ");
        if (parts.length !== 2) {
            throw new Error("Invalid asset string");
        }
        const amount = parts[0].replace(".", "");
        const precision = (parts[0].split(".")[1] || "").length;
        const symbol = Asset.Symbol.fromParts(parts[1], precision);
        return new Asset(Int64.from(amount), symbol);
    }
    static fromFloat(value, symbol) {
        const s = Asset.Symbol.from(symbol);
        return new Asset(s.convertFloat(value), s);
    }
    static fromUnits(value, symbol) {
        return new Asset(Int64.from(value), Asset.Symbol.from(symbol));
    }
    static fromABI(decoder) {
        const units = Int64.fromABI(decoder);
        const symbol = Asset.Symbol.fromABI(decoder);
        return new Asset(units, symbol);
    }
    static abiDefault() {
        return new this(Int64.from(0), Asset.Symbol.abiDefault());
    }
    static formatUnits(units, precision) {
        const digits = Int64.from(units).toString().split("");
        let negative = false;
        if (digits[0] === "-") {
            negative = true;
            digits.shift();
        }
        while (digits.length <= precision) {
            digits.unshift("0");
        }
        if (precision > 0) {
            digits.splice(digits.length - precision, 0, ".");
        }
        let rv = digits.join("");
        if (negative) {
            rv = "-" + rv;
        }
        return rv;
    }
    constructor(units, symbol) {
        this.units = units;
        this.symbol = symbol;
    }
    equals(other) {
        const { symbol, units } = Asset.from(other);
        return this.symbol.value.equals(symbol.value) && this.units.equals(units);
    }
    get value() {
        return this.symbol.convertUnits(this.units);
    }
    set value(newValue) {
        this.units = this.symbol.convertFloat(newValue);
    }
    get quantity() {
        return Asset.formatUnits(this.units, this.symbol.precision);
    }
    toABI(encoder) {
        this.units.toABI(encoder);
        this.symbol.toABI(encoder);
    }
    toString() {
        return this.quantity + " " + this.symbol.name;
    }
    toJSON() {
        return this.toString();
    }
}
(function (Asset) {
    class Symbol {
        static abiName = "symbol";
        static maxPrecision = 18;
        static from(value) {
            if (isInstanceOf(value, Symbol)) {
                return value;
            }
            if (isInstanceOf(value, UInt64)) {
                return new Symbol(value);
            }
            const parts = value.split(",");
            if (parts.length !== 2 && value !== "0,") {
                throw new Error("Invalid symbol string");
            }
            if (value === "0,") {
                parts.push("");
            }
            const precision = Number.parseInt(parts[0]);
            return Symbol.fromParts(parts[1], precision);
        }
        static fromParts(name, precision) {
            return new Symbol(toRawSymbol(name, precision));
        }
        // eslint-disable-next-line @typescript-eslint/ban-types
        static fromABI(decoder) {
            return new Symbol(UInt64.fromABI(decoder));
        }
        static abiDefault() {
            return this.from("4,SYS"); // CORE_SYMBOL = 4,CORE_SYMBOL_NAME
        }
        value;
        constructor(value) {
            if (toSymbolPrecision(value) > Symbol.maxPrecision) {
                throw new Error("Invalid asset symbol, precision too large");
            }
            if (!value.equals(0) && !SymbolCode.pattern.test(toSymbolName(value))) {
                throw new Error("Invalid asset symbol, name must be uppercase A-Z");
            }
            this.value = value;
        }
        equals(other) {
            return this.value.equals(Symbol.from(other).value);
        }
        get name() {
            return toSymbolName(this.value);
        }
        get precision() {
            return toSymbolPrecision(this.value);
        }
        get code() {
            return new SymbolCode(UInt64.from(this.value.value.clone().iushrn(8)));
        }
        toABI(encoder) {
            this.value.toABI(encoder);
        }
        /**
         * Convert units to floating point number according to symbol precision.
         * @throws If the given units can't be represented in 53 bits.
         **/
        convertUnits(units) {
            return units.value.toNumber() / Math.pow(10, this.precision);
        }
        /**
         * Convert floating point to units according to symbol precision.
         * Note that the value will be rounded to closest precision.
         **/
        convertFloat(float) {
            return Int64.from(float.toFixed(this.precision).replace(".", ""));
        }
        toString() {
            return `${this.precision},${this.name}`;
        }
        toJSON() {
            return this.toString();
        }
    }
    Asset.Symbol = Symbol;
    class SymbolCode {
        static abiName = "symbol_code";
        static pattern = /^[A-Z]{0,7}$/;
        static from(value) {
            if (isInstanceOf(value, SymbolCode)) {
                return value;
            }
            if (typeof value === "string") {
                value = UInt64.from(toRawSymbolCode(value));
            }
            return new this(UInt64.from(value));
        }
        static fromABI(decoder) {
            return new SymbolCode(UInt64.fromABI(decoder));
        }
        static abiDefault() {
            return this.from("SYS"); // CORE_SYMBOL_NAME
        }
        value;
        constructor(value) {
            if (!value.equals(0) && !SymbolCode.pattern.test(toSymbolName(value))) {
                throw new Error("Invalid asset symbol, name must be uppercase A-Z");
            }
            this.value = value;
        }
        equals(other) {
            return this.value.equals(SymbolCode.from(other).value);
        }
        toABI(encoder) {
            this.value.toABI(encoder);
        }
        toString() {
            return charsToSymbolName(this.value.value.toArray("be"));
        }
        toJSON() {
            return this.toString();
        }
    }
    Asset.SymbolCode = SymbolCode;
})(Asset || (Asset = {}));
export class ExtendedAsset {
    static abiName = "extended_asset";
    static from(value) {
        if (isInstanceOf(value, ExtendedAsset)) {
            return value;
        }
        return new this(Asset.from(value.quantity), Name.from(value.contract));
    }
    static fromABI(decoder) {
        return new ExtendedAsset(Asset.fromABI(decoder), Name.fromABI(decoder));
    }
    quantity;
    contract;
    constructor(quantity, contract) {
        this.quantity = quantity;
        this.contract = contract;
    }
    equals(other) {
        return (this.quantity.equals(other.quantity) &&
            this.contract.equals(other.contract));
    }
    toABI(encoder) {
        this.quantity.toABI(encoder);
        this.contract.toABI(encoder);
    }
    toJSON() {
        return {
            quantity: this.quantity,
            contract: this.contract
        };
    }
}
export class ExtendedSymbol {
    static abiName = "extended_symbol";
    static from(value) {
        if (isInstanceOf(value, ExtendedSymbol)) {
            return value;
        }
        return new this(Asset.Symbol.from(value.sym), Name.from(value.contract));
    }
    static fromABI(decoder) {
        return new ExtendedSymbol(Asset.Symbol.fromABI(decoder), Name.fromABI(decoder));
    }
    sym;
    contract;
    constructor(sym, contract) {
        this.sym = sym;
        this.contract = contract;
    }
    equals(other) {
        return this.sym.equals(other.sym) && this.contract.equals(other.contract);
    }
    toABI(encoder) {
        this.sym.toABI(encoder);
        this.contract.toABI(encoder);
    }
    toJSON() {
        return {
            sym: this.sym,
            contract: this.contract
        };
    }
}
function toSymbolPrecision(rawSymbol) {
    return rawSymbol.value.and(UInt64.from(0xff).value).toNumber();
}
function toSymbolName(rawSymbol) {
    const chars = rawSymbol.value.toArray("be").slice(0, -1);
    return charsToSymbolName(chars);
}
function charsToSymbolName(chars) {
    return chars
        .map(char => String.fromCharCode(char))
        .reverse()
        .join("");
}
function toRawSymbol(name, precision) {
    const code = toRawSymbolCode(name);
    const bytes = new Uint8Array(code.length + 1);
    bytes[0] = precision;
    bytes.set(code, 1);
    return UInt64.from(bytes);
}
function toRawSymbolCode(name) {
    const length = Math.min(name.length, 7);
    const bytes = new Uint8Array(length);
    for (let i = 0; i < length; i++) {
        bytes[i] = name.charCodeAt(i);
    }
    return bytes;
}
