// @wireio/sdk-core@1.0.18/lib/chain/Authority.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var Authority_1;
import { UInt16, UInt32 } from "./Integer";
import { PermissionLevel } from "./PermissionLevel";
import { PublicKey } from "./PublicKey";
import { Struct } from "./Struct";
import { TypeAlias } from "./TypeAlias";
import { isInstanceOf } from "../Utils";
let Weight = class Weight extends UInt16 {
};
Weight = __decorate([
    TypeAlias("weight_type")
], Weight);
export { Weight };
let KeyWeight = class KeyWeight extends Struct {
};
__decorate([
    Struct.field(PublicKey),
    __metadata("design:type", PublicKey)
], KeyWeight.prototype, "key", void 0);
__decorate([
    Struct.field(Weight),
    __metadata("design:type", Weight)
], KeyWeight.prototype, "weight", void 0);
KeyWeight = __decorate([
    Struct.type("key_weight")
], KeyWeight);
export { KeyWeight };
let PermissionLevelWeight = class PermissionLevelWeight extends Struct {
};
__decorate([
    Struct.field(PermissionLevel),
    __metadata("design:type", PermissionLevel)
], PermissionLevelWeight.prototype, "permission", void 0);
__decorate([
    Struct.field(Weight),
    __metadata("design:type", Weight)
], PermissionLevelWeight.prototype, "weight", void 0);
PermissionLevelWeight = __decorate([
    Struct.type("permission_level_weight")
], PermissionLevelWeight);
export { PermissionLevelWeight };
let Authority = Authority_1 = class Authority extends Struct {
    static from(value) {
        if (isInstanceOf(value, Authority_1)) {
            return value;
        }
        const rv = super.from({
            keys: [],
            accounts: [],
            ...value
        });
        rv.sort();
        return rv;
    }
    /** Weight a key needs to sign for this authority. */
    get keyThreshold() {
        return this.threshold.toNumber();
    }
    /** Return the weight for given public key, or zero if it is not included in this authority. */
    keyWeight(publicKey) {
        const weight = this.keys.find(({ key }) => key.equals(publicKey));
        return weight ? weight.weight.toNumber() : 0;
    }
    /**
     * Check if given public key has permission in this authority,
     * @attention Does not take indirect permissions for the key via account weights into account.
     * @param publicKey The key to check.
     * @param includePartial Whether to consider auths where the key is included but can't be reached alone (e.g. multisig).
     */
    hasPermission(publicKey, includePartial = false) {
        const threshold = includePartial ? 1 : this.keyThreshold;
        const weight = this.keyWeight(publicKey);
        return weight >= threshold;
    }
    /**
     * Sorts the authority weights in place, should be called before including the authority in a `updateauth` action or it might be rejected.
     */
    sort() {
        this.keys.sort((a, b) => String(a.key).localeCompare(String(b.key)));
        this.accounts.sort((a, b) => String(a.permission).localeCompare(String(b.permission)));
    }
};
__decorate([
    Struct.field(UInt32),
    __metadata("design:type", UInt32)
], Authority.prototype, "threshold", void 0);
__decorate([
    Struct.field(KeyWeight, { array: true }),
    __metadata("design:type", Array)
], Authority.prototype, "keys", void 0);
__decorate([
    Struct.field(PermissionLevelWeight, { array: true }),
    __metadata("design:type", Array)
], Authority.prototype, "accounts", void 0);
Authority = Authority_1 = __decorate([
    Struct.type("authority")
], Authority);
export { Authority };
