// @wireio/sdk-core@1.0.18/lib/chain/Blob.js
import { arrayEquals, isInstanceOf } from "../Utils";
export class Blob {
    static abiName = "blob";
    /**
     * Create a new Blob instance.
     */
    static from(value) {
        if (isInstanceOf(value, this)) {
            return value;
        }
        if (typeof value === "string") {
            return this.fromString(value);
        }
        throw new Error("Invalid blob");
    }
    static fromString(value) {
        // fix up base64 padding from nodeop
        switch (value.length % 4) {
            case 2:
                value += "==";
                break;
            case 3:
                value += "=";
                break;
            case 1:
                value = value.substring(0, value.length - 1);
                break;
        }
        const string = atob(value);
        const array = new Uint8Array(string.length);
        for (let i = 0; i < string.length; i++) {
            array[i] = string.charCodeAt(i);
        }
        return new this(array);
    }
    array;
    constructor(array) {
        this.array = array;
    }
    equals(other) {
        const self = this.constructor;
        try {
            return arrayEquals(this.array, self.from(other).array);
        }
        catch {
            return false;
        }
    }
    get base64String() {
        let binary = "";
        for (let i = 0; i < this.array.length; i++) {
            binary += String.fromCharCode(this.array[i]);
        }
        return btoa(binary);
    }
    /** UTF-8 string representation of this instance. */
    get utf8String() {
        return new TextDecoder().decode(this.array);
    }
    toABI(encoder) {
        encoder.writeArray(this.array);
    }
    toString() {
        return this.base64String;
    }
    toJSON() {
        return this.toString();
    }
}
