// @wireio/sdk-core@1.0.18/lib/chain/Checksum.js
import { ripemd160, sha256, sha512 } from "hash.js";
import { arrayEquals, arrayToHex, hexToArray, isInstanceOf } from "../Utils";
import { Bytes } from "./Bytes";
class Checksum {
    static abiName = "__checksum";
    static byteSize;
    static isChecksumType(value) {
        return value?.array && (Array.isArray(value.array) || typeof value.array === "object");
    }
    static from(value) {
        if (isInstanceOf(value, this)) {
            return value;
        }
        if (isInstanceOf(value, Checksum) || Checksum.isChecksumType(value)) {
            return new this(value.array);
        }
        return new this(Bytes.from(value).array);
    }
    static fromHexString(value) {
        if (value.length !== this.byteSize * 2) {
            throw new Error(`Invalid checksum hex string, expected ${this.byteSize * 2} characters got ${value.length}`);
        }
        return new this(hexToArray(value));
    }
    static fromABI(decoder) {
        return new this(decoder.readArray(this.byteSize));
    }
    static abiDefault() {
        return new this(new Uint8Array(this.byteSize));
    }
    array;
    constructor(array) {
        const byteSize = this.constructor.byteSize;
        if (array.byteLength !== byteSize) {
            throw new Error(`Checksum size mismatch, expected ${byteSize} bytes got ${array.byteLength}`);
        }
        this.array = array;
    }
    equals(other) {
        const self = this.constructor;
        try {
            return arrayEquals(this.array, self.from(other).array);
        }
        catch {
            return false;
        }
    }
    get hexString() {
        return arrayToHex(this.array);
    }
    toABI(encoder) {
        encoder.writeArray(this.array);
    }
    toString() {
        return this.hexString;
    }
    toJSON() {
        return this.toString();
    }
}
export class Checksum256 extends Checksum {
    static abiName = "checksum256";
    static byteSize = 32;
    static from(value) {
        return super.from(value);
    }
    static hash(data) {
        const digest = new Uint8Array(sha256().update(Bytes.from(data).array).digest());
        return new Checksum256(digest);
    }
}
export class Checksum512 extends Checksum {
    static abiName = "checksum512";
    static byteSize = 64;
    static from(value) {
        return super.from(value);
    }
    static hash(data) {
        const digest = new Uint8Array(sha512().update(Bytes.from(data).array).digest());
        return new Checksum512(digest);
    }
}
export class Checksum160 extends Checksum {
    static abiName = "checksum160";
    static byteSize = 20;
    static from(value) {
        return super.from(value);
    }
    static hash(data) {
        const digest = new Uint8Array(ripemd160().update(Bytes.from(data).array).digest());
        return new Checksum160(digest);
    }
}
