// @wireio/sdk-core@1.0.18/lib/chain/Float.js
import { ABIDecoder } from "../serializer/Decoder";
import { isInstanceOf, secureRandom } from "../Utils";
import { Bytes } from "./Bytes";
class Float {
    static abiName = "__float";
    static byteWidth;
    static from(value) {
        if (isInstanceOf(value, this)) {
            return value;
        }
        if (typeof value === "string") {
            value = Number.parseFloat(value);
        }
        else if (isInstanceOf(value, Float)) {
            value = value.value;
        }
        return new this(value);
    }
    static fromABI(decoder) {
        return new this(decoder.readFloat(this.byteWidth));
    }
    static abiDefault() {
        return this.from(0);
    }
    static random() {
        const bytes = secureRandom(this.byteWidth);
        const decoder = new ABIDecoder(bytes);
        return this.fromABI(decoder);
    }
    value;
    constructor(value) {
        this.value = value;
    }
    equals(other) {
        const self = this.constructor;
        return this.value === self.from(other).value;
    }
    toABI(encoder) {
        const self = this.constructor;
        encoder.writeFloat(this.value, self.byteWidth);
    }
    toString() {
        return this.value.toString();
    }
    toJSON() {
        return this.toString();
    }
}
export class Float32 extends Float {
    static abiName = "float32";
    static byteWidth = 4;
    toString() {
        return this.value.toFixed(7);
    }
}
export class Float64 extends Float {
    static abiName = "float64";
    static byteWidth = 8;
}
export class Float128 {
    static abiName = "float128";
    static byteWidth = 16;
    static from(value) {
        if (isInstanceOf(value, this)) {
            return value;
        }
        if (typeof value === "string" && value.startsWith("0x")) {
            value = value.slice(2);
        }
        return new this(Bytes.from(value));
    }
    static fromABI(decoder) {
        return new this(new Bytes(decoder.readArray(this.byteWidth)));
    }
    static random() {
        const bytes = secureRandom(16);
        const decoder = new ABIDecoder(bytes);
        return this.fromABI(decoder);
    }
    data;
    constructor(data) {
        if (data.array.length !== 16) {
            throw new Error("Invalid float128");
        }
        this.data = data;
    }
    equals(other) {
        const self = this.constructor;
        return this.data.equals(self.from(other).data);
    }
    toABI(encoder) {
        encoder.writeArray(this.data.array);
    }
    toString() {
        // float128 uses 0x prefixed hex strings as opposed to everywhere else in where there is no prefix ¯\_(ツ)_/¯
        return "0x" + this.data.hexString;
    }
    toJSON() {
        return this.toString();
    }
}
