// @wireio/sdk-core@1.0.18/lib/chain/Integer.js
import BN from "bn.js";
import { ethers } from "ethers";
import { isInstanceOf, secureRandom } from "../Utils";
/**
 * Binary integer with the underlying value represented by a BN.js instance.
 * Follows C++11 standard for arithmetic operators and conversions.
 * @note This type is optimized for correctness not speed, if you plan to manipulate
 *       integers in a tight loop you're advised to use the underlying BN.js value or
 *       convert to a JavaScript number first.
 */
export class Int {
    static abiName = "__int";
    static isSigned;
    static byteWidth;
    /** Largest value that can be represented by this integer type. */
    static get max() {
        return new BN(2)
            .pow(new BN(this.byteWidth * 8 - (this.isSigned ? 1 : 0)))
            .isubn(1);
    }
    /** Smallest value that can be represented by this integer type. */
    static get min() {
        return this.isSigned ? this.max.ineg().isubn(1) : new BN(0);
    }
    /** Add `lhs` to `rhs` and return the resulting value. */
    static add(lhs, rhs, overflow = "truncate") {
        return Int.operator(lhs, rhs, overflow, (a, b) => a.add(b));
    }
    /** Add `lhs` to `rhs` and return the resulting value. */
    static sub(lhs, rhs, overflow) {
        return Int.operator(lhs, rhs, overflow, (a, b) => a.sub(b));
    }
    /** Multiply `lhs` by `rhs` and return the resulting value. */
    static mul(lhs, rhs, overflow) {
        return Int.operator(lhs, rhs, overflow, (a, b) => a.mul(b));
    }
    /**
     * Divide `lhs` by `rhs` and return the quotient, dropping the remainder.
     * @throws When dividing by zero.
     */
    static div(lhs, rhs, overflow) {
        return Int.operator(lhs, rhs, overflow, (a, b) => {
            if (b.isZero()) {
                throw new Error("Division by zero");
            }
            return a.div(b);
        });
    }
    /**
     * Divide `lhs` by `rhs` and return the quotient + remainder rounded to the closest integer.
     * @throws When dividing by zero.
     */
    static divRound(lhs, rhs, overflow) {
        return Int.operator(lhs, rhs, overflow, (a, b) => {
            if (b.isZero()) {
                throw new Error("Division by zero");
            }
            return a.divRound(b);
        });
    }
    /**
     * Divide `lhs` by `rhs` and return the quotient + remainder rounded up to the closest integer.
     * @throws When dividing by zero.
     */
    static divCeil(lhs, rhs, overflow) {
        return Int.operator(lhs, rhs, overflow, (a, b) => {
            if (b.isZero()) {
                throw new Error("Division by zero");
            }
            const dm = a.divmod(b);
            if (dm.mod.isZero())
                return dm.div;
            return dm.div.negative !== 0 ? dm.div.isubn(1) : dm.div.iaddn(1);
        });
    }
    /** Compare `lhs` to `rhs` and return true if `lhs` is greater than `rhs`. */
    static gt(lhs, rhs) {
        return lhs.value.gt(rhs.value);
    }
    /** Compare `lhs` to `rhs` and return true if `lhs` is less than `rhs`. */
    static lt(lhs, rhs) {
        return lhs.value.lt(rhs.value);
    }
    /** Compare `lhs` to `rhs` and return true if `lhs` is greater than or equal to `rhs`. */
    static gte(lhs, rhs) {
        return lhs.value.gte(rhs.value);
    }
    /** Compare `lhs` to `rhs` and return true if `lhs` is less than or equal to `rhs`. */
    static lte(lhs, rhs) {
        return lhs.value.lte(rhs.value);
    }
    /**
     * Can be used to implement custom operator.
     * @internal
     */
    static operator(lhs, rhs, overflow = "truncate", fn) {
        const { a, b } = convert(lhs, rhs);
        const type = a.constructor;
        const result = fn(a.value, b.value);
        return type.from(result, overflow);
    }
    static from(value, overflow) {
        if (isInstanceOf(value, this)) {
            return value;
        }
        let fromType = this;
        let bn;
        if (isInstanceOf(value, Int)) {
            fromType = value.constructor;
            bn = value.value.clone();
        }
        else if (value instanceof Uint8Array) {
            bn = new BN(value, undefined, "le");
            if (fromType.isSigned) {
                bn = bn.fromTwos(fromType.byteWidth * 8);
            }
        }
        else {
            if ((typeof value === "string" && !/[0-9]+/.test(value)) ||
                (typeof value === "number" && !Number.isFinite(value))) {
                throw new Error("Invalid number");
            }
            bn = BN.isBN(value) ? value.clone() : new BN(value, 10);
            if (bn.isNeg() && !fromType.isSigned) {
                fromType = { byteWidth: fromType.byteWidth, isSigned: true };
            }
        }
        switch (overflow) {
            case "clamp":
                bn = clamp(bn, this.min, this.max);
                break;
            case "truncate":
                bn = truncate(bn, fromType, this);
                break;
        }
        return new this(bn);
    }
    static fromABI(decoder) {
        return this.from(decoder.readArray(this.byteWidth));
    }
    static abiDefault() {
        return this.from(0);
    }
    static random() {
        return this.from(secureRandom(this.byteWidth));
    }
    /**
     * The underlying BN.js instance – don't modify this
     * directly – take a copy first using `.clone()`.
     */
    value;
    /**
     * Create a new instance, don't use this directly. Use the `.from` factory method instead.
     * @throws If the value over- or under-flows the integer type.
     */
    constructor(value) {
        const self = this.constructor;
        if (self.isSigned === undefined || self.byteWidth === undefined) {
            throw new Error("Cannot instantiate abstract class Int");
        }
        if (value.gt(self.max)) {
            throw new Error(`Number ${value} overflows ${self.abiName}`);
        }
        if (value.lt(self.min)) {
            throw new Error(`Number ${value} underflows ${self.abiName}`);
        }
        this.value = value;
    }
    cast(type, overflow = "truncate") {
        if (this.constructor === type) {
            return this;
        }
        return type.from(this, overflow);
    }
    /** Number as bytes in little endian (matches memory layout in C++ contract). */
    get byteArray() {
        const self = this.constructor;
        const value = self.isSigned
            ? this.value.toTwos(self.byteWidth * 8)
            : this.value;
        return value.toArrayLike(Uint8Array, "le", self.byteWidth);
    }
    /**
     * Compare two integers, if strict is set to true the test will only consider integers
     * of the exact same type. I.e. Int64.from(1).equals(UInt64.from(1)) will return false.
     */
    equals(other, strict = false) {
        const self = this.constructor;
        if (strict === true && isInstanceOf(other, Int)) {
            const otherType = other.constructor;
            if (self.byteWidth !== otherType.byteWidth ||
                self.isSigned !== otherType.isSigned) {
                return false;
            }
        }
        try {
            return this.value.eq(self.from(other).value);
        }
        catch {
            return false;
        }
    }
    /** Mutating add. */
    add(num) {
        this.value = this.operator(num, Int.add).value;
    }
    /** Non-mutating add. */
    adding(num) {
        return this.operator(num, Int.add);
    }
    /** Mutating subtract. */
    subtract(num) {
        this.value = this.operator(num, Int.sub).value;
    }
    /** Non-mutating subtract. */
    subtracting(num) {
        return this.operator(num, Int.sub);
    }
    /** Mutating multiply. */
    multiply(by) {
        this.value = this.operator(by, Int.mul).value;
    }
    /** Non-mutating multiply. */
    multiplying(by) {
        return this.operator(by, Int.mul);
    }
    /**
     * Mutating divide.
     * @param behavior How to handle the remainder, default is to floor (round down).
     * @throws When dividing by zero.
     */
    divide(by, behavior) {
        this.value = this.dividing(by, behavior).value;
    }
    /**
     * Non-mutating divide.
     * @param behavior How to handle the remainder, default is to floor (round down).
     * @throws When dividing by zero.
     */
    dividing(by, behavior) {
        let op = Int.div;
        switch (behavior) {
            case "ceil":
                op = Int.divCeil;
                break;
            case "round":
                op = Int.divRound;
                break;
        }
        return this.operator(by, op);
    }
    /** Greater than comparision operator */
    gt(other) {
        return Int.gt(this, other);
    }
    /** Less than comparision operator */
    lt(other) {
        return Int.lt(this, other);
    }
    /** Greater than or equal comparision operator */
    gte(other) {
        return Int.gte(this, other);
    }
    /** Less than or equal comparision operator */
    lte(other) {
        return Int.lte(this, other);
    }
    /**
     * Run operator with C++11 implicit conversion.
     * @internal
     */
    operator(other, fn) {
        let rhs;
        if (isInstanceOf(other, Int)) {
            rhs = other;
        }
        else {
            rhs = Int64.from(other, "truncate");
        }
        return fn(this, rhs).cast(this.constructor);
    }
    /**
     * Convert to a JavaScript number.
     * @throws If the number cannot be represented by 53-bits.
     **/
    toNumber() {
        return this.value.toNumber();
    }
    toString() {
        return this.value.toString();
    }
    [Symbol.toPrimitive](type) {
        if (type === "number") {
            return this.toNumber();
        }
        else {
            return this.toString();
        }
    }
    toABI(encoder) {
        encoder.writeArray(this.byteArray);
    }
    toJSON() {
        // match FCs behavior and return strings for anything above 32-bit
        if (this.value.bitLength() > 32) {
            return this.value.toString();
        }
        else {
            return this.value.toNumber();
        }
    }
}
export class Int8 extends Int {
    static abiName = "int8";
    static byteWidth = 1;
    static isSigned = true;
}
export class Int16 extends Int {
    static abiName = "int16";
    static byteWidth = 2;
    static isSigned = true;
}
export class Int32 extends Int {
    static abiName = "int32";
    static byteWidth = 4;
    static isSigned = true;
}
export class Int64 extends Int {
    static abiName = "int64";
    static byteWidth = 8;
    static isSigned = true;
}
export class Int128 extends Int {
    static abiName = "int128";
    static byteWidth = 16;
    static isSigned = true;
}
export class UInt8 extends Int {
    static abiName = "uint8";
    static byteWidth = 1;
    static isSigned = false;
}
export class UInt16 extends Int {
    static abiName = "uint16";
    static byteWidth = 2;
    static isSigned = false;
}
export class UInt32 extends Int {
    static abiName = "uint32";
    static byteWidth = 4;
    static isSigned = false;
}
export class UInt64 extends Int {
    static abiName = "uint64";
    static byteWidth = 8;
    static isSigned = false;
}
export class UInt128 extends Int {
    static abiName = "uint128";
    static byteWidth = 16;
    static isSigned = false;
}
// Struct to match the 256-bit integer in Wire C++ Core contract
export class UInt256 {
    static abiName = "uint256";
    static byteWidth = 32;
    static isSigned = false;
    static DECIMALS = 18;
    static SCALE = new BN(10).pow(new BN(UInt256.DECIMALS));
    static MAX_UINT256 = new BN(1).shln(256).isubn(1); // 2^256 - 1
    low;
    high;
    constructor(low, high) {
        this.low = low;
        this.high = high;
    }
    /**
     * Create a UInt256 from:
     *  • number|string    → interpreted as an 18‐decimal fixed‐point (scaled by 10^18).
     *  • ethers.BigNumber → interpreted as a “whole integer,” multiply by 10^18.
     *  • BN               → interpreted as a “whole integer,” multiply by 10^18.
     *  • UInt128           → interpreted as a “whole integer,” multiply by 10^18.
     *  • UInt256           → just return it (clone).
     *  • { low, high }     → raw halves (H<<128 + L) as a “whole integer,” multiply by 10^18.
     */
    static from(value) {
        // 1) If it's already a UInt256, return it.
        if (value instanceof UInt256) {
            return value;
        }
        // 2) If it's a UInt128: get raw BN128, multiply by SCALE, then split.
        if (value instanceof UInt128) {
            const as128 = UInt256.u128ToBN(value);
            const scaled = as128.mul(this.SCALE);
            return UInt256.fromRaw(scaled);
        }
        // 3) If it's an ethers.BigNumber: get raw BN, multiply by SCALE, then split.
        if (ethers.BigNumber.isBigNumber(value)) {
            const bnWhole = new BN(value._hex.slice(2), 16);
            const scaled = bnWhole.mul(this.SCALE);
            return UInt256.fromRaw(scaled);
        }
        // 4) If it's a BN: treat as raw BN, multiply by SCALE, then split.
        if (BN.isBN(value)) {
            const scaled = value.mul(this.SCALE);
            return UInt256.fromRaw(scaled);
        }
        // 5) If it's an object { low, high }, reconstruct rawInteger = (high<<128) + low,
        //    then multiply by SCALE, then split.
        if (typeof value === "object" &&
            value !== null &&
            "low" in value &&
            "high" in value) {
            const parts = value;
            const lowPart = UInt128.from(parts.low);
            const highPart = UInt128.from(parts.high);
            return new UInt256(lowPart, highPart);
            // const lowBN = UInt256.u128ToBN(lowPart);
            // const highBN = UInt256.u128ToBN(highPart);
            // const rawInteger = highBN.shln(128).add(lowBN);
            // const scaled = rawInteger.mul(this.SCALE);
            // return UInt256.fromRaw(scaled);
        }
        // 6) Otherwise (number|string): parse as 18‐decimal fixed‐point
        //    (e.g. "1.5" → BN(1.5 × 10^18)).
        const valueStr = value.toString();
        const [whole, frac = ""] = valueStr.split(".");
        const wholeBN = new BN(whole || "0", 10).mul(this.SCALE);
        const fracStr = frac.padEnd(this.DECIMALS, "0").slice(0, this.DECIMALS);
        const fracBN = new BN(fracStr, 10);
        const scaled = wholeBN.add(fracBN);
        if (scaled.gt(this.MAX_UINT256)) {
            throw new Error(`Value ${value} exceeds 256 bits once scaled`);
        }
        const mask128 = new BN("ffffffffffffffffffffffffffffffff", 16);
        const lowBN = scaled.and(mask128);
        const highBN = scaled.shrn(128);
        return new UInt256(UInt128.from(lowBN), UInt128.from(highBN));
    }
    /**
     * Construct a UInt256 from a raw 256-bit BN (no scaling).
     * Internal helper for add/sub/mul/div.
     */
    static fromRaw(raw) {
        if (raw.isNeg()) {
            throw new Error("Cannot represent negative values in UInt256");
        }
        if (raw.gt(this.MAX_UINT256)) {
            throw new Error("Uint256 overflow (larger than 2^256 - 1)");
        }
        const mask128 = new BN("ffffffffffffffffffffffffffffffff", 16);
        const lowBN = raw.and(mask128);
        const highBN = raw.shrn(128);
        return new UInt256(UInt128.from(lowBN), UInt128.from(highBN));
    }
    /**
     * Helper to combine `low` + `high` into a single BN that includes the
     * _already-scaled_ 10^18 factor.
     */
    raw() {
        const lowBN = UInt256.u128ToBN(this.low);
        const highBN = UInt256.u128ToBN(this.high).shln(128);
        return highBN.add(lowBN);
    }
    /**
     * Convert the UInt256 to a human-readable string,
     * e.g. "123.456" for internal BN "123456000000000000000".
     */
    toString() {
        const scaled = this.raw();
        const intPart = scaled.div(UInt256.SCALE);
        const fracPart = scaled.mod(UInt256.SCALE);
        if (fracPart.isZero()) {
            // No fractional digits
            return intPart.toString(10);
        }
        else {
            // We have a fractional component
            const fracStr = fracPart
                .toString(10)
                .padStart(UInt256.DECIMALS, "0")
                .replace(/0+$/, ""); // remove trailing zeros
            return `${intPart}.${fracStr}`;
        }
    }
    /**
     * Convert the UInt256 to a JS number if safe; otherwise returns a BN.
     * This "descale" by 10^18 first, so "123.456" comes back as ~123.456 in JS.
     */
    toNumber() {
        const scaled = this.raw(); // The big BN, e.g. 123.456 => 123456000000000000
        const integer = scaled.div(UInt256.SCALE);
        const remainder = scaled.mod(UInt256.SCALE);
        // 1) If the integer part alone exceeds 2^53, return BN (or throw)
        if (integer.bitLength() > 53) {
            return new BN(this.toString());
        }
        // 2) Convert integer part safely
        const intNum = integer.toNumber(); // Guaranteed safe
        // 3) If no remainder, we have a whole number (like 123.000...)
        if (remainder.isZero()) {
            return intNum;
        }
        // 4) We have a fractional part. Instead of remainder.toNumber(),
        //    convert remainder to decimal string, pad left to 18 digits,
        //    then parse as float in '0.xxxxx' form.
        const remainderStr = remainder.toString(10).padStart(UInt256.DECIMALS, "0");
        // e.g. "456000000000000000" => parseFloat("0.456000000000000000") => ~0.456
        // parseFloat of an 18-digit fraction is near the limit of JS float precision,
        // but it won't throw an error. You will get a float ~0.456
        const fracNum = parseFloat("0." + remainderStr);
        // 5) Combine integer + fraction. If that sum is still <= 2^53,
        //    we return it; otherwise, return BN.
        const result = intNum + fracNum;
        if (!Number.isFinite(result) || result > Number.MAX_SAFE_INTEGER) {
            return new BN(this.toString());
        }
        return result;
    }
    /**
     * Add another UInt256 (both 18-decimal scaled).
     */
    add(other) {
        const sum = this.raw().add(other.raw());
        return UInt256.fromRaw(sum);
    }
    /**
     * Subtract another UInt256. Throws if result < 0 (underflow).
     */
    subtract(other) {
        const diff = this.raw().sub(other.raw());
        if (diff.isNeg()) {
            throw new Error("Underflow in subtract");
        }
        return UInt256.fromRaw(diff);
    }
    /**
     * Multiply this UInt256 by another, then scale back down by 10^18
     * so final is still 18-decimals.
     *
     * So effectively: (a * b) / 10^18
     */
    multiply(other) {
        const product = this.raw().mul(other.raw()).div(UInt256.SCALE);
        return UInt256.fromRaw(product);
    }
    /**
     * Divide this UInt256 by another, scaling up the dividend by 10^18
     * first so final is still 18 decimals.
     *
     * So effectively: (a * 10^18) / b
     */
    divide(divisor) {
        const b = divisor.raw();
        if (b.isZero()) {
            throw new Error("Division by zero");
        }
        const numerator = this.raw().mul(UInt256.SCALE);
        const quotient = numerator.div(b);
        return UInt256.fromRaw(quotient);
    }
    /**
     * Modulo. Because both sides are scaled, we just do raw mod.
     * The result is still scaled with 18 decimals.
     */
    modulo(divisor) {
        const b = divisor.raw();
        if (b.isZero()) {
            throw new Error("Division by zero in modulo");
        }
        const remainder = this.raw().mod(b);
        return UInt256.fromRaw(remainder);
    }
    /**
     * Compare: -1 if this < other, 0 if equal, +1 if this > other.
     */
    compare(other) {
        const aRaw = this.raw();
        const bRaw = other.raw();
        return aRaw.cmp(bRaw);
    }
    equals(other) {
        return this.compare(other) === 0;
    }
    greaterThan(other) {
        return this.compare(other) > 0;
    }
    lessThan(other) {
        return this.compare(other) < 0;
    }
    static u128ToBN(u128) {
        // Access the raw bytes in LE
        const bytes = u128.byteArray;
        return new BN(bytes, "le");
    }
}
export class VarInt extends Int {
    static abiName = "varint32";
    static byteWidth = 32;
    static isSigned = true;
    static fromABI(decoder) {
        return new this(new BN(decoder.readVarint32()));
    }
    toABI(encoder) {
        encoder.writeVarint32(Number(this));
    }
}
export class VarUInt extends Int {
    static abiName = "varuint32";
    static byteWidth = 32;
    static isSigned = false;
    static fromABI(decoder) {
        return new this(new BN(decoder.readVaruint32()));
    }
    toABI(encoder) {
        encoder.writeVaruint32(Number(this));
    }
}
/** Clamp number between min and max. */
function clamp(num, min, max) {
    return BN.min(BN.max(num, min), max);
}
/**
 * Create new BN with the same bit pattern as the passed value,
 * extending or truncating the value’s representation as necessary.
 */
function truncate(value, from, to) {
    const fill = value.isNeg() ? 255 : 0;
    const fromValue = from.isSigned ? value.toTwos(from.byteWidth * 8) : value;
    const fromBytes = fromValue.toArrayLike(Uint8Array, "le");
    const toBytes = new Uint8Array(to.byteWidth);
    toBytes.fill(fill);
    toBytes.set(fromBytes.slice(0, to.byteWidth));
    const toValue = new BN(toBytes, undefined, "le");
    return to.isSigned ? toValue.fromTwos(to.byteWidth * 8) : toValue;
}
/** C++11 implicit integer conversions. */
function convert(a, b) {
    // The integral promotions (4.5) shall be performed on both operands.
    a = promote(a);
    b = promote(b);
    const aType = a.constructor;
    const bType = b.constructor;
    // If both operands have the same type, no further conversion is needed
    if (aType !== bType) {
        // Otherwise, if both operands have signed integer types or both have unsigned integer types,
        // the operand with the type of lesser integer conversion rank shall be converted to the type
        // of the operand with greater rank.
        if (aType.isSigned === bType.isSigned) {
            if (aType.byteWidth > bType.byteWidth) {
                b = b.cast(aType);
            }
            else if (bType.byteWidth > aType.byteWidth) {
                a = a.cast(bType);
            }
        }
        else {
            // Otherwise, if the operand that has unsigned integer type has rank greater than or equal
            // to the rank of the type of the other operand, the operand with signed integer type
            // shall be converted to the type of the operand with unsigned integer type.
            if (aType.isSigned === false && aType.byteWidth >= bType.byteWidth) {
                b = b.cast(aType);
            }
            else if (bType.isSigned === false &&
                bType.byteWidth >= aType.byteWidth) {
                a = a.cast(bType);
            }
            else {
                // Otherwise, if the type of the operand with signed integer type can represent all of the
                // values of the type of the operand with unsigned integer type, the operand with unsigned
                // integer type shall be converted to the type of the operand with signed integer type.
                if (aType.isSigned === true &&
                    aType.max.gte(bType.max) &&
                    aType.min.lte(bType.min)) {
                    b = b.cast(aType);
                }
                else if (bType.isSigned === true &&
                    bType.max.gte(aType.max) &&
                    bType.min.lte(aType.min)) {
                    a = a.cast(bType);
                }
                else {
                    // Otherwise, both operands shall be converted to the unsigned integer type
                    // corresponding to the type of the operand with signed integer type.
                    // ---
                    // Dead code: this can't happen™ with the types we have.
                    // ---
                    // const signedType = aType.isSigned ? aType : bType
                    // let unsignedType: typeof Int
                    // switch (signedType.byteWidth) {
                    //     case 4:
                    //         unsignedType = UInt32
                    //         break
                    //     case 8:
                    //         unsignedType = UInt64
                    //         break
                    //     case 16:
                    //         unsignedType = UInt128
                    //         break
                    //     default:
                    //         throw new Error(
                    //             `No corresponding unsigned type for ${signedType.abiName}`
                    //         )
                    // }
                    // a = a.cast(unsignedType)
                    // b = b.cast(unsignedType)
                }
            }
        }
    }
    return { a, b };
}
/** C++11 integral promotion. */
function promote(n) {
    // An rvalue of type char, signed char, unsigned char, short int, or
    // unsigned short int can be converted to an rvalue of type int if int
    // can represent all the values of the source type; otherwise, the source
    // rvalue can be converted to an rvalue of type unsigned int.
    let rv = n;
    const type = n.constructor;
    if (type.byteWidth < 4) {
        rv = n.cast(Int32);
    }
    return rv;
}
