// @wireio/sdk-core@1.0.18/lib/chain/KeyType.js
/** Supported Wire curve types. */
export var KeyType;
(function (KeyType) {
    KeyType["K1"] = "K1";
    KeyType["R1"] = "R1";
    KeyType["WA"] = "WA";
    // Note: WA is not a curve type but a protocol, it is included for compatibility
    KeyType["EM"] = "EM";
    KeyType["ED"] = "ED"; // Ed25519 - used for Solana and other ED chains
})(KeyType || (KeyType = {}));
(function (KeyType) {
    function indexFor(value) {
        switch (value) {
            case KeyType.K1:
                return 0;
            case KeyType.R1:
                return 1;
            case KeyType.WA:
                return 2;
            case KeyType.EM:
                return 3;
            case KeyType.ED:
                return 4;
            default:
                throw new Error(`Unknown curve type: ${value}`);
        }
    }
    KeyType.indexFor = indexFor;
    function from(value) {
        let index;
        if (typeof value !== "number") {
            index = KeyType.indexFor(value);
        }
        else {
            index = value;
        }
        switch (index) {
            case 0:
                return KeyType.K1;
            case 1:
                return KeyType.R1;
            case 2:
                return KeyType.WA;
            case 3:
                return KeyType.EM;
            case 4:
                return KeyType.ED;
            default:
                throw new Error("Unknown curve type");
        }
    }
    KeyType.from = from;
})(KeyType || (KeyType = {}));
