// @wireio/sdk-core@1.0.18/lib/chain/PublicKey.js
/* eslint-disable no-console */
// src/crypto/public_key.ts
import { Base58 } from "../Base58";
import { arrayToHex, hexToArray, isInstanceOf } from "../Utils";
import { Bytes } from "./Bytes";
import { KeyType } from "./KeyType";
export class PublicKey {
    static abiName = "public_key";
    /** Type, e.g. `K1`, `R1`, `EM`, or `ED` */
    type;
    /** Compressed public key point. */
    data;
    /** Create PublicKey object from representing types. */
    static from(value) {
        if (isInstanceOf(value, PublicKey)) {
            return value;
        }
        if (typeof value === "object" && value.type && value.compressed) {
            return new PublicKey(KeyType.from(value.type), new Bytes(value.compressed));
        }
        if (typeof value !== "string") {
            throw new Error("Invalid public key");
        }
        if (value.startsWith("PUB_")) {
            const parts = value.split("_");
            if (parts.length !== 3) {
                throw new Error("Invalid public key string");
            }
            const type = KeyType.from(parts[1]);
            // ECDSA curves use 33-byte compressed pubs; ED25519 uses 32-byte
            const size = type === KeyType.K1 || type === KeyType.R1 || type === KeyType.EM
                ? 33
                : type === KeyType.ED
                    ? 32
                    : undefined;
            let data;
            try {
                data = Base58.decodeRipemd160Check(parts[2], size, type);
            }
            catch (e) {
                try {
                    data = hexToArray(parts[2]);
                }
                catch (e2) {
                    console.error("Both base58 and hex failed to parse", e, e2);
                    throw e;
                }
            }
            // const data = hexToArray(parts[2]);
            return new PublicKey(type, data);
        }
        else if (value.length >= 50) {
            // Legacy SYS key
            const data = Base58.decodeRipemd160Check(value.slice(-50));
            return new PublicKey(KeyType.K1, data);
        }
        else {
            throw new Error("Invalid public key string");
        }
    }
    /** @internal */
    static fromABI(decoder) {
        const type = KeyType.from(decoder.readByte());
        if (type == KeyType.WA) {
            const startPos = decoder.getPosition();
            decoder.advance(33); // key_data
            decoder.advance(1); // user presence
            decoder.advance(decoder.readVaruint32()); // rpid
            const len = decoder.getPosition() - startPos;
            decoder.setPosition(startPos);
            const data = Bytes.from(decoder.readArray(len));
            return new PublicKey(KeyType.WA, data);
        }
        // ECDSA compressed keys = 33 bytes; ED25519 keys = 32 bytes
        const len = type === KeyType.ED ? 32 : 33;
        return new PublicKey(type, new Bytes(decoder.readArray(len)));
    }
    /** @internal */
    constructor(type, data) {
        this.type = type;
        this.data = data instanceof Bytes ? data : new Bytes(data);
    }
    equals(other) {
        const otherKey = PublicKey.from(other);
        return this.type === otherKey.type && this.data.equals(otherKey.data);
    }
    /**
     * Return Antelope/SYSIO legacy (`SYS<base58data>`) formatted key.
     * @throws If the key type isn't `K1` or `EM`.
     */
    toLegacyString(prefix = "SYS") {
        if (this.type !== KeyType.K1 && this.type !== KeyType.EM) {
            throw new Error("Unable to create legacy formatted string for non-K1/EM key");
        }
        return `${prefix}${Base58.encodeRipemd160Check(this.data)}`;
    }
    /** Return key in modern Antelope/SYSIO format (`PUB_<type>_<base58data>`) */
    toString() {
        // Ensure the key is compressed
        if ((this.type === KeyType.K1 ||
            this.type === KeyType.R1 ||
            this.type === KeyType.EM) &&
            this.data.array.length !== 33) {
            throw new Error(`Expected 33-byte compressed key for ${this.type}, got ${this.data.array.length}`);
        }
        if (this.type === KeyType.K1 || this.type === KeyType.R1) {
            return `PUB_${this.type}_${Base58.encodeRipemd160Check(this.data, this.type)}`;
        }
        return `PUB_${this.type}_${arrayToHex(this.data.array)}`;
    }
    toHexString() {
        // Ensure the key is compressed
        if ((this.type === KeyType.K1 ||
            this.type === KeyType.R1 ||
            this.type === KeyType.EM) &&
            this.data.array.length !== 33) {
            throw new Error(`Expected 33-byte compressed key for ${this.type}, got ${this.data.array.length}`);
        }
        return `PUB_${this.type}_${arrayToHex(this.data.array)}`;
    }
    /** @internal */
    toABI(encoder) {
        encoder.writeByte(KeyType.indexFor(this.type));
        encoder.writeArray(this.data.array);
    }
    /** @internal */
    toJSON() {
        return this.toString();
    }
}
