// @wireio/sdk-core@1.0.18/lib/chain/Transaction.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var Transaction_1;
import pako from "pako";
import { abiEncode } from "../serializer/Encoder";
import { padEdForTx, stripEdPad } from "./Signature";
import { abiDecode } from "../serializer/Decoder";
import { Action } from "./Action";
import { Bytes } from "./Bytes";
import { Checksum256 } from "./Checksum";
import { UInt16, UInt32, UInt8, VarUInt } from "./Integer";
import { KeyType } from "./KeyType";
import { Name } from "./Name";
import { PublicKey } from "./PublicKey";
import { Struct } from "./Struct";
import { TimePointSec } from "./Time";
import { ethers } from "ethers";
import { concatBytes } from "../Utils";
let TransactionExtension = class TransactionExtension extends Struct {
};
__decorate([
    Struct.field("uint16"),
    __metadata("design:type", UInt16)
], TransactionExtension.prototype, "type", void 0);
__decorate([
    Struct.field("bytes"),
    __metadata("design:type", Bytes)
], TransactionExtension.prototype, "data", void 0);
TransactionExtension = __decorate([
    Struct.type("transaction_extension")
], TransactionExtension);
export { TransactionExtension };
export var TransactionExtensionType;
(function (TransactionExtensionType) {
    TransactionExtensionType[TransactionExtensionType["PubKey"] = 32768] = "PubKey";
    // Add other extension types here as needed
})(TransactionExtensionType || (TransactionExtensionType = {}));
let TransactionHeader = class TransactionHeader extends Struct {
    static from(object) {
        return super.from({
            max_net_usage_words: 0,
            max_cpu_usage_ms: 0,
            delay_sec: 0,
            ...object
        });
    }
};
__decorate([
    Struct.field("time_point_sec"),
    __metadata("design:type", TimePointSec
    /** *Specifies a block num in the last 2^16 blocks. */
    )
], TransactionHeader.prototype, "expiration", void 0);
__decorate([
    Struct.field("uint16"),
    __metadata("design:type", UInt16
    /** Specifies the lower 32 bits of the block id. */
    )
], TransactionHeader.prototype, "ref_block_num", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32
    /** Upper limit on total network bandwidth (in 8 byte words) billed for this transaction. */
    )
], TransactionHeader.prototype, "ref_block_prefix", void 0);
__decorate([
    Struct.field("varuint32"),
    __metadata("design:type", VarUInt
    /** Upper limit on the total CPU time billed for this transaction. */
    )
], TransactionHeader.prototype, "max_net_usage_words", void 0);
__decorate([
    Struct.field("uint8"),
    __metadata("design:type", UInt8
    /** Number of seconds to delay this transaction for during which it may be canceled. */
    )
], TransactionHeader.prototype, "max_cpu_usage_ms", void 0);
__decorate([
    Struct.field("varuint32"),
    __metadata("design:type", VarUInt)
], TransactionHeader.prototype, "delay_sec", void 0);
TransactionHeader = __decorate([
    Struct.type("transaction_header")
], TransactionHeader);
export { TransactionHeader };
let Transaction = Transaction_1 = class Transaction extends TransactionHeader {
    static from(object, abis) {
        const abiFor = (contract) => {
            if (!abis) {
                return;
            }
            else if (Array.isArray(abis)) {
                return abis
                    .filter(abi => Name.from(abi.contract).equals(contract))
                    .map(({ abi }) => abi)[0];
            }
            else {
                return abis;
            }
        };
        const resolveAction = (action) => {
            if (action instanceof Action) {
                return action;
            }
            else {
                return Action.from(action, abiFor(action.account));
            }
        };
        const actions = (object.actions || []).map(resolveAction);
        const context_free_actions = (object.context_free_actions || []).map(resolveAction);
        const transaction = {
            transaction_extensions: [],
            ...object,
            context_free_actions,
            actions
        };
        return super.from(transaction);
    }
    /** Return true if this transaction is equal to given transaction. */
    equals(other) {
        const tx = Transaction_1.from(other);
        return this.id.equals(tx.id);
    }
    get id() {
        return Checksum256.hash(abiEncode({ object: this }));
    }
    /**
     * Computes the signing digest and message bytes for this transaction.
     *
     * @param chainId - The chain ID to use for the signing digest.
     * @param keyType - (Optional) The key type of the signer. If provided, the message bytes (`msgBytes`)
     *   are formatted according to the expected format for the given key type.
     *   - For `KeyType.EM`, the digest is prefixed with `0x` and arrayified.
     *   - For `KeyType.ED`, the digest's hex string is encoded as UTF-8 bytes.
     *   - If not provided, the default digest bytes are used.
     * @returns An object containing the signing digest (`msgDigest`) and the formatted message bytes (`msgBytes`).
     */
    signingDigest(chainId, keyType) {
        const data = this.signingData(chainId);
        const msgDigest = Checksum256.hash(data);
        const msgDigestHex = msgDigest.hexString.toLowerCase();
        let msgBytes = msgDigest.array;
        // Prepare custom msgBytes based on key type
        switch (keyType) {
            case KeyType.EM: // Prefix with 0x and arrayify
                msgBytes = ethers.utils.arrayify("0x" + msgDigestHex);
                break;
            case KeyType.ED: // Encode as UTF-8 bytes for Phantom signing
                msgBytes = new TextEncoder().encode(msgDigestHex);
                break;
        }
        return { msgDigest, msgBytes };
    }
    signingData(chainId) {
        let data = Bytes.from(Checksum256.from(chainId).array);
        data = data.appending(abiEncode({ object: this }));
        data = data.appending(new Uint8Array(32));
        return data;
    }
    extPubKey(pubKeys) {
        if (!Array.isArray(pubKeys))
            pubKeys = [pubKeys];
        for (const pkey of pubKeys) {
            const pubKey = PublicKey.from(pkey);
            const rawKey = pubKey.data.array;
            const keyTypeTag = KeyType.indexFor(pubKey.type);
            const tag = Uint8Array.from([keyTypeTag]);
            const ext = TransactionExtension.from({
                type: TransactionExtensionType.PubKey,
                data: concatBytes(tag, rawKey)
            });
            this.transaction_extensions.push(ext);
        }
    }
};
__decorate([
    Struct.field(Action, { array: true }),
    __metadata("design:type", Array)
], Transaction.prototype, "context_free_actions", void 0);
__decorate([
    Struct.field(Action, { array: true }),
    __metadata("design:type", Array)
], Transaction.prototype, "actions", void 0);
__decorate([
    Struct.field(TransactionExtension, { array: true }),
    __metadata("design:type", Array)
], Transaction.prototype, "transaction_extensions", void 0);
Transaction = Transaction_1 = __decorate([
    Struct.type("transaction")
], Transaction);
export { Transaction };
let SignedTransaction = class SignedTransaction extends Transaction {
    /** The transaction without the signatures. */
    get transaction() {
        return Transaction.from({
            ...this,
            signatures: undefined,
            context_free_data: undefined
        });
    }
    get id() {
        return this.transaction.id;
    }
    static from(object) {
        return super.from({
            signatures: [],
            context_free_data: [],
            ...object
        });
    }
};
__decorate([
    Struct.field("signature[]"),
    __metadata("design:type", Array)
], SignedTransaction.prototype, "signatures", void 0);
__decorate([
    Struct.field("bytes[]"),
    __metadata("design:type", Array)
], SignedTransaction.prototype, "context_free_data", void 0);
SignedTransaction = __decorate([
    Struct.type("signed_transaction")
], SignedTransaction);
export { SignedTransaction };
// reference: https://github.com/AntelopeIO/leap/blob/339d98eed107b9fd94736988996082c7002fa52a/libraries/chain/include/sysio/chain/transaction.hpp#L131-L134
export var CompressionType;
(function (CompressionType) {
    CompressionType[CompressionType["none"] = 0] = "none";
    CompressionType[CompressionType["zlib"] = 1] = "zlib";
})(CompressionType || (CompressionType = {}));
let PackedTransaction = class PackedTransaction extends Struct {
    static from(object) {
        return super.from({
            signatures: [],
            packed_context_free_data: "",
            compression: 0,
            ...object
        });
    }
    static fromSigned(signed, compression = 1) {
        // Pad ED sigs to 65 bytes, return unmodified K1/R1/EM sigs
        const wireSigs = signed.signatures.map(padEdForTx);
        // Encode data
        let packed_trx = abiEncode({ object: Transaction.from(signed) });
        let packed_context_free_data = abiEncode({
            object: signed.context_free_data,
            type: "bytes[]"
        });
        switch (compression) {
            case CompressionType.zlib: {
                // compress data
                packed_trx = pako.deflate(packed_trx.array);
                packed_context_free_data = pako.deflate(packed_context_free_data.array);
                break;
            }
            case CompressionType.none: {
                break;
            }
        }
        return this.from({
            compression,
            signatures: wireSigs,
            packed_context_free_data,
            packed_trx
        });
    }
    getTransaction() {
        switch (Number(this.compression)) {
            // none
            case CompressionType.none: {
                return abiDecode({ data: this.packed_trx, type: Transaction });
            }
            // zlib compressed
            case CompressionType.zlib: {
                const inflated = pako.inflate(this.packed_trx.array);
                return abiDecode({ data: inflated, type: Transaction });
            }
            default: {
                throw new Error(`Unknown transaction compression ${this.compression}`);
            }
        }
    }
    getSignedTransaction() {
        const transaction = this.getTransaction();
        // TODO: decode context free data
        return SignedTransaction.from({
            ...transaction,
            signatures: this.signatures.map(stripEdPad)
        });
    }
};
__decorate([
    Struct.field("signature[]"),
    __metadata("design:type", Array)
], PackedTransaction.prototype, "signatures", void 0);
__decorate([
    Struct.field("uint8"),
    __metadata("design:type", UInt8)
], PackedTransaction.prototype, "compression", void 0);
__decorate([
    Struct.field("bytes"),
    __metadata("design:type", Bytes)
], PackedTransaction.prototype, "packed_context_free_data", void 0);
__decorate([
    Struct.field("bytes"),
    __metadata("design:type", Bytes)
], PackedTransaction.prototype, "packed_trx", void 0);
PackedTransaction = __decorate([
    Struct.type("packed_transaction")
], PackedTransaction);
export { PackedTransaction };
let TransactionReceipt = class TransactionReceipt extends Struct {
};
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], TransactionReceipt.prototype, "status", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], TransactionReceipt.prototype, "cpu_usage_us", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], TransactionReceipt.prototype, "net_usage_words", void 0);
TransactionReceipt = __decorate([
    Struct.type("transaction_receipt")
], TransactionReceipt);
export { TransactionReceipt };
