// @wireio/sdk-core@1.0.18/lib/resources/Powerup.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Asset } from "../chain/Asset";
import { UInt32, UInt8 } from "../chain/Integer";
import { Struct } from "../chain/Struct";
import { PowerUpStateResourceCPU } from "./powerup/Cpu";
import { PowerUpStateResourceNET } from "./powerup/Net";
let PowerUpState = class PowerUpState extends Struct {
};
__decorate([
    Struct.field("uint8"),
    __metadata("design:type", UInt8)
], PowerUpState.prototype, "version", void 0);
__decorate([
    Struct.field(PowerUpStateResourceNET),
    __metadata("design:type", PowerUpStateResourceNET)
], PowerUpState.prototype, "net", void 0);
__decorate([
    Struct.field(PowerUpStateResourceCPU),
    __metadata("design:type", PowerUpStateResourceCPU)
], PowerUpState.prototype, "cpu", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], PowerUpState.prototype, "powerup_days", void 0);
__decorate([
    Struct.field("asset"),
    __metadata("design:type", Asset)
], PowerUpState.prototype, "min_powerup_fee", void 0);
PowerUpState = __decorate([
    Struct.type("powerupstate")
], PowerUpState);
export { PowerUpState };
export class PowerUpAPI {
    parent;
    constructor(parent) {
        this.parent = parent;
    }
    async get_state() {
        const response = await this.parent.api.v1.chain.get_table_rows({
            code: "sysio",
            scope: "",
            table: "powup.state",
            type: PowerUpState
        });
        return response.rows[0];
    }
}
