// @wireio/sdk-core@1.0.18/lib/resources/Ram.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Asset } from "../chain/Asset";
import { Float64 } from "../chain/Float";
import { Int64 } from "../chain/Integer";
import { Struct } from "../chain/Struct";
let Connector = class Connector extends Struct {
};
__decorate([
    Struct.field("asset"),
    __metadata("design:type", Asset)
], Connector.prototype, "balance", void 0);
__decorate([
    Struct.field("float64"),
    __metadata("design:type", Float64)
], Connector.prototype, "weight", void 0);
Connector = __decorate([
    Struct.type("connector")
], Connector);
export { Connector };
let ExchangeState = class ExchangeState extends Struct {
};
__decorate([
    Struct.field("asset"),
    __metadata("design:type", Asset)
], ExchangeState.prototype, "supply", void 0);
__decorate([
    Struct.field(Connector),
    __metadata("design:type", Connector)
], ExchangeState.prototype, "base", void 0);
__decorate([
    Struct.field(Connector),
    __metadata("design:type", Connector)
], ExchangeState.prototype, "quote", void 0);
ExchangeState = __decorate([
    Struct.type("exchange_state")
], ExchangeState);
export { ExchangeState };
let RAMState = class RAMState extends ExchangeState {
    price_per(bytes) {
        const base = this.base.balance.units;
        const quote = this.quote.balance.units;
        return Asset.fromUnits(this.get_input(base, quote, Int64.from(bytes)), this.quote.balance.symbol);
    }
    price_per_kb(kilobytes) {
        return this.price_per(kilobytes * 1000);
    }
    // Derived from https://github.com/SYSIO/sysio.contracts/blob/f6578c45c83ec60826e6a1eeb9ee71de85abe976/contracts/sysio.system/src/exchange_state.cpp#L96
    get_input(base, quote, value) {
        // (quote * value) / (base - value), using 'ceil' to round up
        return quote.multiplying(value).dividing(base.subtracting(value), "ceil");
    }
};
RAMState = __decorate([
    Struct.type("ramstate")
], RAMState);
export { RAMState };
export class RAMAPI {
    parent;
    constructor(parent) {
        this.parent = parent;
    }
    async get_state() {
        const response = await this.parent.api.v1.chain.get_table_rows({
            code: "sysio",
            scope: "sysio",
            table: "rammarket",
            type: RAMState
        });
        return response.rows[0];
    }
}
