// @wireio/sdk-core@1.0.18/lib/resources/Rex.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { BNPrecision } from "./IndexResources";
import { Asset } from "../chain/Asset";
import { UInt64, UInt8 } from "../chain/Integer";
import { Struct } from "../chain/Struct";
let REXState = class REXState extends Struct {
    get reserved() {
        return Number(this.total_lent.units) / Number(this.total_lendable.units);
    }
    get symbol() {
        return this.total_lent.symbol;
    }
    get precision() {
        return this.total_lent.symbol.precision;
    }
    get value() {
        return ((Number(this.total_lent.units) + Number(this.total_unlent.units)) /
            Number(this.total_rex.units));
    }
    exchange(amount) {
        return Asset.from((amount.value * this.total_lendable.value) / this.total_rex.value, this.symbol);
    }
    price_per(sample, unit = 1000) {
        // Sample token units
        const tokens = Asset.fromUnits(10000, this.symbol);
        // Spending 1 SYS (10000 units) on REX gives this many tokens
        const bancor = Number(tokens.units) / (this.total_rent.value / this.total_unlent.value);
        // The ratio of the number of tokens received vs the sampled values
        const unitPrice = bancor * (Number(sample.cpu) / BNPrecision.toNumber());
        // The token units spent per unit
        const perunit = Number(tokens.units) / unitPrice;
        // Multiply the per unit cost by the units requested
        const cost = perunit * unit;
        // Converting to an Asset
        return cost / Math.pow(10, this.precision);
    }
};
__decorate([
    Struct.field("uint8"),
    __metadata("design:type", UInt8)
], REXState.prototype, "version", void 0);
__decorate([
    Struct.field("asset"),
    __metadata("design:type", Asset)
], REXState.prototype, "total_lent", void 0);
__decorate([
    Struct.field("asset"),
    __metadata("design:type", Asset)
], REXState.prototype, "total_unlent", void 0);
__decorate([
    Struct.field("asset"),
    __metadata("design:type", Asset)
], REXState.prototype, "total_rent", void 0);
__decorate([
    Struct.field("asset"),
    __metadata("design:type", Asset)
], REXState.prototype, "total_lendable", void 0);
__decorate([
    Struct.field("asset"),
    __metadata("design:type", Asset)
], REXState.prototype, "total_rex", void 0);
__decorate([
    Struct.field("asset"),
    __metadata("design:type", Asset)
], REXState.prototype, "namebid_proceeds", void 0);
__decorate([
    Struct.field("uint64"),
    __metadata("design:type", UInt64)
], REXState.prototype, "loan_num", void 0);
REXState = __decorate([
    Struct.type("rexstate")
], REXState);
export { REXState };
export class REXAPI {
    parent;
    constructor(parent) {
        this.parent = parent;
    }
    async get_state() {
        const response = await this.parent.api.v1.chain.get_table_rows({
            code: "sysio",
            scope: "sysio",
            table: "rexpool",
            type: REXState
        });
        return response.rows[0];
    }
}
