// @wireio/sdk-core@1.0.18/lib/api/Client.js
import { FetchProvider } from "./Provider";
import { abiDecode } from "../serializer/Decoder";
import { ChainAPI } from "./v1/Chain";
import { HistoryAPI } from "./v1/History";
import { HistoryAPIv2 } from "./v2/History";
import { StateAPIv2 } from "./v2/State";
import { StatsAPIv2 } from "./v2/Stats";
export { ChainAPI, HistoryAPI };
import { ABI } from "../chain/Abi";
import { Action } from "../chain/Action";
import { KeyType } from "../chain/KeyType";
import { Signature } from "../chain/Signature";
import { Transaction, SignedTransaction, PackedTransaction, CompressionType } from "../chain/Transaction";
export class APIError extends Error {
    static __className = "APIError";
    static formatError(error) {
        if (error.what === "unspecified" &&
            error.details[0].file &&
            error.details[0].file === "http_plugin.cpp" &&
            error.details[0].message.slice(0, 11) === "unknown key") {
            // fix cryptic error messages from nodeop for missing accounts
            return "Account not found";
        }
        else if (error.what === "unspecified" &&
            error.details &&
            error.details.length > 0) {
            return error.details[0].message;
        }
        else if (error.what && error.what.length > 0) {
            return error.what;
        }
        else {
            return "Unknown API error";
        }
    }
    /** The path to the API that failed, e.g. `/v1/chain/get_info`. */
    path;
    /** The full response from the API that failed. */
    response;
    constructor(path, response) {
        let message;
        if (response.json && response.json.error) {
            message = `${APIError.formatError(response.json.error)} at ${path}`;
        }
        else {
            message = `HTTP ${response.status} at ${path}`;
        }
        super(message);
        this.path = path;
        this.response = response;
    }
    /** The nodeop error object. */
    get error() {
        const { json } = this.response;
        return (json ? json.error : undefined);
    }
    /** The nodeop error name, e.g. `tx_net_usage_exceeded` */
    get name() {
        const { error } = this;
        return error ? error.name : "unspecified";
    }
    /** The nodeop error code, e.g. `3080002`. */
    get code() {
        const { error } = this;
        return error ? error.code : 0;
    }
    /** List of exceptions, if any. */
    get details() {
        const { error } = this;
        return error ? error.details : [];
    }
}
export class APIClient {
    static __className = "APIClient";
    v1Provider;
    v2Provider;
    _signer;
    get signer() {
        return this._signer;
    }
    constructor(options) {
        if (options.provider)
            this.v1Provider = options.provider;
        else if (options.url)
            this.v1Provider = new FetchProvider(options.url, options);
        else
            throw new Error("Missing v1 url or provider");
        if (options.hyperionUrl && options.hyperionUrl != "")
            this.v2Provider = new FetchProvider(options.hyperionUrl, options);
        if (options.signer)
            this._signer = options.signer;
    }
    v1 = {
        chain: new ChainAPI(this),
        history: new HistoryAPI(this)
    };
    v2 = {
        history: new HistoryAPIv2(this),
        state: new StateAPIv2(this),
        stats: new StatsAPIv2(this)
    };
    async call(args) {
        const isV2 = args.path.startsWith("/v2/");
        if (isV2 && !this.v2Provider)
            throw new Error("HyperionAPI requires a v2 provider");
        const response = isV2 && this.v2Provider
            ? await this.v2Provider.call(args)
            : await this.v1Provider.call(args);
        const { json } = response;
        if (Math.floor(response.status / 100) !== 2 ||
            (json && typeof json.error === "object"))
            throw new APIError(args.path, response);
        if (args.responseType)
            return abiDecode({ type: args.responseType, object: response.json });
        return response.json || response.text;
    }
    setSigner(signer) {
        this._signer = signer;
    }
    /**
     * Fetches rows based on the provided options.
     * @param options The options for fetching rows.
     * @returns A Promise that resolves to the fetched rows.
     */
    async getRows(options) {
        try {
            // Trim string fields
            for (const key in options) {
                if (typeof options[key] === "string") {
                    options[key] = options[key].trim();
                }
            }
            if (!options.key_type)
                options.key_type = "i64"; // default to int keytype
            const result = await this.v1.chain.get_table_rows({
                json: true,
                code: options.contract,
                scope: options.scope !== undefined ? options.scope : options.contract,
                table: options.table,
                index_position: options.index_position,
                limit: options.limit != null ? options.limit : 50,
                lower_bound: options.lower_bound,
                upper_bound: options.upper_bound,
                key_type: options.key_type,
                reverse: options.reverse !== undefined ? options.reverse : true
            });
            return result;
        }
        catch (e) {
            if (e instanceof APIError) {
                const details = e.details || [];
                const msg = details
                    .map((d) => d.message.replace(/Error:/g, "").trim())
                    .join(", ");
                throw new Error(msg);
            }
            else
                throw e;
        }
    }
    async pushTransaction(action, opts) {
        if (!this.signer)
            throw new Error("Signer is required to push transaction");
        try {
            const signedTrx = await this.buildSignedTransaction(action, opts);
            const packedTrx = PackedTransaction.fromSigned(signedTrx, CompressionType.none);
            const pushTrx = await this.v1.chain.push_transaction(packedTrx);
            if (opts && opts.wait_final) {
                const result = await pushTrx;
                await this.awaitBlock(result.processed.block_num);
            }
            return pushTrx;
        }
        catch (e) {
            if (e instanceof APIError) {
                const details = e.details || [];
                const msg = details
                    .map((d) => d.message
                    .replace(/Error:/g, "")
                    .replace(/assertion failure with message: /g, "")
                    .trim())
                    .join(", ");
                throw new Error(msg);
            }
            else {
                const msg = e.message.replace(/Error:/g, "") || e;
                throw new Error(msg);
            }
        }
    }
    async buildSignedTransaction(action, opts) {
        if (!this.signer)
            throw new Error("No signer function provided in APIClient options");
        const keyType = this.signer.pubKey.type;
        const actions = await this.anyToAction(action);
        const info = await this.v1.chain.get_info();
        const header = info.getTransactionHeader();
        const transaction = Transaction.from({
            ...header,
            actions,
            context_free_actions: opts && opts.context_free_actions ? opts.context_free_actions : []
        });
        // If ED signer, add the public key to the transaction extensions
        if (keyType === KeyType.ED)
            transaction.extPubKey(this.signer.pubKey);
        const { msgBytes } = transaction.signingDigest(info.chain_id, keyType);
        const sigBytes = await this.signer.sign(msgBytes).catch(err => {
            throw new Error(err);
        });
        const signature = Signature.fromRaw(sigBytes, keyType);
        return SignedTransaction.from({ ...transaction, signatures: [signature] });
    }
    async anyToAction(action) {
        if (!Array.isArray(action))
            action = [action];
        const actions = [];
        const knownAbis = new Map();
        for (const act of action) {
            if (!knownAbis.has(act.account)) {
                const abi_res = await this.v1.chain.get_abi(act.account);
                knownAbis.set(act.account, ABI.from(abi_res.abi));
            }
            actions.push(Action.from(act, knownAbis.get(act.account)));
        }
        return actions;
    }
    /**
     * Waits until the specified block number is irreversible.
     * @param blocknum The block number to wait for.
     * @param onProgress Optional callback to report progress (blocksLeft, secondsLeft).
     * @param timeoutMs Optional timeout in milliseconds (default: 2 minutes).
     * @throws Error if the timeout is reached before the block becomes irreversible.
     */
    async awaitBlock(blocknum, onProgress, timeoutMs = 120_000, // 2 minutes
    pollInterval = 3000 // 3 seconds 6 blocks half of 12 block round
    ) {
        const start = Date.now();
        let waiting = true;
        while (waiting) {
            const info = await this.v1.chain.get_info();
            const libNum = info.last_irreversible_block_num.toNumber();
            const blocksLeft = Math.max(0, +blocknum - libNum);
            const secondsLeft = blocksLeft * 0.5; // assuming 0.5s per block
            if (onProgress)
                onProgress(blocksLeft, secondsLeft);
            if (libNum >= +blocknum) {
                waiting = false;
                return;
            }
            if (Date.now() - start > timeoutMs) {
                throw new Error(`Timeout waiting for block ${blocknum} to become irreversible`);
            }
            await new Promise(res => setTimeout(res, pollInterval));
        }
    }
}
