// @wireio/sdk-core@1.0.18/lib/api/Provider.js
/** Default provider that uses the Fetch API to call a single node. */
export class FetchProvider {
    url;
    fetch;
    headers = {};
    constructor(url, options = {}) {
        url = url.trim();
        if (url.endsWith("/"))
            url = url.slice(0, -1);
        this.url = url;
        if (options.headers) {
            this.headers = options.headers;
        }
        if (!options.fetch) {
            if (typeof window !== "undefined" && window.fetch) {
                this.fetch = window.fetch.bind(window);
            }
            else if (typeof global !== "undefined" && global.fetch) {
                this.fetch = global.fetch.bind(global);
            }
            else {
                throw new Error("Missing fetch");
            }
        }
        else {
            this.fetch = options.fetch;
        }
    }
    async call(args) {
        const method = args.method || "POST";
        let url = this.url + args.path;
        const headers = { ...this.headers, ...args.headers };
        // Filter out undefined, null, and empty string values
        const params = args.params
            ? Object.entries(args.params)
                .filter(([_, value]) => value != null && value !== "")
                .reduce((acc, [key, value]) => ({ ...acc, [key]: value }), {})
            : {};
        let body;
        // If GET method, convert params to query string
        if (method === "GET" && Object.keys(params).length > 0) {
            url +=
                "?" + new URLSearchParams(params).toString();
        }
        else if (Object.keys(params).length > 0) {
            body = JSON.stringify(params);
        }
        const response = await this.fetch(url, {
            method,
            body: method === "GET" ? undefined : body,
            headers
        });
        const text = await response.text();
        let json;
        try {
            json = JSON.parse(text);
        }
        catch {
            // Ignore JSON parse errors
        }
        return {
            headers: Object.fromEntries(response.headers.entries()),
            status: response.status,
            json,
            text
        };
    }
}
