// @wireio/sdk-core@1.0.18/lib/api/v2/History.js
export class HistoryAPIv2 {
    client;
    constructor(client) {
        this.client = client;
    }
    /**
     * Fetch a transaction by ID
     * @param id - Transaction ID
     * @param block_hint - Optional block hint for performance
     */
    async get_transaction(id, block_hint) {
        return this.client.call({
            method: "GET",
            path: "/v2/history/get_transaction",
            params: { id, block_hint }
        });
    }
    /**
     * Fetch actions based on specified parameters
     * @param params - Query parameters for fetching actions
     * @returns A promise that resolves to a GetActionsResponse object
     */
    async get_actions(params) {
        return this.client.call({
            method: "GET",
            path: `/v2/history/get_actions`,
            params: params || {}
        });
    }
    /**
     * Fetch all accounts created by a specified account
     * @param params - Query parameters for fetching created accounts
     * @returns A promise that resolves to a GetCreatedAccountsResponse object containing an array of CreatedAccounts
     */
    async get_created_accounts(params) {
        return this.client.call({
            method: "GET",
            path: `/v2/history/get_created_accounts`,
            params
        });
    }
}
