// @wireio/sdk-core@1.0.18/lib/api/v2/Types.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Signature, Struct } from "../../chain";
let AccountRamDelta = class AccountRamDelta extends Struct {
};
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], AccountRamDelta.prototype, "account", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], AccountRamDelta.prototype, "delta", void 0);
AccountRamDelta = __decorate([
    Struct.type("account_ram_delta")
], AccountRamDelta);
export { AccountRamDelta };
let AuthSequence = class AuthSequence extends Struct {
};
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], AuthSequence.prototype, "account", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], AuthSequence.prototype, "sequence", void 0);
AuthSequence = __decorate([
    Struct.type("auth_sequence")
], AuthSequence);
export { AuthSequence };
let Receipt = class Receipt extends Struct {
};
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], Receipt.prototype, "receiver", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], Receipt.prototype, "global_sequence", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], Receipt.prototype, "recv_sequence", void 0);
__decorate([
    Struct.field(AuthSequence, { array: true }),
    __metadata("design:type", Array)
], Receipt.prototype, "auth_sequence", void 0);
Receipt = __decorate([
    Struct.type("receipt")
], Receipt);
export { Receipt };
let GetTransactionResponseAction = class GetTransactionResponseAction extends Struct {
};
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], GetTransactionResponseAction.prototype, "action_ordinal", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], GetTransactionResponseAction.prototype, "creator_action_ordinal", void 0);
__decorate([
    Struct.field("AnyAction"),
    __metadata("design:type", Object)
], GetTransactionResponseAction.prototype, "act", void 0);
__decorate([
    Struct.field(AccountRamDelta, { array: true }),
    __metadata("design:type", Array)
], GetTransactionResponseAction.prototype, "account_ram_deltas", void 0);
__decorate([
    Struct.field(Signature, { array: true }),
    __metadata("design:type", Array)
], GetTransactionResponseAction.prototype, "signatures", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], GetTransactionResponseAction.prototype, "@timestamp", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], GetTransactionResponseAction.prototype, "block_num", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], GetTransactionResponseAction.prototype, "block_id", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], GetTransactionResponseAction.prototype, "producer", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], GetTransactionResponseAction.prototype, "trx_id", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], GetTransactionResponseAction.prototype, "global_sequence", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], GetTransactionResponseAction.prototype, "cpu_usage_us", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], GetTransactionResponseAction.prototype, "net_usage_words", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], GetTransactionResponseAction.prototype, "code_sequence", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], GetTransactionResponseAction.prototype, "abi_sequence", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], GetTransactionResponseAction.prototype, "act_digest", void 0);
__decorate([
    Struct.field(Receipt, { array: true }),
    __metadata("design:type", Array)
], GetTransactionResponseAction.prototype, "receipts", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], GetTransactionResponseAction.prototype, "timestamp", void 0);
GetTransactionResponseAction = __decorate([
    Struct.type("get_transaction_response_action")
], GetTransactionResponseAction);
export { GetTransactionResponseAction };
let GetTransactionResponse = class GetTransactionResponse extends Struct {
};
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], GetTransactionResponse.prototype, "query_time_ms", void 0);
__decorate([
    Struct.field("boolean"),
    __metadata("design:type", Boolean)
], GetTransactionResponse.prototype, "executed", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], GetTransactionResponse.prototype, "trx_id", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], GetTransactionResponse.prototype, "lib", void 0);
__decorate([
    Struct.field("boolean"),
    __metadata("design:type", Boolean)
], GetTransactionResponse.prototype, "cached_lib", void 0);
__decorate([
    Struct.field(GetTransactionResponseAction, { array: true }),
    __metadata("design:type", Array)
], GetTransactionResponse.prototype, "actions", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], GetTransactionResponse.prototype, "last_indexed_block", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], GetTransactionResponse.prototype, "last_indexed_block_time", void 0);
GetTransactionResponse = __decorate([
    Struct.type("get_transaction_response_v2")
], GetTransactionResponse);
export { GetTransactionResponse };
let ActionDataHeader = class ActionDataHeader extends Struct {
};
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], ActionDataHeader.prototype, "timestamp", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], ActionDataHeader.prototype, "producer", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], ActionDataHeader.prototype, "confirmed", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], ActionDataHeader.prototype, "previous", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], ActionDataHeader.prototype, "transaction_mroot", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], ActionDataHeader.prototype, "action_mroot", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], ActionDataHeader.prototype, "schedule_version", void 0);
__decorate([
    Struct.field("any"),
    __metadata("design:type", Object)
], ActionDataHeader.prototype, "new_producers", void 0);
ActionDataHeader = __decorate([
    Struct.type("action_data_header")
], ActionDataHeader);
export { ActionDataHeader };
let ActionObject = class ActionObject extends Struct {
};
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], ActionObject.prototype, "action_ordinal", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], ActionObject.prototype, "creator_action_ordinal", void 0);
__decorate([
    Struct.field("AnyAction"),
    __metadata("design:type", Object)
], ActionObject.prototype, "act", void 0);
__decorate([
    Struct.field(AccountRamDelta, { array: true }),
    __metadata("design:type", Array)
], ActionObject.prototype, "account_ram_deltas", void 0);
__decorate([
    Struct.field(Signature, { array: true }),
    __metadata("design:type", Array)
], ActionObject.prototype, "signatures", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], ActionObject.prototype, "@timestamp", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], ActionObject.prototype, "timestamp", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], ActionObject.prototype, "block_num", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], ActionObject.prototype, "block_id", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], ActionObject.prototype, "trx_id", void 0);
__decorate([
    Struct.field(Receipt, { array: true }),
    __metadata("design:type", Array)
], ActionObject.prototype, "receipts", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], ActionObject.prototype, "cpu_usage_us", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], ActionObject.prototype, "global_sequence", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], ActionObject.prototype, "producer", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], ActionObject.prototype, "net_usage_words", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], ActionObject.prototype, "code_sequence", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], ActionObject.prototype, "abi_sequence", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], ActionObject.prototype, "act_digest", void 0);
ActionObject = __decorate([
    Struct.type("get_actions_response_action")
], ActionObject);
export { ActionObject };
let GetActionsTotal = class GetActionsTotal extends Struct {
};
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], GetActionsTotal.prototype, "value", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], GetActionsTotal.prototype, "relation", void 0);
GetActionsTotal = __decorate([
    Struct.type("get_actions_total")
], GetActionsTotal);
export { GetActionsTotal };
let GetActionsResponse = class GetActionsResponse extends Struct {
};
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], GetActionsResponse.prototype, "query_time_ms", void 0);
__decorate([
    Struct.field("boolean"),
    __metadata("design:type", Boolean)
], GetActionsResponse.prototype, "cached", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], GetActionsResponse.prototype, "lib", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], GetActionsResponse.prototype, "last_indexed_block", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], GetActionsResponse.prototype, "last_indexed_block_time", void 0);
__decorate([
    Struct.field(GetActionsTotal),
    __metadata("design:type", GetActionsTotal)
], GetActionsResponse.prototype, "total", void 0);
__decorate([
    Struct.field(ActionObject, { array: true }),
    __metadata("design:type", Array)
], GetActionsResponse.prototype, "actions", void 0);
GetActionsResponse = __decorate([
    Struct.type("get_actions_response")
], GetActionsResponse);
export { GetActionsResponse };
let HealthService = class HealthService extends Struct {
};
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], HealthService.prototype, "service", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], HealthService.prototype, "status", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], HealthService.prototype, "time", void 0);
__decorate([
    Struct.field("any", { optional: true }),
    __metadata("design:type", Object)
], HealthService.prototype, "service_data", void 0);
HealthService = __decorate([
    Struct.type("health_service")
], HealthService);
export { HealthService };
let HealthFeaturesStreaming = class HealthFeaturesStreaming extends Struct {
};
__decorate([
    Struct.field("boolean"),
    __metadata("design:type", Boolean)
], HealthFeaturesStreaming.prototype, "enable", void 0);
__decorate([
    Struct.field("boolean"),
    __metadata("design:type", Boolean)
], HealthFeaturesStreaming.prototype, "traces", void 0);
__decorate([
    Struct.field("boolean"),
    __metadata("design:type", Boolean)
], HealthFeaturesStreaming.prototype, "deltas", void 0);
HealthFeaturesStreaming = __decorate([
    Struct.type("health_features_streaming")
], HealthFeaturesStreaming);
export { HealthFeaturesStreaming };
let HealthFeaturesTables = class HealthFeaturesTables extends Struct {
};
__decorate([
    Struct.field("boolean"),
    __metadata("design:type", Boolean)
], HealthFeaturesTables.prototype, "proposals", void 0);
__decorate([
    Struct.field("boolean"),
    __metadata("design:type", Boolean)
], HealthFeaturesTables.prototype, "accounts", void 0);
__decorate([
    Struct.field("boolean"),
    __metadata("design:type", Boolean)
], HealthFeaturesTables.prototype, "voters", void 0);
HealthFeaturesTables = __decorate([
    Struct.type("health_features_tables")
], HealthFeaturesTables);
export { HealthFeaturesTables };
let HealthFeatures = class HealthFeatures extends Struct {
};
__decorate([
    Struct.field(HealthFeaturesStreaming),
    __metadata("design:type", HealthFeaturesStreaming)
], HealthFeatures.prototype, "streaming", void 0);
__decorate([
    Struct.field(HealthFeaturesTables),
    __metadata("design:type", HealthFeaturesTables)
], HealthFeatures.prototype, "tables", void 0);
__decorate([
    Struct.field("boolean"),
    __metadata("design:type", Boolean)
], HealthFeatures.prototype, "index_deltas", void 0);
__decorate([
    Struct.field("boolean"),
    __metadata("design:type", Boolean)
], HealthFeatures.prototype, "index_transfer_memo", void 0);
__decorate([
    Struct.field("boolean"),
    __metadata("design:type", Boolean)
], HealthFeatures.prototype, "index_all_deltas", void 0);
__decorate([
    Struct.field("boolean"),
    __metadata("design:type", Boolean)
], HealthFeatures.prototype, "deferred_trx", void 0);
__decorate([
    Struct.field("boolean"),
    __metadata("design:type", Boolean)
], HealthFeatures.prototype, "failed_trx", void 0);
__decorate([
    Struct.field("boolean"),
    __metadata("design:type", Boolean)
], HealthFeatures.prototype, "resource_limits", void 0);
__decorate([
    Struct.field("boolean"),
    __metadata("design:type", Boolean)
], HealthFeatures.prototype, "resource_usage", void 0);
HealthFeatures = __decorate([
    Struct.type("health_features")
], HealthFeatures);
export { HealthFeatures };
let HealthResponse = class HealthResponse extends Struct {
};
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], HealthResponse.prototype, "version", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], HealthResponse.prototype, "version_hash", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], HealthResponse.prototype, "host", void 0);
__decorate([
    Struct.field(HealthService, { array: true }),
    __metadata("design:type", Array)
], HealthResponse.prototype, "health", void 0);
__decorate([
    Struct.field(HealthFeatures),
    __metadata("design:type", HealthFeatures
    // HyperionBaseResponse
    )
], HealthResponse.prototype, "features", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], HealthResponse.prototype, "query_time_ms", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], HealthResponse.prototype, "last_indexed_block", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], HealthResponse.prototype, "last_indexed_block_time", void 0);
HealthResponse = __decorate([
    Struct.type("health_response")
], HealthResponse);
export { HealthResponse };
let ApiUsageTotal = class ApiUsageTotal extends Struct {
};
__decorate([
    Struct.field("any"),
    __metadata("design:type", Object)
], ApiUsageTotal.prototype, "responses", void 0);
ApiUsageTotal = __decorate([
    Struct.type("api_usage_total")
], ApiUsageTotal);
export { ApiUsageTotal };
let ApiUsageBucket = class ApiUsageBucket extends Struct {
};
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], ApiUsageBucket.prototype, "timestamp", void 0);
__decorate([
    Struct.field("any"),
    __metadata("design:type", Object)
], ApiUsageBucket.prototype, "responses", void 0);
ApiUsageBucket = __decorate([
    Struct.type("api_usage_bucket")
], ApiUsageBucket);
export { ApiUsageBucket };
let ApiUsageResponse = class ApiUsageResponse extends Struct {
};
__decorate([
    Struct.field(ApiUsageTotal),
    __metadata("design:type", ApiUsageTotal)
], ApiUsageResponse.prototype, "total", void 0);
__decorate([
    Struct.field(ApiUsageBucket, { array: true }),
    __metadata("design:type", Array)
], ApiUsageResponse.prototype, "buckets", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], ApiUsageResponse.prototype, "query_time_ms", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], ApiUsageResponse.prototype, "last_indexed_block", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], ApiUsageResponse.prototype, "last_indexed_block_time", void 0);
ApiUsageResponse = __decorate([
    Struct.type("api_usage_response")
], ApiUsageResponse);
export { ApiUsageResponse };
let MissedBlocksStats = class MissedBlocksStats extends Struct {
};
__decorate([
    Struct.field("any"),
    __metadata("design:type", Object)
], MissedBlocksStats.prototype, "by_producer", void 0);
MissedBlocksStats = __decorate([
    Struct.type("missed_blocks_stats")
], MissedBlocksStats);
export { MissedBlocksStats };
let MissedBlocksResponse = class MissedBlocksResponse extends Struct {
};
__decorate([
    Struct.field(MissedBlocksStats),
    __metadata("design:type", MissedBlocksStats)
], MissedBlocksResponse.prototype, "stats", void 0);
__decorate([
    Struct.field("any", { array: true }),
    __metadata("design:type", Array)
], MissedBlocksResponse.prototype, "events", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], MissedBlocksResponse.prototype, "query_time_ms", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], MissedBlocksResponse.prototype, "last_indexed_block", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], MissedBlocksResponse.prototype, "last_indexed_block_time", void 0);
MissedBlocksResponse = __decorate([
    Struct.type("missed_blocks_response")
], MissedBlocksResponse);
export { MissedBlocksResponse };
let ResourceUsageStats = class ResourceUsageStats extends Struct {
};
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], ResourceUsageStats.prototype, "count", void 0);
__decorate([
    Struct.field("number", { optional: true }),
    __metadata("design:type", Number)
], ResourceUsageStats.prototype, "min", void 0);
__decorate([
    Struct.field("number", { optional: true }),
    __metadata("design:type", Number)
], ResourceUsageStats.prototype, "max", void 0);
__decorate([
    Struct.field("number", { optional: true }),
    __metadata("design:type", Number)
], ResourceUsageStats.prototype, "avg", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], ResourceUsageStats.prototype, "sum", void 0);
__decorate([
    Struct.field("number", { optional: true }),
    __metadata("design:type", Number)
], ResourceUsageStats.prototype, "sum_of_squares", void 0);
__decorate([
    Struct.field("number", { optional: true }),
    __metadata("design:type", Number)
], ResourceUsageStats.prototype, "variance", void 0);
__decorate([
    Struct.field("number", { optional: true }),
    __metadata("design:type", Number)
], ResourceUsageStats.prototype, "variance_population", void 0);
__decorate([
    Struct.field("number", { optional: true }),
    __metadata("design:type", Number)
], ResourceUsageStats.prototype, "variance_sampling", void 0);
__decorate([
    Struct.field("number", { optional: true }),
    __metadata("design:type", Number)
], ResourceUsageStats.prototype, "std_deviation", void 0);
__decorate([
    Struct.field("number", { optional: true }),
    __metadata("design:type", Number)
], ResourceUsageStats.prototype, "std_deviation_population", void 0);
__decorate([
    Struct.field("number", { optional: true }),
    __metadata("design:type", Number)
], ResourceUsageStats.prototype, "std_deviation_sampling", void 0);
__decorate([
    Struct.field("number", { optional: true }),
    __metadata("design:type", Number)
], ResourceUsageStats.prototype, "std_deviation_bounds_upper", void 0);
__decorate([
    Struct.field("number", { optional: true }),
    __metadata("design:type", Number)
], ResourceUsageStats.prototype, "std_deviation_bounds_lower", void 0);
__decorate([
    Struct.field("number", { optional: true }),
    __metadata("design:type", Number)
], ResourceUsageStats.prototype, "std_deviation_bounds_upper_population", void 0);
__decorate([
    Struct.field("number", { optional: true }),
    __metadata("design:type", Number)
], ResourceUsageStats.prototype, "std_deviation_bounds_lower_population", void 0);
__decorate([
    Struct.field("number", { optional: true }),
    __metadata("design:type", Number)
], ResourceUsageStats.prototype, "std_deviation_bounds_upper_sampling", void 0);
__decorate([
    Struct.field("number", { optional: true }),
    __metadata("design:type", Number)
], ResourceUsageStats.prototype, "std_deviation_bounds_lower_sampling", void 0);
ResourceUsageStats = __decorate([
    Struct.type("resource_usage_stats")
], ResourceUsageStats);
export { ResourceUsageStats };
let ResourceUsagePercentiles = class ResourceUsagePercentiles extends Struct {
};
__decorate([
    Struct.field("number", { optional: true }),
    __metadata("design:type", Number)
], ResourceUsagePercentiles.prototype, "1.0", void 0);
__decorate([
    Struct.field("number", { optional: true }),
    __metadata("design:type", Number)
], ResourceUsagePercentiles.prototype, "5.0", void 0);
__decorate([
    Struct.field("number", { optional: true }),
    __metadata("design:type", Number)
], ResourceUsagePercentiles.prototype, "25.0", void 0);
__decorate([
    Struct.field("number", { optional: true }),
    __metadata("design:type", Number)
], ResourceUsagePercentiles.prototype, "50.0", void 0);
__decorate([
    Struct.field("number", { optional: true }),
    __metadata("design:type", Number)
], ResourceUsagePercentiles.prototype, "75.0", void 0);
__decorate([
    Struct.field("number", { optional: true }),
    __metadata("design:type", Number)
], ResourceUsagePercentiles.prototype, "95.0", void 0);
__decorate([
    Struct.field("number", { optional: true }),
    __metadata("design:type", Number)
], ResourceUsagePercentiles.prototype, "99.0", void 0);
ResourceUsagePercentiles = __decorate([
    Struct.type("resource_usage_percentiles")
], ResourceUsagePercentiles);
export { ResourceUsagePercentiles };
let ResourceUsage = class ResourceUsage extends Struct {
};
__decorate([
    Struct.field(ResourceUsageStats),
    __metadata("design:type", ResourceUsageStats)
], ResourceUsage.prototype, "stats", void 0);
__decorate([
    Struct.field(ResourceUsagePercentiles),
    __metadata("design:type", ResourceUsagePercentiles)
], ResourceUsage.prototype, "percentiles", void 0);
ResourceUsage = __decorate([
    Struct.type("resource_usage")
], ResourceUsage);
export { ResourceUsage };
let GetResourceUsageResponse = class GetResourceUsageResponse extends Struct {
};
__decorate([
    Struct.field(ResourceUsage),
    __metadata("design:type", ResourceUsage)
], GetResourceUsageResponse.prototype, "cpu", void 0);
__decorate([
    Struct.field(ResourceUsage),
    __metadata("design:type", ResourceUsage)
], GetResourceUsageResponse.prototype, "net", void 0);
__decorate([
    Struct.field("boolean"),
    __metadata("design:type", Boolean)
], GetResourceUsageResponse.prototype, "cached", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], GetResourceUsageResponse.prototype, "query_time_ms", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], GetResourceUsageResponse.prototype, "last_indexed_block", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], GetResourceUsageResponse.prototype, "last_indexed_block_time", void 0);
GetResourceUsageResponse = __decorate([
    Struct.type("get_resource_usage_response")
], GetResourceUsageResponse);
export { GetResourceUsageResponse };
let Permission = class Permission extends Struct {
};
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], Permission.prototype, "perm_name", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], Permission.prototype, "parent", void 0);
__decorate([
    Struct.field("any"),
    __metadata("design:type", Object)
], Permission.prototype, "required_auth", void 0);
Permission = __decorate([
    Struct.type("permission")
], Permission);
export { Permission };
let Limit = class Limit extends Struct {
};
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], Limit.prototype, "used", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], Limit.prototype, "available", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], Limit.prototype, "max", void 0);
Limit = __decorate([
    Struct.type("limit")
], Limit);
export { Limit };
let TotalResources = class TotalResources extends Struct {
};
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], TotalResources.prototype, "owner", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], TotalResources.prototype, "net_weight", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], TotalResources.prototype, "cpu_weight", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], TotalResources.prototype, "ram_bytes", void 0);
TotalResources = __decorate([
    Struct.type("total_resources")
], TotalResources);
export { TotalResources };
let SelfDelegatedBandwidth = class SelfDelegatedBandwidth extends Struct {
};
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], SelfDelegatedBandwidth.prototype, "from", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], SelfDelegatedBandwidth.prototype, "to", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], SelfDelegatedBandwidth.prototype, "net_weight", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], SelfDelegatedBandwidth.prototype, "cpu_weight", void 0);
SelfDelegatedBandwidth = __decorate([
    Struct.type("self_delegated_bandwidth")
], SelfDelegatedBandwidth);
export { SelfDelegatedBandwidth };
let VoterInfo = class VoterInfo extends Struct {
};
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], VoterInfo.prototype, "owner", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], VoterInfo.prototype, "proxy", void 0);
__decorate([
    Struct.field("string[]"),
    __metadata("design:type", Array)
], VoterInfo.prototype, "producers", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], VoterInfo.prototype, "staked", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], VoterInfo.prototype, "last_vote_weight", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], VoterInfo.prototype, "proxied_vote_weight", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], VoterInfo.prototype, "is_proxy", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], VoterInfo.prototype, "flags1", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], VoterInfo.prototype, "reserved2", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], VoterInfo.prototype, "reserved3", void 0);
VoterInfo = __decorate([
    Struct.type("voter_info")
], VoterInfo);
export { VoterInfo };
let Account = class Account extends Struct {
};
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], Account.prototype, "account_name", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], Account.prototype, "head_block_num", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], Account.prototype, "head_block_time", void 0);
__decorate([
    Struct.field("boolean"),
    __metadata("design:type", Boolean)
], Account.prototype, "privileged", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], Account.prototype, "last_code_update", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], Account.prototype, "created", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], Account.prototype, "core_liquid_balance", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], Account.prototype, "ram_quota", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], Account.prototype, "net_weight", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], Account.prototype, "cpu_weight", void 0);
__decorate([
    Struct.field(Limit),
    __metadata("design:type", Limit)
], Account.prototype, "net_limit", void 0);
__decorate([
    Struct.field(Limit),
    __metadata("design:type", Limit)
], Account.prototype, "cpu_limit", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], Account.prototype, "ram_usage", void 0);
__decorate([
    Struct.field(Permission, { array: true }),
    __metadata("design:type", Array)
], Account.prototype, "permissions", void 0);
__decorate([
    Struct.field(TotalResources),
    __metadata("design:type", TotalResources)
], Account.prototype, "total_resources", void 0);
__decorate([
    Struct.field(SelfDelegatedBandwidth),
    __metadata("design:type", SelfDelegatedBandwidth)
], Account.prototype, "self_delegated_bandwidth", void 0);
__decorate([
    Struct.field("any", { optional: true }),
    __metadata("design:type", Object)
], Account.prototype, "refund_request", void 0);
__decorate([
    Struct.field(VoterInfo),
    __metadata("design:type", VoterInfo)
], Account.prototype, "voter_info", void 0);
__decorate([
    Struct.field("any", { optional: true }),
    __metadata("design:type", Object)
], Account.prototype, "rex_info", void 0);
__decorate([
    Struct.field(Limit),
    __metadata("design:type", Limit)
], Account.prototype, "subjective_cpu_bill_limit", void 0);
Account = __decorate([
    Struct.type("account")
], Account);
export { Account };
let Link = class Link extends Struct {
};
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], Link.prototype, "timestamp", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], Link.prototype, "permission", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], Link.prototype, "code", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], Link.prototype, "action", void 0);
Link = __decorate([
    Struct.type("link")
], Link);
export { Link };
let Token = class Token extends Struct {
};
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], Token.prototype, "symbol", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], Token.prototype, "precision", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], Token.prototype, "amount", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], Token.prototype, "contract", void 0);
Token = __decorate([
    Struct.type("token")
], Token);
export { Token };
let Action = class Action extends Struct {
};
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], Action.prototype, "@timestamp", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], Action.prototype, "timestamp", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], Action.prototype, "block_num", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], Action.prototype, "block_id", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], Action.prototype, "trx_id", void 0);
__decorate([
    Struct.field("any"),
    __metadata("design:type", Object)
], Action.prototype, "act", void 0);
__decorate([
    Struct.field(Receipt, { array: true }),
    __metadata("design:type", Array)
], Action.prototype, "receipts", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], Action.prototype, "cpu_usage_us", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], Action.prototype, "net_usage_words", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], Action.prototype, "global_sequence", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], Action.prototype, "producer", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], Action.prototype, "action_ordinal", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], Action.prototype, "creator_action_ordinal", void 0);
__decorate([
    Struct.field(Signature, { array: true }),
    __metadata("design:type", Array)
], Action.prototype, "signatures", void 0);
Action = __decorate([
    Struct.type("action")
], Action);
export { Action };
let GetAccountResponse = class GetAccountResponse extends Struct {
};
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], GetAccountResponse.prototype, "query_time_ms", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], GetAccountResponse.prototype, "last_indexed_block", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], GetAccountResponse.prototype, "last_indexed_block_time", void 0);
__decorate([
    Struct.field(Account),
    __metadata("design:type", Account)
], GetAccountResponse.prototype, "account", void 0);
__decorate([
    Struct.field(Link, { array: true }),
    __metadata("design:type", Array)
], GetAccountResponse.prototype, "links", void 0);
__decorate([
    Struct.field(Token, { array: true }),
    __metadata("design:type", Array)
], GetAccountResponse.prototype, "tokens", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], GetAccountResponse.prototype, "total_actions", void 0);
__decorate([
    Struct.field(Action, { array: true }),
    __metadata("design:type", Array)
], GetAccountResponse.prototype, "actions", void 0);
GetAccountResponse = __decorate([
    Struct.type("get_account_response")
], GetAccountResponse);
export { GetAccountResponse };
let GetKeyAccountsResponse = class GetKeyAccountsResponse extends Struct {
};
__decorate([
    Struct.field("string[]"),
    __metadata("design:type", Array)
], GetKeyAccountsResponse.prototype, "account_names", void 0);
GetKeyAccountsResponse = __decorate([
    Struct.type("get_key_accounts_response")
], GetKeyAccountsResponse);
export { GetKeyAccountsResponse };
let GetCreatedAccountsResponse = class GetCreatedAccountsResponse extends Struct {
};
__decorate([
    Struct.field("CreatedAccount[]"),
    __metadata("design:type", Array)
], GetCreatedAccountsResponse.prototype, "accounts", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], GetCreatedAccountsResponse.prototype, "query_time_ms", void 0);
__decorate([
    Struct.field("number"),
    __metadata("design:type", Number)
], GetCreatedAccountsResponse.prototype, "last_indexed_block", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], GetCreatedAccountsResponse.prototype, "last_indexed_block_time", void 0);
GetCreatedAccountsResponse = __decorate([
    Struct.type("get_account_response")
], GetCreatedAccountsResponse);
export { GetCreatedAccountsResponse };
