// @wireio/sdk-core@1.0.18/lib/p2p/Types.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var BlockHeader_1;
import { BlockId } from "../chain/BlockId";
import { Bytes } from "../chain/Bytes";
import { Checksum256 } from "../chain/Checksum";
import { Int16, Int64, UInt16, UInt32, UInt8, VarUInt } from "../chain/Integer";
import { Name } from "../chain/Name";
import { PackedTransaction } from "../chain/Transaction";
import { PublicKey } from "../chain/PublicKey";
import { Signature } from "../chain/Signature";
import { Struct } from "../chain/Struct";
import { Variant } from "../chain/Variant";
import { Serializer } from "../serializer/index";
let HandshakeMessage = class HandshakeMessage extends Struct {
};
__decorate([
    Struct.field("uint16"),
    __metadata("design:type", UInt16)
], HandshakeMessage.prototype, "networkVersion", void 0);
__decorate([
    Struct.field("checksum256"),
    __metadata("design:type", Checksum256)
], HandshakeMessage.prototype, "chainId", void 0);
__decorate([
    Struct.field("checksum256"),
    __metadata("design:type", Checksum256)
], HandshakeMessage.prototype, "nodeId", void 0);
__decorate([
    Struct.field("public_key"),
    __metadata("design:type", PublicKey)
], HandshakeMessage.prototype, "key", void 0);
__decorate([
    Struct.field("int64"),
    __metadata("design:type", Int64)
], HandshakeMessage.prototype, "time", void 0);
__decorate([
    Struct.field("checksum256"),
    __metadata("design:type", Checksum256)
], HandshakeMessage.prototype, "token", void 0);
__decorate([
    Struct.field("signature"),
    __metadata("design:type", Signature)
], HandshakeMessage.prototype, "sig", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], HandshakeMessage.prototype, "p2pAddress", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], HandshakeMessage.prototype, "lastIrreversibleBlockNumber", void 0);
__decorate([
    Struct.field(BlockId),
    __metadata("design:type", BlockId)
], HandshakeMessage.prototype, "lastIrreversibleBlockId", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], HandshakeMessage.prototype, "headNum", void 0);
__decorate([
    Struct.field(BlockId),
    __metadata("design:type", BlockId)
], HandshakeMessage.prototype, "headId", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], HandshakeMessage.prototype, "os", void 0);
__decorate([
    Struct.field("string"),
    __metadata("design:type", String)
], HandshakeMessage.prototype, "agent", void 0);
__decorate([
    Struct.field("int16"),
    __metadata("design:type", Int16)
], HandshakeMessage.prototype, "generation", void 0);
HandshakeMessage = __decorate([
    Struct.type("handshake_message")
], HandshakeMessage);
export { HandshakeMessage };
let ChainSizeMessage = class ChainSizeMessage extends Struct {
};
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], ChainSizeMessage.prototype, "lastIrreversibleBlockNumber", void 0);
__decorate([
    Struct.field(BlockId),
    __metadata("design:type", BlockId)
], ChainSizeMessage.prototype, "lastIrreversibleBlockId", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], ChainSizeMessage.prototype, "headNum", void 0);
__decorate([
    Struct.field(BlockId),
    __metadata("design:type", BlockId)
], ChainSizeMessage.prototype, "headId", void 0);
ChainSizeMessage = __decorate([
    Struct.type("chain_size_message")
], ChainSizeMessage);
export { ChainSizeMessage };
let GoAwayMessage = class GoAwayMessage extends Struct {
};
__decorate([
    Struct.field("uint8"),
    __metadata("design:type", UInt8)
], GoAwayMessage.prototype, "reason", void 0);
__decorate([
    Struct.field("checksum256"),
    __metadata("design:type", Checksum256)
], GoAwayMessage.prototype, "nodeId", void 0);
GoAwayMessage = __decorate([
    Struct.type("go_away_message")
], GoAwayMessage);
export { GoAwayMessage };
let TimeMessage = class TimeMessage extends Struct {
};
__decorate([
    Struct.field("int64"),
    __metadata("design:type", Int64)
], TimeMessage.prototype, "org", void 0);
__decorate([
    Struct.field("int64"),
    __metadata("design:type", Int64)
], TimeMessage.prototype, "rec", void 0);
__decorate([
    Struct.field("int64"),
    __metadata("design:type", Int64)
], TimeMessage.prototype, "xmt", void 0);
__decorate([
    Struct.field("int64"),
    __metadata("design:type", Int64)
], TimeMessage.prototype, "dst", void 0);
TimeMessage = __decorate([
    Struct.type("time_message")
], TimeMessage);
export { TimeMessage };
let NoticeMessage = class NoticeMessage extends Struct {
};
__decorate([
    Struct.field("checksum256", { array: true }),
    __metadata("design:type", Array)
], NoticeMessage.prototype, "knownTrx", void 0);
__decorate([
    Struct.field(BlockId, { array: true }),
    __metadata("design:type", Array)
], NoticeMessage.prototype, "knownBlocks", void 0);
NoticeMessage = __decorate([
    Struct.type("notice_message")
], NoticeMessage);
export { NoticeMessage };
let RequestMessage = class RequestMessage extends Struct {
};
__decorate([
    Struct.field("checksum256", { array: true }),
    __metadata("design:type", Array)
], RequestMessage.prototype, "reqTrx", void 0);
__decorate([
    Struct.field(BlockId, { array: true }),
    __metadata("design:type", Array)
], RequestMessage.prototype, "reqBlocks", void 0);
RequestMessage = __decorate([
    Struct.type("request_message")
], RequestMessage);
export { RequestMessage };
let SyncRequestMessage = class SyncRequestMessage extends Struct {
};
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], SyncRequestMessage.prototype, "startBlock", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], SyncRequestMessage.prototype, "endBlock", void 0);
SyncRequestMessage = __decorate([
    Struct.type("sync_request_message")
], SyncRequestMessage);
export { SyncRequestMessage };
let NewProducersEntry = class NewProducersEntry extends Struct {
};
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], NewProducersEntry.prototype, "producer_name", void 0);
__decorate([
    Struct.field("public_key"),
    __metadata("design:type", PublicKey)
], NewProducersEntry.prototype, "block_signing_key", void 0);
NewProducersEntry = __decorate([
    Struct.type("new_producers_entry")
], NewProducersEntry);
export { NewProducersEntry };
let NewProducers = class NewProducers extends Struct {
};
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], NewProducers.prototype, "version", void 0);
__decorate([
    Struct.field(NewProducersEntry, { array: true }),
    __metadata("design:type", NewProducersEntry)
], NewProducers.prototype, "producers", void 0);
NewProducers = __decorate([
    Struct.type("new_producers")
], NewProducers);
export { NewProducers };
let BlockExtension = class BlockExtension extends Struct {
};
__decorate([
    Struct.field("uint16"),
    __metadata("design:type", UInt16)
], BlockExtension.prototype, "type", void 0);
__decorate([
    Struct.field("bytes"),
    __metadata("design:type", Bytes)
], BlockExtension.prototype, "data", void 0);
BlockExtension = __decorate([
    Struct.type("block_extension")
], BlockExtension);
export { BlockExtension };
let HeaderExtension = class HeaderExtension extends Struct {
};
__decorate([
    Struct.field("uint16"),
    __metadata("design:type", UInt16)
], HeaderExtension.prototype, "type", void 0);
__decorate([
    Struct.field("bytes"),
    __metadata("design:type", Bytes)
], HeaderExtension.prototype, "data", void 0);
HeaderExtension = __decorate([
    Struct.type("header_extension")
], HeaderExtension);
export { HeaderExtension };
let TrxVariant = class TrxVariant extends Variant {
};
TrxVariant = __decorate([
    Variant.type("trx_variant", [Checksum256, PackedTransaction])
], TrxVariant);
let FullTransactionReceipt = class FullTransactionReceipt extends Struct {
};
__decorate([
    Struct.field(UInt8),
    __metadata("design:type", UInt8)
], FullTransactionReceipt.prototype, "status", void 0);
__decorate([
    Struct.field(UInt32),
    __metadata("design:type", UInt32)
], FullTransactionReceipt.prototype, "cpu_usage_us", void 0);
__decorate([
    Struct.field(VarUInt),
    __metadata("design:type", VarUInt)
], FullTransactionReceipt.prototype, "net_usage_words", void 0);
__decorate([
    Struct.field(TrxVariant),
    __metadata("design:type", TrxVariant)
], FullTransactionReceipt.prototype, "trx", void 0);
FullTransactionReceipt = __decorate([
    Struct.type("full_transaction_receipt")
], FullTransactionReceipt);
export { FullTransactionReceipt };
let BlockHeader = BlockHeader_1 = class BlockHeader extends Struct {
    new_producers;
    get blockNum() {
        return this.previous.blockNum.adding(1);
    }
    get id() {
        const id = Checksum256.hash(Serializer.encode({ object: this, type: BlockHeader_1 }));
        return BlockId.fromBlockChecksum(id, this.blockNum);
    }
};
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], BlockHeader.prototype, "timeSlot", void 0);
__decorate([
    Struct.field("name"),
    __metadata("design:type", Name)
], BlockHeader.prototype, "producer", void 0);
__decorate([
    Struct.field("uint16"),
    __metadata("design:type", UInt16)
], BlockHeader.prototype, "confirmed", void 0);
__decorate([
    Struct.field(BlockId),
    __metadata("design:type", BlockId)
], BlockHeader.prototype, "previous", void 0);
__decorate([
    Struct.field(BlockId),
    __metadata("design:type", BlockId)
], BlockHeader.prototype, "transaction_mroot", void 0);
__decorate([
    Struct.field(BlockId),
    __metadata("design:type", BlockId)
], BlockHeader.prototype, "action_mroot", void 0);
__decorate([
    Struct.field("uint32"),
    __metadata("design:type", UInt32)
], BlockHeader.prototype, "schedule_version", void 0);
__decorate([
    Struct.field(NewProducers, { optional: true }),
    __metadata("design:type", NewProducers)
], BlockHeader.prototype, "new_producers", void 0);
__decorate([
    Struct.field(HeaderExtension, { array: true }),
    __metadata("design:type", Array)
], BlockHeader.prototype, "header_extensions", void 0);
BlockHeader = BlockHeader_1 = __decorate([
    Struct.type("block_header")
], BlockHeader);
export { BlockHeader };
let SignedBlock = class SignedBlock extends BlockHeader {
};
__decorate([
    Struct.field("signature"),
    __metadata("design:type", Signature)
], SignedBlock.prototype, "producer_signature", void 0);
__decorate([
    Struct.field(FullTransactionReceipt, { array: true }),
    __metadata("design:type", Array)
], SignedBlock.prototype, "transactions", void 0);
__decorate([
    Struct.field(BlockExtension, { array: true }),
    __metadata("design:type", Array)
], SignedBlock.prototype, "block_extensions", void 0);
SignedBlock = __decorate([
    Struct.type("signed_block")
], SignedBlock);
export { SignedBlock };
let NetMessage = class NetMessage extends Variant {
};
NetMessage = __decorate([
    Variant.type("net_message", [
        HandshakeMessage,
        ChainSizeMessage,
        GoAwayMessage,
        TimeMessage,
        NoticeMessage,
        RequestMessage,
        SyncRequestMessage,
        SignedBlock,
        PackedTransaction
    ])
], NetMessage);
export { NetMessage };
